{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import csv\n",
    "import re\n",
    "import pandas as pd\n",
    "import collections\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from math import trunc\n",
    "import threading\n",
    "import multiprocessing as mp\n",
    "from Bio import SeqIO\n",
    "import pysam\n",
    "import scipy\n",
    "import math\n",
    "import random\n",
    "import seaborn as sns"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. 3nt periodicity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[E::idx_find_and_load] Could not retrieve index file for '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/1-BAM/BAM/eIF1/eIF1-Ribo_Aligned.sortedByCoord.out.bam'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I\n",
      "II\n",
      "III\n",
      "IV\n",
      "V\n",
      "VI\n",
      "VII\n",
      "VIII\n",
      "IX\n",
      "X\n",
      "XI\n",
      "XII\n",
      "XIII\n",
      "XIV\n",
      "XV\n",
      "XVI\n",
      "Mito\n"
     ]
    }
   ],
   "source": [
    "### Do the following for both Ribo-seq and RNA-seq data\n",
    "\n",
    "sample = 'eIF1'\n",
    "#rep = '1'\n",
    "name = sample# + '-' + rep\n",
    "SeqType = 'Ribo'\n",
    "path = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/1-BAM/BAM/' + sample + '/') ## The pathway of bam file\n",
    "path2 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/1-BAM/BAM_gene/' + SeqType + '/' + sample + '/') ## The pathway to output results\n",
    "\n",
    "# Put the information from the bam file into the csv file\n",
    "\n",
    "# BAM_file = pysam.AlignmentFile((path + sample + '-' + SeqType + '-' + rep + '_Aligned.sortedByCoord.out.bam'), 'rb')\n",
    "BAM_file = pysam.AlignmentFile((path + sample + '-' + SeqType + '_Aligned.sortedByCoord.out.bam'), 'rb')\n",
    "\n",
    "BAM_csv = open((path2 + name + '_bam.csv'),'w') \n",
    "BAM_csv_writer = csv.writer(BAM_csv)\n",
    "BAM_csv_writer.writerow(['gene','ReadsStartPos','cigar','seq'])\n",
    "\n",
    "for line in BAM_file:\n",
    "    BAM_csv_writer.writerow([line.reference_name, line.pos+1, line.cigarstring[0:-1], line.seq])\n",
    "BAM_csv.close()\n",
    "\n",
    "# Assign reads to individual genes\n",
    "\n",
    "BAM_csv = pd.read_csv((path2 + name + '_bam.csv'))\n",
    "BAM_csv.columns = ['seqid','ReadsStartPos','length','seq']\n",
    "BAM_csv['length'] = BAM_csv['seq'].apply(len)\n",
    "BAM_csv['ReadsEndPos'] = BAM_csv['ReadsStartPos'] + BAM_csv['length'] - 1\n",
    "\n",
    "seqid = BAM_csv['seqid'].unique()\n",
    "gene_position = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/GeneMergeUTR_annotation.txt') ## Positional annotation of full-length gene transcripts, including UTRs\n",
    "gene_position['start'] = gene_position['start'] - 20\n",
    "gene_position['stop'] = gene_position['stop'] + 20\n",
    "\n",
    "seqid_annotation = gene_position['chr'].unique()\n",
    "seqid = [i for i in seqid if i in seqid_annotation]\n",
    "\n",
    "ReadsAssignToGene = open((path2 + name + '_bam_AssignGene_transcript±20.csv'),'w')\n",
    "ReadsAssignToGene_writer = csv.writer(ReadsAssignToGene)\n",
    "ReadsAssignToGene_writer.writerow(['seqid','ReadsStartPos','ReadsEndPos','length','gene','strand','start','stop'])\n",
    "\n",
    "for i in seqid:\n",
    "    print(i)\n",
    "    gene_position_subset = gene_position[gene_position['chr'] == i].reset_index()\n",
    "    BAM_csv_subset = BAM_csv[BAM_csv['seqid'] == i].reset_index()\n",
    "    for n in range(len(gene_position_subset['chr'])):\n",
    "        start = gene_position_subset['start'][n]\n",
    "        stop = gene_position_subset['stop'][n]\n",
    "        gene = gene_position_subset['gene_id'][n]\n",
    "        strand = gene_position_subset['strand'][n]\n",
    "        BAM_csv_subset_pos = BAM_csv_subset[(BAM_csv_subset['ReadsStartPos'] >= start) & (BAM_csv_subset['ReadsEndPos'] <= stop)].reset_index()\n",
    "        for m in range(len(BAM_csv_subset_pos['seqid'])):\n",
    "            length = BAM_csv_subset_pos['length'][m]\n",
    "            ReadsStartPos = BAM_csv_subset_pos['ReadsStartPos'][m]\n",
    "            ReadsEndPos = BAM_csv_subset_pos['ReadsEndPos'][m]\n",
    "            ReadsAssignToGene_writer.writerow([i,ReadsStartPos,ReadsEndPos,length,gene,strand,start,stop])\n",
    "ReadsAssignToGene.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. All genes: 28nt reads in Ribo-seq and all length reads in RNA-seq\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### align all genes at start or stop codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_242461/2363197165.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  GenomeAnnotation_subset['transcript_id'][i] = pattern.findall(GenomeAnnotation_subset['transcript_id'][i].split(';')[0])[0]\n"
     ]
    }
   ],
   "source": [
    "GenomeAnnotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/Saccharomyces_cerevisiae.R64-1-1.107.gtf',skiprows=5,\\\n",
    "    names=['chr', 'DataBase', 'type','start','stop','6','strand','8','transcript_id'])\n",
    "GenomeAnnotation = GenomeAnnotation[['chr','type','start','stop','strand','transcript_id']]\n",
    "GenomeAnnotation_subset = GenomeAnnotation[(GenomeAnnotation['type'] == 'start_codon')|(GenomeAnnotation['type'] == 'stop_codon')]\n",
    "GenomeAnnotation_subset = GenomeAnnotation_subset.reset_index().drop(['index'], axis = 1)\n",
    "\n",
    "pattern = re.compile('\"(.*)\"')\n",
    "for i in range(len(GenomeAnnotation_subset['transcript_id'])):\n",
    "    GenomeAnnotation_subset['transcript_id'][i] = pattern.findall(GenomeAnnotation_subset['transcript_id'][i].split(';')[0])[0]\n",
    "\n",
    "StartCodonPos = GenomeAnnotation_subset[GenomeAnnotation_subset['type'] == 'start_codon'].reset_index().drop(['index'], axis = 1)\n",
    "StartCodonPos = StartCodonPos[['start','stop','strand','transcript_id']]\n",
    "StartCodonPos.columns = ['StartCodonStart','StartCodonStop','strand','gene']\n",
    "StopCodonPos = GenomeAnnotation_subset[GenomeAnnotation_subset['type'] == 'stop_codon'].reset_index().drop(['index'], axis = 1)\n",
    "StopCodonPos = StopCodonPos[['start','stop','strand','transcript_id']]\n",
    "StopCodonPos.columns = ['StopCodonStart','StopCodonStop','strand','gene']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Do the following for both Ribo-seq and RNA-seq data\n",
    "\n",
    "sample = 'HO'\n",
    "#rep = '1'\n",
    "name = sample# + '-' + rep\n",
    "SeqType = 'RNA'\n",
    "path1 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/1-BAM/BAM_gene/' + SeqType + '/' + sample + '/') ## The pathway of bam_AssignGene_transcript±20.csv\n",
    "path2 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/2-TripletPeriodicity/Periodicity_AlignAtStartStopCodon/' + SeqType + '/' + sample + '/') ## The pathway to output results\n",
    "\n",
    "# Judging the 5'/3' ends of reads relative to the start/stop codon\n",
    "\n",
    "ReadsAssignToGene = pd.read_csv((path1 + name + '_bam_AssignGene_transcript±20.csv'))\n",
    "ReadsAssignToGene = ReadsAssignToGene[(ReadsAssignToGene['gene'] != 'YJL189W')&(ReadsAssignToGene['gene'] != 'YLR075W')].reset_index(drop = True)\n",
    "ReadsAssignToGene = pd.merge(ReadsAssignToGene,StartCodonPos,on = 'gene')\n",
    "ReadsAssignToGene = pd.merge(ReadsAssignToGene,StopCodonPos,on = 'gene')\n",
    "ReadsAssignToGene = ReadsAssignToGene[['seqid','ReadsStartPos','length','gene','StartCodonStart','StartCodonStop','StopCodonStart','StopCodonStop','strand']]\n",
    "\n",
    "ReadsAssignToGene_1 = ReadsAssignToGene[ReadsAssignToGene['strand'] == '+'].reset_index().drop(['index'], axis = 1)\n",
    "ReadsAssignToGene_1['StartPos_to_StartCodon'] = ReadsAssignToGene_1['ReadsStartPos'] - ReadsAssignToGene_1['StartCodonStart']\n",
    "ReadsAssignToGene_1['StopPos_to_StartCodon'] = ReadsAssignToGene_1['StartPos_to_StartCodon'] + ReadsAssignToGene_1['length'] - 1\n",
    "ReadsAssignToGene_1['StartPos_to_StopCodon'] = ReadsAssignToGene_1['ReadsStartPos'] - ReadsAssignToGene_1['StopCodonStart']\n",
    "ReadsAssignToGene_1['StopPos_to_StopCodon'] = ReadsAssignToGene_1['StartPos_to_StopCodon'] + ReadsAssignToGene_1['length'] - 1\n",
    "ReadsAssignToGene_1 = ReadsAssignToGene_1[['length','gene','strand','StartPos_to_StartCodon','StopPos_to_StartCodon','StartPos_to_StopCodon','StopPos_to_StopCodon']]\n",
    "\n",
    "ReadsAssignToGene_0 = ReadsAssignToGene[ReadsAssignToGene['strand'] == '-'].reset_index().drop(['index'], axis = 1)\n",
    "ReadsAssignToGene_0['ReadsEndPos'] = ReadsAssignToGene_0['ReadsStartPos'] + ReadsAssignToGene_0['length'] - 1\n",
    "ReadsAssignToGene_0['StartPos_to_StartCodon'] = ReadsAssignToGene_0['StartCodonStop'] - ReadsAssignToGene_0['ReadsEndPos']\n",
    "ReadsAssignToGene_0['StopPos_to_StartCodon'] = ReadsAssignToGene_0['StartCodonStop'] - ReadsAssignToGene_0['ReadsStartPos']\n",
    "ReadsAssignToGene_0['StartPos_to_StopCodon'] = ReadsAssignToGene_0['StopCodonStop'] - ReadsAssignToGene_0['ReadsEndPos']\n",
    "ReadsAssignToGene_0['StopPos_to_StopCodon'] = ReadsAssignToGene_0['StopCodonStop'] - ReadsAssignToGene_0['ReadsStartPos']\n",
    "ReadsAssignToGene_0 = ReadsAssignToGene_0[['length','gene','strand','StartPos_to_StartCodon','StopPos_to_StartCodon','StartPos_to_StopCodon','StopPos_to_StopCodon']]\n",
    "\n",
    "ReadsAssignToGene = pd.concat([ReadsAssignToGene_1,ReadsAssignToGene_0])\n",
    "ReadsAssignToGene['frame'] = ReadsAssignToGene['StartPos_to_StartCodon'] % 3\n",
    "ReadsAssignToGene.to_csv((path2 + name + '_ToStartAndStop_Frame.txt'),sep = '\\t')\n",
    "\n",
    "\n",
    "Length_Frame = ReadsAssignToGene.groupby('length')['frame'].value_counts()\n",
    "Length_Frame.to_csv((path2 + name + '_Frame_length.txt'),sep = '\\t')\n",
    "\n",
    "Start_codon_around_count = ReadsAssignToGene.groupby('length')['StartPos_to_StartCodon'].value_counts().to_frame()\n",
    "Start_codon_around_count.columns = ['counts']\n",
    "Start_codon_around_count = Start_codon_around_count.reset_index()\n",
    "Start_codon_around_count['frame'] = Start_codon_around_count['StartPos_to_StartCodon'] % 3\n",
    "Start_codon_around_count.to_csv((path2 + name + '_StartCodonFrame_Length.txt'),sep = '\\t')\n",
    "\n",
    "Start_codon_total_count = ReadsAssignToGene['StartPos_to_StartCodon'].value_counts().to_frame()\n",
    "Start_codon_total_count.columns = ['counts']\n",
    "Start_codon_total_count = Start_codon_total_count.reset_index()\n",
    "Start_codon_total_count.columns = ['StartPos_to_StartCodon','counts']\n",
    "Start_codon_total_count['frame'] = Start_codon_total_count['StartPos_to_StartCodon'] % 3\n",
    "Start_codon_total_count.to_csv((path2 + name + '_StartCodonFrame.txt'),sep = '\\t')\n",
    "\n",
    "Stop_codon_around_count = ReadsAssignToGene.groupby('length')['StartPos_to_StopCodon'].value_counts().to_frame()\n",
    "Stop_codon_around_count.columns = ['counts']\n",
    "Stop_codon_around_count = Stop_codon_around_count.reset_index()\n",
    "Stop_codon_around_count['frame'] = Stop_codon_around_count['StartPos_to_StopCodon'] % 3\n",
    "Stop_codon_around_count.to_csv((path2 + name + '_StopCodonFrame_Length.txt'),sep = '\\t')\n",
    "\n",
    "Stop_codon_total_count = ReadsAssignToGene['StartPos_to_StopCodon'].value_counts().to_frame()\n",
    "Stop_codon_total_count.columns = ['counts']\n",
    "Stop_codon_total_count = Stop_codon_total_count.reset_index()\n",
    "Stop_codon_total_count.columns = ['StartPos_to_StopCodon','counts']\n",
    "Stop_codon_total_count['frame'] = Stop_codon_total_count['StartPos_to_StopCodon'] % 3\n",
    "Stop_codon_total_count.to_csv((path2 + name + '_StopCodonFrame.txt'),sep = '\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "### combine RNA-seq and 28nt Ribo-seq \n",
    "\n",
    "sample = 'HO'\n",
    "rep = '1'\n",
    "name = sample + '-' + rep\n",
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/2-TripletPeriodicity/Periodicity_AlignAtStartStopCodon/'\n",
    "\n",
    "Ribo_Start_codon = pd.read_table((path + 'Ribo/' + sample + '/' + name + '_StartCodonFrame_Length.txt'),usecols = ['length','StartPos_to_StartCodon','counts'])\n",
    "Ribo_Start_codon = Ribo_Start_codon[Ribo_Start_codon['length'] == 28].reset_index(drop = True).drop(['length'], axis = 1)\n",
    "Ribo_Start_codon.columns = ['StartPos_to_StartCodon','Ribo_counts']\n",
    "\n",
    "RNA_Start_codon = pd.read_table((path + 'RNA/' + sample + '/' + name + '_StartCodonFrame.txt'), usecols = ['StartPos_to_StartCodon','counts'])\n",
    "RNA_Start_codon.columns = ['StartPos_to_StartCodon','RNA_counts']\n",
    "\n",
    "Ribo_RNA_Start_codon = pd.merge(Ribo_Start_codon,RNA_Start_codon,on = 'StartPos_to_StartCodon', how = 'outer')\n",
    "Ribo_RNA_Start_codon['frame'] = Ribo_RNA_Start_codon['StartPos_to_StartCodon'] % 3\n",
    "Ribo_RNA_Start_codon.to_csv((path + 'Ribo_RNA_Merge/' + sample + '/' + name + '_StartFrame.txt'), sep = '\\t', na_rep = 0)\n",
    "\n",
    "Ribo_Stop_codon = pd.read_table((path + 'Ribo/' + sample + '/' + name + '_StopCodonFrame_Length.txt'),usecols = ['length','StartPos_to_StopCodon','counts'])\n",
    "Ribo_Stop_codon = Ribo_Stop_codon[Ribo_Stop_codon['length'] == 28].reset_index(drop = True).drop(['length'], axis = 1)\n",
    "Ribo_Stop_codon.columns = ['StartPos_to_StopCodon','Ribo_counts']\n",
    "\n",
    "RNA_Stop_codon = pd.read_table((path + 'RNA/' + sample + '/' + name + '_StopCodonFrame.txt'), usecols = ['StartPos_to_StopCodon','counts'])\n",
    "RNA_Stop_codon.columns = ['StartPos_to_StopCodon','RNA_counts']\n",
    "\n",
    "Ribo_RNA_Stop_codon = pd.merge(Ribo_Stop_codon,RNA_Stop_codon,on = 'StartPos_to_StopCodon', how = 'outer')\n",
    "Ribo_RNA_Stop_codon['frame'] = Ribo_RNA_Stop_codon['StartPos_to_StopCodon'] % 3\n",
    "Ribo_RNA_Stop_codon.to_csv((path + 'Ribo_RNA_Merge/' + sample + '/' + name + '_StopFrame.txt'), sep = '\\t', na_rep = 0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Frame 0/1/2 Fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>frame</th>\n",
       "      <th>counts</th>\n",
       "      <th>fraction</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>2003541</td>\n",
       "      <td>0.344145</td>\n",
       "      <td>HO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1895669</td>\n",
       "      <td>0.325616</td>\n",
       "      <td>HO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>1922575</td>\n",
       "      <td>0.330238</td>\n",
       "      <td>HO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   frame   counts  fraction sample\n",
       "0      0  2003541  0.344145     HO\n",
       "1      1  1895669  0.325616     HO\n",
       "2      2  1922575  0.330238     HO"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample = 'HO'\n",
    "#rep = '1'\n",
    "name = sample# + '-' + rep\n",
    "SeqType = 'RNA'\n",
    "path1 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/2-TripletPeriodicity/Periodicity_AlignAtStartStopCodon/' + SeqType + '/' + sample + '/') ## _StartCodonFrame_Length.txt\n",
    "path2 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/2-TripletPeriodicity/Frame_Fraction/' + SeqType + '/' + sample + '/') ## The pathway to output results\n",
    "\n",
    "StartCodon_Frame = pd.read_table((path1 + name + '_StartCodonFrame_Length.txt'))\n",
    "#StartCodon_Frame = StartCodon_Frame[StartCodon_Frame['length'] == 28]\n",
    "Frame_Fraction = StartCodon_Frame.groupby('frame')['counts'].sum().reset_index()\n",
    "Frame_Fraction['fraction'] = Frame_Fraction['counts'] / Frame_Fraction['counts'].sum()\n",
    "Frame_Fraction['sample'] = sample\n",
    "#Frame_Fraction['rep'] = rep\n",
    "Frame_Fraction.to_csv((path2 + name + '_Frame_Fraction.txt'), sep = '\\t')\n",
    "Frame_Fraction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>frame</th>\n",
       "      <th>counts</th>\n",
       "      <th>fraction</th>\n",
       "      <th>sample</th>\n",
       "      <th>rep</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>86431</td>\n",
       "      <td>0.853336</td>\n",
       "      <td>HO</td>\n",
       "      <td>1</td>\n",
       "      <td>Ribo_Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>6533</td>\n",
       "      <td>0.064501</td>\n",
       "      <td>HO</td>\n",
       "      <td>1</td>\n",
       "      <td>Ribo_Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>8322</td>\n",
       "      <td>0.082163</td>\n",
       "      <td>HO</td>\n",
       "      <td>1</td>\n",
       "      <td>Ribo_Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>270502</td>\n",
       "      <td>0.849628</td>\n",
       "      <td>HO</td>\n",
       "      <td>2</td>\n",
       "      <td>Ribo_Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>20595</td>\n",
       "      <td>0.064687</td>\n",
       "      <td>HO</td>\n",
       "      <td>2</td>\n",
       "      <td>Ribo_Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>115</th>\n",
       "      <td>1</td>\n",
       "      <td>2530998</td>\n",
       "      <td>0.331780</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>1</td>\n",
       "      <td>RNA_Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116</th>\n",
       "      <td>2</td>\n",
       "      <td>2529417</td>\n",
       "      <td>0.331573</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>1</td>\n",
       "      <td>RNA_Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117</th>\n",
       "      <td>0</td>\n",
       "      <td>2694722</td>\n",
       "      <td>0.338078</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>2</td>\n",
       "      <td>RNA_Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>118</th>\n",
       "      <td>1</td>\n",
       "      <td>2641394</td>\n",
       "      <td>0.331388</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>2</td>\n",
       "      <td>RNA_Seq</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>119</th>\n",
       "      <td>2</td>\n",
       "      <td>2634594</td>\n",
       "      <td>0.330534</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>2</td>\n",
       "      <td>RNA_Seq</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     frame   counts  fraction sample  rep      Type\n",
       "0        0    86431  0.853336     HO    1  Ribo_Seq\n",
       "1        1     6533  0.064501     HO    1  Ribo_Seq\n",
       "2        2     8322  0.082163     HO    1  Ribo_Seq\n",
       "3        0   270502  0.849628     HO    2  Ribo_Seq\n",
       "4        1    20595  0.064687     HO    2  Ribo_Seq\n",
       "..     ...      ...       ...    ...  ...       ...\n",
       "115      1  2530998  0.331780  eIF4B    1   RNA_Seq\n",
       "116      2  2529417  0.331573  eIF4B    1   RNA_Seq\n",
       "117      0  2694722  0.338078  eIF4B    2   RNA_Seq\n",
       "118      1  2641394  0.331388  eIF4B    2   RNA_Seq\n",
       "119      2  2634594  0.330534  eIF4B    2   RNA_Seq\n",
       "\n",
       "[120 rows x 6 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Ribo-seq and RNA-seq file merge\n",
    "\n",
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/2-TripletPeriodicity/Frame_Fraction/'\n",
    "\n",
    "MutantList = ['HO-1','HO-2','CAF20-1','CAF20-2','EAP1-1','EAP1-2','eIF4E-1','eIF4E-2','eIF4G1-1','eIF4G1-2',\\\n",
    "    'PAB1-1','PAB1-2','eIF1-1','eIF1-2','eIF1A-1','eIF1A-2','eIF4A-1','eIF4A-2','eIF4B-1','eIF4B-2']         \n",
    "\n",
    "for i in MutantList:\n",
    "    locals()[('Ribo_Frame_' + i)] = pd.read_table((path + 'Ribo/' + i[:-2] + '/' + i + '_Frame_Fraction.txt'),usecols = ['frame','counts','fraction','sample','rep'])\n",
    "    locals()[('Ribo_Frame_' + i)]['Type'] = 'Ribo_Seq'\n",
    "    locals()[('RNA_Frame_' + i)] = pd.read_table((path + 'RNA/' + i[:-2] + '/' + i + '_Frame_Fraction.txt'),usecols = ['frame','counts','fraction','sample','rep'])\n",
    "    locals()[('RNA_Frame_' + i)]['Type'] = 'RNA_Seq'\n",
    "\n",
    "\n",
    "RiboAndRNA_Frame = pd.concat([locals()[('Ribo_Frame_HO-1')],locals()[('Ribo_Frame_HO-2')],locals()[('Ribo_Frame_CAF20-1')],locals()[('Ribo_Frame_CAF20-2')],\\\n",
    "    locals()[('Ribo_Frame_EAP1-1')],locals()[('Ribo_Frame_EAP1-2')],locals()[('Ribo_Frame_eIF4E-1')],locals()[('Ribo_Frame_eIF4E-2')],\\\n",
    "        locals()[('Ribo_Frame_eIF4G1-1')],locals()[('Ribo_Frame_eIF4G1-2')],locals()[('Ribo_Frame_PAB1-1')],locals()[('Ribo_Frame_PAB1-2')],\\\n",
    "            locals()[('Ribo_Frame_eIF1-1')],locals()[('Ribo_Frame_eIF1-2')],locals()[('Ribo_Frame_eIF1A-1')],locals()[('Ribo_Frame_eIF1A-2')],\\\n",
    "                locals()[('Ribo_Frame_eIF4A-1')],locals()[('Ribo_Frame_eIF4A-2')],locals()[('Ribo_Frame_eIF4B-1')],locals()[('Ribo_Frame_eIF4B-2')],\\\n",
    "    locals()[('RNA_Frame_HO-1')],locals()[('RNA_Frame_HO-2')],locals()[('RNA_Frame_CAF20-1')],locals()[('RNA_Frame_CAF20-2')],\\\n",
    "        locals()[('RNA_Frame_EAP1-1')],locals()[('RNA_Frame_EAP1-2')],locals()[('RNA_Frame_eIF4E-1')],locals()[('RNA_Frame_eIF4E-2')],\\\n",
    "            locals()[('RNA_Frame_eIF4G1-1')],locals()[('RNA_Frame_eIF4G1-2')],locals()[('RNA_Frame_PAB1-1')],locals()[('RNA_Frame_PAB1-2')],\\\n",
    "                locals()[('RNA_Frame_eIF1-1')],locals()[('RNA_Frame_eIF1-2')],locals()[('RNA_Frame_eIF1A-1')],locals()[('RNA_Frame_eIF1A-2')],\\\n",
    "                    locals()[('RNA_Frame_eIF4A-1')],locals()[('RNA_Frame_eIF4A-2')],locals()[('RNA_Frame_eIF4B-1')],locals()[('RNA_Frame_eIF4B-2')]]).reset_index(drop = 'True')\n",
    "RiboAndRNA_Frame.to_csv((path +'AllSample_RegionFrame.txt'),sep = '\\t')\n",
    "RiboAndRNA_Frame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. Genes have uORF: 28nt reads in Ribo-seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### align genes at first uAUG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Finding the position of first uAUG at the 5'UTR\n",
    "\n",
    "\n",
    "UTR3_annotation = pd.read_table('~/3UTR_seq.txt') ## Only consider genes with 5'UTR and 3'UTR annotations and 5'UTR length > 1\n",
    "UTR3_gene = UTR3_annotation['gene_id'].unique()\n",
    "\n",
    "UTR5_annotation = pd.read_table('~/5UTR_seq.txt')\n",
    "UTR5_annotation['length'] = UTR5_annotation['seq'].apply(len)\n",
    "UTR5_annotation = UTR5_annotation[UTR5_annotation['length'] != 1] \n",
    "UTR5_gene = UTR5_annotation['gene_id'].unique()\n",
    "\n",
    "gene = list(set(UTR5_gene).intersection(set(UTR3_gene)))\n",
    "UTR5_annotation = UTR5_annotation[UTR5_annotation['gene_id'].isin(gene)].reset_index(drop = True) \n",
    "\n",
    "\n",
    "First_uAUG = UTR5_annotation[UTR5_annotation['seq'].str.contains('ATG')]\n",
    "uAUG_Gene = First_uAUG['gene_id'].unique()\n",
    "print('n Gene have uAUG = ' + str(len(uAUG_Gene)))\n",
    "\n",
    "First_uAUG_1 = First_uAUG[First_uAUG['strand'] == '+']\n",
    "First_uAUG_1['position'] = First_uAUG_1['seq'].str.find('ATG')\n",
    "First_uAUG_1['uAUG_start'] = First_uAUG_1['start'] + First_uAUG_1['position']\n",
    "First_uAUG_1['uAUG_stop'] = First_uAUG_1['start'] + First_uAUG_1['position'] + 2\n",
    "First_uAUG_1 = First_uAUG_1[['uAUG_start','uAUG_stop','strand','gene_id']]\n",
    "First_uAUG_1.columns = ['uAUG_start','uAUG_stop','strand','gene']\n",
    "First_uAUG_0 = First_uAUG[First_uAUG['strand'] == '-']\n",
    "First_uAUG_0['position'] = First_uAUG_0['seq'].str.find('ATG')\n",
    "First_uAUG_0['uAUG_start'] = First_uAUG_0['stop'] - First_uAUG_0['position'] - 2\n",
    "First_uAUG_0['uAUG_stop'] = First_uAUG_0['stop'] - First_uAUG_0['position']\n",
    "First_uAUG_0 = First_uAUG_0[['uAUG_start','uAUG_stop','strand','gene_id']]\n",
    "First_uAUG_0.columns = ['uAUG_start','uAUG_stop','strand','gene']\n",
    "First_uAUG = pd.concat([First_uAUG_0,First_uAUG_1]).reset_index(drop = True)\n",
    "\n",
    "First_uAUG.to_csv('~/uAUG_Gene.txt', sep = '\\t')\n",
    "\n",
    "# Genes with uORFs annotated in the Ingolia article\n",
    "Ingolia_uORF_GeneList = pd.read_table('~/uORF_Gene_2009-ingolia.txt')\n",
    "Ingolia_uORF_Gene = Ingolia_uORF_GeneList['Gene'].unique()\n",
    "First_uAUG = First_uAUG[First_uAUG['gene'].isin(Ingolia_uORF_Gene)].reset_index(drop = True)\n",
    "\n",
    "First_uAUG.to_csv('~/uAUG_Gene_Ingolia_First.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = 'HO'\n",
    "#rep = '1'\n",
    "name = sample# + '-' + rep\n",
    "path1 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/1-BAM/BAM_gene/Ribo/' + sample + '/') ## The pathway of bam_AssignGene_transcript±20.csv\n",
    "path2 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/2-TripletPeriodicity/Periodicity_AlignAt_uAUG-2009-ingolia/First/' + sample + '/') ## The pathway to output results\n",
    "\n",
    "First_uAUG = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uAUG_Gene_Ingolia_First.txt', usecols = ['uAUG_start','uAUG_stop','gene'])\n",
    "\n",
    "ReadsAssignToGene = pd.read_csv((path1 + name + '_bam_AssignGene_transcript±20.csv'))\n",
    "ReadsAssignToGene = pd.merge(ReadsAssignToGene, First_uAUG, on = 'gene')\n",
    "ReadsAssignToGene = ReadsAssignToGene[['seqid','ReadsStartPos','ReadsEndPos','length','gene','uAUG_start','uAUG_stop','strand']]\n",
    "\n",
    "ReadsAssignToGene_1 = ReadsAssignToGene[ReadsAssignToGene['strand'] == '+'].reset_index(drop = True)\n",
    "ReadsAssignToGene_1['StartPos_to_uAUG_start'] = ReadsAssignToGene_1['ReadsStartPos'] - ReadsAssignToGene_1['uAUG_start']\n",
    "ReadsAssignToGene_1['StopPos_to_uAUG_start'] = ReadsAssignToGene_1['ReadsEndPos'] - ReadsAssignToGene_1['uAUG_start']\n",
    "ReadsAssignToGene_1 = ReadsAssignToGene_1[['length','gene','strand','StartPos_to_uAUG_start','StopPos_to_uAUG_start']]\n",
    "\n",
    "ReadsAssignToGene_0 = ReadsAssignToGene[ReadsAssignToGene['strand'] == '-'].reset_index(drop = True)\n",
    "ReadsAssignToGene_0['StartPos_to_uAUG_start'] = ReadsAssignToGene_0['uAUG_stop'] - ReadsAssignToGene_0['ReadsEndPos']\n",
    "ReadsAssignToGene_0['StopPos_to_uAUG_start'] = ReadsAssignToGene_0['uAUG_stop'] - ReadsAssignToGene_0['ReadsStartPos']\n",
    "ReadsAssignToGene_0 = ReadsAssignToGene_0[['length','gene','strand','StartPos_to_uAUG_start','StopPos_to_uAUG_start']]\n",
    "ReadsAssignToGene = pd.concat([ReadsAssignToGene_1,ReadsAssignToGene_0]).reset_index(drop = True)\n",
    "ReadsAssignToGene.to_csv((path2 + name + '_To_First_uAUG_Frame.txt'),sep = '\\t')\n",
    "\n",
    "LengthCount = ReadsAssignToGene['length'].value_counts().reset_index()\n",
    "LengthCount.columns = ['length','count']\n",
    "LengthCount.to_csv((path2 + name + '_Length_count.txt'),sep = '\\t')\n",
    "\n",
    "Start_codon_around_count = ReadsAssignToGene.groupby(['length'])['StartPos_to_uAUG_start'].value_counts().to_frame()\n",
    "Start_codon_around_count.columns = ['counts']\n",
    "Start_codon_around_count = Start_codon_around_count.reset_index()\n",
    "Start_codon_around_count['frame'] = Start_codon_around_count['StartPos_to_uAUG_start'] % 3\n",
    "Start_codon_around_count.to_csv((path2 + name + '_StartCodonFrame_Length.txt'),sep = '\\t')\n",
    "\n",
    "Gene_Start_codon_around_count = ReadsAssignToGene.groupby(['length','gene'])['StartPos_to_uAUG_start'].value_counts().to_frame() ## single gene level\n",
    "Gene_Start_codon_around_count.columns = ['counts']\n",
    "Gene_Start_codon_around_count = Gene_Start_codon_around_count.reset_index()\n",
    "Gene_Start_codon_around_count['frame'] = Gene_Start_codon_around_count['StartPos_to_uAUG_start'] % 3\n",
    "Gene_Start_codon_around_count.to_csv((path2 + name + '_Gene_StartCodonFrame_Length.txt'),sep = '\\t')\n",
    "\n",
    "Start_codon_total_count = ReadsAssignToGene['StartPos_to_uAUG_start'].value_counts().reset_index()\n",
    "Start_codon_total_count.columns = ['StartPos_to_uAUG_start','counts']\n",
    "Start_codon_total_count.to_csv((path2 + name + '_StartPos_count.txt'),sep = '\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### align genes at second uAUG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Finding the position of second uAUG at the 5'UTR\n",
    "\n",
    "UTR3_annotation = pd.read_table('~/3UTR_seq.txt') ## Only consider genes with 5'UTR and 3'UTR annotations and 5'UTR length > 1\n",
    "UTR3_gene = UTR3_annotation['gene_id'].unique()\n",
    "\n",
    "UTR5_annotation = pd.read_table('~/5UTR_seq.txt')\n",
    "UTR5_annotation['length'] = UTR5_annotation['seq'].apply(len)\n",
    "UTR5_annotation = UTR5_annotation[UTR5_annotation['length'] != 1] \n",
    "UTR5_gene = UTR5_annotation['gene_id'].unique()\n",
    "\n",
    "gene = list(set(UTR5_gene).intersection(set(UTR3_gene)))\n",
    "UTR5_annotation = UTR5_annotation[UTR5_annotation['gene_id'].isin(gene)].reset_index(drop = True) \n",
    "\n",
    "\n",
    "Second_uAUG = UTR5_annotation[UTR5_annotation['seq'].str.count('ATG') >= 2]\n",
    "uAUG_Gene = Second_uAUG['gene_id'].unique()\n",
    "print('n Gene have uAUG = ' + str(len(uAUG_Gene)))\n",
    "\n",
    "\n",
    "def find_second_occurrence(text, substring):\n",
    "    first_occurrence = text.find(substring)\n",
    "    if first_occurrence != -1:\n",
    "        second_occurrence = text.find(substring, first_occurrence + 1)\n",
    "        return second_occurrence\n",
    "    else:\n",
    "        return -1\n",
    "\n",
    "Second_uAUG_1 = Second_uAUG[Second_uAUG['strand'] == '+'].reset_index(drop = True)\n",
    "Second_uAUG_1['position'] = ''\n",
    "for i in range(len(Second_uAUG_1['position'])):\n",
    "    Second_uAUG_1['position'][i] = find_second_occurrence(Second_uAUG_1['seq'][i],'ATG')\n",
    "Second_uAUG_1['Second_uAUG_start'] = Second_uAUG_1['start'] + Second_uAUG_1['position']\n",
    "Second_uAUG_1['Second_uAUG_stop'] = Second_uAUG_1['start'] + Second_uAUG_1['position'] + 2\n",
    "Second_uAUG_1 = Second_uAUG_1[['Second_uAUG_start','Second_uAUG_stop','strand','gene_id']]\n",
    "Second_uAUG_1.columns = ['Second_uAUG_start','Second_uAUG_stop','strand','gene']\n",
    "\n",
    "Second_uAUG_0 = Second_uAUG[Second_uAUG['strand'] == '-'].reset_index(drop = True)\n",
    "Second_uAUG_0['position'] = ''\n",
    "for i in range(len(Second_uAUG_0['position'])):\n",
    "    Second_uAUG_0['position'][i] = find_second_occurrence(Second_uAUG_0['seq'][i],'ATG')\n",
    "Second_uAUG_0['Second_uAUG_start'] = Second_uAUG_0['stop'] - Second_uAUG_0['position'] - 2\n",
    "Second_uAUG_0['Second_uAUG_stop'] = Second_uAUG_0['stop'] - Second_uAUG_0['position']\n",
    "Second_uAUG_0 = Second_uAUG_0[['Second_uAUG_start','Second_uAUG_stop','strand','gene_id']]\n",
    "Second_uAUG_0.columns = ['Second_uAUG_start','Second_uAUG_stop','strand','gene']\n",
    "Second_uAUG = pd.concat([Second_uAUG_0,Second_uAUG_1]).reset_index(drop = True)\n",
    "\n",
    "# Genes with uORFs annotated in the Ingolia article\n",
    "Ingolia_uORF_GeneList = pd.read_table('~/uORF_Gene_2009-ingolia.txt')\n",
    "Ingolia_uORF_Gene = Ingolia_uORF_GeneList['Gene'].unique()\n",
    "Second_uAUG = Second_uAUG[Second_uAUG['gene'].isin(Ingolia_uORF_Gene)].reset_index(drop = True)\n",
    "\n",
    "uAUG_First = pd.read_table('~/uAUG_Gene_Ingolia_First.txt', usecols = ['uAUG_start','uAUG_stop','gene'])\n",
    "Second_uAUG = pd.merge(uAUG_First, Second_uAUG, on = 'gene')\n",
    "Second_uAUG.to_csv('~/uAUG_Gene_Ingolia_Second.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = 'HO'\n",
    "#rep = '2'\n",
    "name = sample# + '-' + rep\n",
    "path1 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/1-BAM/BAM_gene/Ribo/' + sample + '/') ## The pathway of bam_AssignGene_transcript±20.csv\n",
    "path2 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/2-TripletPeriodicity/Periodicity_AlignAt_uAUG-2009-ingolia/Second/' + sample + '/') ## The pathway to output results\n",
    "\n",
    "\n",
    "Second_uAUG = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uAUG_Gene_Ingolia_Second.txt', usecols = ['Second_uAUG_start','Second_uAUG_stop','gene'])\n",
    "\n",
    "ReadsAssignToGene = pd.read_csv((path1 + name + '_bam_AssignGene_transcript±20.csv'))\n",
    "ReadsAssignToGene = pd.merge(ReadsAssignToGene,Second_uAUG,on = 'gene')\n",
    "ReadsAssignToGene = ReadsAssignToGene[['seqid','ReadsStartPos','ReadsEndPos','length','gene','Second_uAUG_start','Second_uAUG_stop','strand']]\n",
    "\n",
    "ReadsAssignToGene_1 = ReadsAssignToGene[ReadsAssignToGene['strand'] == '+'].reset_index(drop = True)\n",
    "ReadsAssignToGene_1['StartPos_to_Second_uAUG_start'] = ReadsAssignToGene_1['ReadsStartPos'] - ReadsAssignToGene_1['Second_uAUG_start']\n",
    "ReadsAssignToGene_1['StopPos_to_Second_uAUG_start'] = ReadsAssignToGene_1['ReadsEndPos'] - ReadsAssignToGene_1['Second_uAUG_start']\n",
    "ReadsAssignToGene_1 = ReadsAssignToGene_1[['length','gene','strand','StartPos_to_Second_uAUG_start','StopPos_to_Second_uAUG_start']]\n",
    "\n",
    "ReadsAssignToGene_0 = ReadsAssignToGene[ReadsAssignToGene['strand'] == '-'].reset_index(drop = True)\n",
    "ReadsAssignToGene_0['StartPos_to_Second_uAUG_start'] = ReadsAssignToGene_0['Second_uAUG_stop'] - ReadsAssignToGene_0['ReadsEndPos']\n",
    "ReadsAssignToGene_0['StopPos_to_Second_uAUG_start'] = ReadsAssignToGene_0['Second_uAUG_stop'] - ReadsAssignToGene_0['ReadsStartPos']\n",
    "ReadsAssignToGene_0 = ReadsAssignToGene_0[['length','gene','strand','StartPos_to_Second_uAUG_start','StopPos_to_Second_uAUG_start']]\n",
    "ReadsAssignToGene = pd.concat([ReadsAssignToGene_1,ReadsAssignToGene_0]).reset_index(drop = True)\n",
    "ReadsAssignToGene.to_csv((path2 + name + '_To_Second_uAUG_Frame.txt'),sep = '\\t')\n",
    "\n",
    "LengthCount = ReadsAssignToGene['length'].value_counts().reset_index()\n",
    "LengthCount.columns = ['length','count']\n",
    "LengthCount.to_csv((path2 + name + '_Length_count.txt'),sep = '\\t')\n",
    "\n",
    "Start_codon_around_count = ReadsAssignToGene.groupby(['length'])['StartPos_to_Second_uAUG_start'].value_counts().to_frame()\n",
    "Start_codon_around_count.columns = ['counts']\n",
    "Start_codon_around_count = Start_codon_around_count.reset_index()\n",
    "Start_codon_around_count['frame'] = Start_codon_around_count['StartPos_to_Second_uAUG_start'] % 3\n",
    "Start_codon_around_count.to_csv((path2 + name + '_StartCodonFrame_Length.txt'),sep = '\\t')\n",
    "\n",
    "Gene_Start_codon_around_count = ReadsAssignToGene.groupby(['length','gene'])['StartPos_to_Second_uAUG_start'].value_counts().to_frame()  ## single gene level\n",
    "Gene_Start_codon_around_count.columns = ['counts']\n",
    "Gene_Start_codon_around_count = Gene_Start_codon_around_count.reset_index()\n",
    "Gene_Start_codon_around_count['frame'] = Gene_Start_codon_around_count['StartPos_to_Second_uAUG_start'] % 3\n",
    "Gene_Start_codon_around_count.to_csv((path2 + name + '_Gene_StartCodonFrame_Length.txt'),sep = '\\t')\n",
    "\n",
    "Start_codon_total_count = ReadsAssignToGene['StartPos_to_Second_uAUG_start'].value_counts().reset_index()\n",
    "Start_codon_total_count.columns = ['StartPos_to_Second_uAUG_start','counts']\n",
    "Start_codon_total_count.to_csv((path2 + name + '_StartPos_count.txt'),sep = '\\t')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. Calculation of RPF abundance and RNA abundance"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. RPF abundance and RNA abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 217,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "I\n",
      "II\n",
      "III\n",
      "IV\n",
      "V\n",
      "VI\n",
      "VII\n",
      "VIII\n",
      "IX\n",
      "X\n",
      "XI\n",
      "XII\n",
      "XIII\n",
      "XIV\n",
      "XV\n",
      "XVI\n",
      "Mito\n"
     ]
    }
   ],
   "source": [
    "### When assigning reads to genes, only the region with CDS±50 is selected here\n",
    "### Do the following for both Ribo-seq and RNA-seq data\n",
    "\n",
    "sample = 'EAP1'\n",
    "rep = '1'\n",
    "name = sample + '-' + rep\n",
    "SeqType = 'RNA'\n",
    "path2 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/1-BAM/BAM_gene/' + SeqType + '/' + sample + '/') ## The pathway of bam.csv file\n",
    "\n",
    "BAM_csv = pd.read_csv((path2 + name + '_bam.csv'))\n",
    "BAM_csv.columns = ['seqid','start_pos','length','seq']\n",
    "\n",
    "BAM_csv['length'] = BAM_csv['seq'].apply(len)\n",
    "BAM_csv['stop_pos'] = BAM_csv['start_pos'] + BAM_csv['length'] - 1\n",
    "\n",
    "\n",
    "seqid = BAM_csv['seqid'].unique()\n",
    "gene_position = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/Saccharomyces_cerevisiae_GeneInformation.txt') ## Positional annotation of gene CDS\n",
    "gene_position['start'] = gene_position['start'] - 50\n",
    "gene_position['stop'] = gene_position['stop'] + 50\n",
    "\n",
    "seqid_annotation = gene_position['chr'].unique()\n",
    "seqid = [i for i in seqid if i in seqid_annotation]\n",
    "\n",
    "ReadsAssignToGene = open((path2 + name + '_bam_AssignGene_CDS±50.csv'),'w')\n",
    "ReadsAssignToGene_writer = csv.writer(ReadsAssignToGene)\n",
    "ReadsAssignToGene_writer.writerow(['seqid','ReadsStartPos','ReadsEndPos','length','gene','strand','start','stop'])\n",
    "for i in seqid:\n",
    "    print(i)\n",
    "    gene_position_subset = gene_position[gene_position['chr'] == i].reset_index()\n",
    "    BAM_csv_subset = BAM_csv[BAM_csv['seqid'] == i].reset_index()\n",
    "    for n in range(len(gene_position_subset['chr'])):\n",
    "        start = gene_position_subset['start'][n]\n",
    "        stop = gene_position_subset['stop'][n]\n",
    "        gene = gene_position_subset['gene_id'][n]\n",
    "        strand = gene_position_subset['strand'][n]\n",
    "        BAM_csv_subset_pos = BAM_csv_subset[(BAM_csv_subset['start_pos'] >= start) & (BAM_csv_subset['stop_pos'] <= stop)].reset_index()\n",
    "        for m in range(len(BAM_csv_subset_pos['seqid'])):\n",
    "            length = BAM_csv_subset_pos['length'][m]\n",
    "            start_pos = BAM_csv_subset_pos['start_pos'][m]\n",
    "            stop_pos = BAM_csv_subset_pos['stop_pos'][m]\n",
    "            ReadsAssignToGene_writer.writerow([i,start_pos,stop_pos,length,gene,strand,start,stop])\n",
    "ReadsAssignToGene.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = 'EAP1'\n",
    "rep = '2'\n",
    "name = sample + '-' + rep\n",
    "SeqType = 'Ribo'\n",
    "path1 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/1-BAM/BAM_gene/' + SeqType + '/' + sample + '/') ## The pathway of bam_AssignGene_CDS±50.csv file\n",
    "path2 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/' + SeqType + '_Abundance/' + sample + '/') ## The pathway to output results\n",
    "\n",
    "### Calculate all gene abundance\n",
    "\n",
    "GeneReads = pd.read_csv((path1 + name + '_bam_AssignGene_CDS±50.csv'))\n",
    "GeneReads.value_counts('gene').to_csv((path2 + name + '_ReadCount.txt'), sep = '\\t')\n",
    "\n",
    "GeneReads = pd.read_table((path2 + name + '_ReadCount.txt'))\n",
    "\n",
    "gene_position = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/Saccharomyces_cerevisiae_GeneInformation.txt') ## Positional annotation of gene CDS\n",
    "gene_position = gene_position[['gene_id','start','stop']]\n",
    "gene_position['length'] = gene_position['stop'] - gene_position['start'] + 1\n",
    "\n",
    "GeneReads = pd.merge(GeneReads, gene_position, left_on = 'gene', right_on = 'gene_id')\n",
    "GeneReads = GeneReads[['gene','0','length']]\n",
    "GeneReads.columns = ['gene','count','length']\n",
    "GeneReads_sum = GeneReads['count'].sum()\n",
    "GeneReads[(SeqType + '_abundance')] = (GeneReads['count'] / GeneReads['length']) / GeneReads_sum\n",
    "GeneReads['sample'] = name\n",
    "GeneReads.to_csv((path2 + name + '-' + SeqType + '-Abundance.txt'), sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "YER165W\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>RNA_abundance</th>\n",
       "      <th>rep</th>\n",
       "      <th>sample</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YER165W</td>\n",
       "      <td>0.858013</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YER165W</td>\n",
       "      <td>0.700748</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YER165W</td>\n",
       "      <td>0.131107</td>\n",
       "      <td>rep1</td>\n",
       "      <td>PAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YER165W</td>\n",
       "      <td>0.114185</td>\n",
       "      <td>rep2</td>\n",
       "      <td>PAB1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      gene  RNA_abundance   rep sample\n",
       "0  YER165W       0.858013  rep1     HO\n",
       "1  YER165W       0.700748  rep2     HO\n",
       "2  YER165W       0.131107  rep1   PAB1\n",
       "3  YER165W       0.114185  rep2   PAB1"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Knock down and Knock out efficiency\n",
    "\n",
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/RNA_Abundance/'\n",
    "Gene_dict = {'ACT1':'YFL039C','EAP1':'YKL204W','CAF20':'YOR276W','eIF1':'YNL244C','eIF1A':'YMR260C',\\\n",
    "             'eIF4A':'YKR059W','eIF4B':'YPR163C','eIF4E':'YOL139C','eIF4G1':'YGR162W','PAB1':'YER165W'}\n",
    "Gene_List = ['YFL039C','YKL204W','YOR276W','YNL244C','YMR260C','YKR059W','YPR163C','YOL139C','YGR162W','YER165W']\n",
    "\n",
    "HO_1 = pd.read_table((path + 'HO/HO-1-RNA-Abundance.txt'),usecols = ['gene','RNA_abundance'])\n",
    "HO_1['rep'] = 'rep1'\n",
    "HO_1['RNA_abundance'] = HO_1['RNA_abundance'] / HO_1[HO_1['gene'] == 'YFL039C'].reset_index(drop = True)['RNA_abundance'][0]\n",
    "HO_2 = pd.read_table((path + 'HO/HO-2-RNA-Abundance.txt'),usecols = ['gene','RNA_abundance'])\n",
    "HO_2['rep'] = 'rep2'\n",
    "HO_2['RNA_abundance'] = HO_2['RNA_abundance'] / HO_2[HO_2['gene'] == 'YFL039C'].reset_index(drop = True)['RNA_abundance'][0]\n",
    "HO = pd.concat([HO_1,HO_2]).reset_index(drop = True)\n",
    "HO['sample'] = 'HO'\n",
    "HO = HO[HO['gene'].isin(Gene_List)].reset_index(drop = True)\n",
    "\n",
    "sample = 'PAB1'\n",
    "gene = Gene_dict[sample]\n",
    "print(gene)\n",
    "\n",
    "Mutant_1 = pd.read_table((path + sample + '/'  + sample + '-1-RNA-Abundance.txt'), usecols = ['gene','RNA_abundance'])\n",
    "Mutant_1['rep'] = 'rep1'\n",
    "Mutant_1['RNA_abundance'] = Mutant_1['RNA_abundance'] / Mutant_1[Mutant_1['gene'] == 'YFL039C'].reset_index(drop = True)['RNA_abundance'][0]\n",
    "\n",
    "if gene not in Mutant_1['gene'].values: # 判断列中是否存在目标元素\n",
    "    new_row = {'gene':gene,'RNA_abundance':0,'rep':'rep1'} # 如果目标元素不存在，则添加一行填充数据\n",
    "    Mutant_1 = Mutant_1.append(new_row, ignore_index=True)\n",
    "\n",
    "Mutant_2 = pd.read_table((path + sample + '/'  + sample + '-2-RNA-Abundance.txt'), usecols = ['gene','RNA_abundance'])\n",
    "Mutant_2['rep'] = 'rep2'\n",
    "Mutant_2['RNA_abundance'] = Mutant_2['RNA_abundance'] / Mutant_2[Mutant_2['gene'] == 'YFL039C'].reset_index(drop = True)['RNA_abundance'][0]\n",
    "\n",
    "if gene not in Mutant_2['gene'].values:\n",
    "    new_row = {'gene':gene,'RNA_abundance':0,'rep':'rep2'}\n",
    "    Mutant_2 = Mutant_2.append(new_row, ignore_index=True)\n",
    "\n",
    "Mutant = pd.concat([Mutant_1,Mutant_2]).reset_index(drop = True)\n",
    "Mutant['sample'] = sample\n",
    "Mutant = Mutant[Mutant['gene'].isin(Gene_List)].reset_index(drop = True)\n",
    "\n",
    "HO_Mutant = pd.concat([HO,Mutant]).reset_index(drop = True)\n",
    "HO_Mutant.to_csv((path + sample + '/' + sample + '_RNA_abundance_ACT1normalized.txt'), sep = '\\t')\n",
    "HO_Mutant[HO_Mutant['gene'] == gene].reset_index(drop = True).to_csv((path + sample + '/' + sample + '_RNA_abundance_ACT1normalized_subset.txt'), sep = '\\t')\n",
    "HO_Mutant[HO_Mutant['gene'] == gene].reset_index(drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>RNA_abundance</th>\n",
       "      <th>rep</th>\n",
       "      <th>sample</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YOR276W</td>\n",
       "      <td>0.085511</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "      <td>CAF20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YOR276W</td>\n",
       "      <td>0.094737</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "      <td>CAF20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YOR276W</td>\n",
       "      <td>0.001083</td>\n",
       "      <td>rep1</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>CAF20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YOR276W</td>\n",
       "      <td>0.000298</td>\n",
       "      <td>rep2</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>CAF20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YKL204W</td>\n",
       "      <td>0.029471</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "      <td>EAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>YKL204W</td>\n",
       "      <td>0.029167</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "      <td>EAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>YKL204W</td>\n",
       "      <td>0.001967</td>\n",
       "      <td>rep1</td>\n",
       "      <td>EAP1</td>\n",
       "      <td>EAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>YKL204W</td>\n",
       "      <td>0.001484</td>\n",
       "      <td>rep2</td>\n",
       "      <td>EAP1</td>\n",
       "      <td>EAP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>YOL139C</td>\n",
       "      <td>0.270439</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF4E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>YOL139C</td>\n",
       "      <td>0.287945</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF4E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>YOL139C</td>\n",
       "      <td>0.033997</td>\n",
       "      <td>rep1</td>\n",
       "      <td>eIF4E</td>\n",
       "      <td>eIF4E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>YOL139C</td>\n",
       "      <td>0.033419</td>\n",
       "      <td>rep2</td>\n",
       "      <td>eIF4E</td>\n",
       "      <td>eIF4E</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>YGR162W</td>\n",
       "      <td>0.089606</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF4G1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>YGR162W</td>\n",
       "      <td>0.089966</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF4G1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>YGR162W</td>\n",
       "      <td>0.001264</td>\n",
       "      <td>rep1</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>eIF4G1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>YGR162W</td>\n",
       "      <td>0.001328</td>\n",
       "      <td>rep2</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>eIF4G1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>YER165W</td>\n",
       "      <td>0.858013</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "      <td>PAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>YER165W</td>\n",
       "      <td>0.700748</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "      <td>PAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>YER165W</td>\n",
       "      <td>0.131107</td>\n",
       "      <td>rep1</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>PAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>YER165W</td>\n",
       "      <td>0.114185</td>\n",
       "      <td>rep2</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>PAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>YNL244C</td>\n",
       "      <td>1.282763</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>YNL244C</td>\n",
       "      <td>1.303210</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>YNL244C</td>\n",
       "      <td>0.194668</td>\n",
       "      <td>rep1</td>\n",
       "      <td>eIF1</td>\n",
       "      <td>eIF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>YNL244C</td>\n",
       "      <td>0.435358</td>\n",
       "      <td>rep2</td>\n",
       "      <td>eIF1</td>\n",
       "      <td>eIF1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>YMR260C</td>\n",
       "      <td>0.344222</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>YMR260C</td>\n",
       "      <td>0.374655</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>YMR260C</td>\n",
       "      <td>0.034718</td>\n",
       "      <td>rep1</td>\n",
       "      <td>eIF1A</td>\n",
       "      <td>eIF1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>YMR260C</td>\n",
       "      <td>0.037287</td>\n",
       "      <td>rep2</td>\n",
       "      <td>eIF1A</td>\n",
       "      <td>eIF1A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>YKR059W</td>\n",
       "      <td>0.082713</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>YKR059W</td>\n",
       "      <td>0.090916</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>YKR059W</td>\n",
       "      <td>0.002942</td>\n",
       "      <td>rep1</td>\n",
       "      <td>eIF4A</td>\n",
       "      <td>eIF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>YKR059W</td>\n",
       "      <td>0.001064</td>\n",
       "      <td>rep2</td>\n",
       "      <td>eIF4A</td>\n",
       "      <td>eIF4A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>YPR163C</td>\n",
       "      <td>0.288257</td>\n",
       "      <td>rep1</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>YPR163C</td>\n",
       "      <td>0.264323</td>\n",
       "      <td>rep2</td>\n",
       "      <td>HO</td>\n",
       "      <td>eIF4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>YPR163C</td>\n",
       "      <td>0.000943</td>\n",
       "      <td>rep1</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>eIF4B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>YPR163C</td>\n",
       "      <td>0.000154</td>\n",
       "      <td>rep2</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>eIF4B</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene  RNA_abundance   rep  sample   group\n",
       "0   YOR276W       0.085511  rep1      HO   CAF20\n",
       "1   YOR276W       0.094737  rep2      HO   CAF20\n",
       "2   YOR276W       0.001083  rep1   CAF20   CAF20\n",
       "3   YOR276W       0.000298  rep2   CAF20   CAF20\n",
       "4   YKL204W       0.029471  rep1      HO    EAP1\n",
       "5   YKL204W       0.029167  rep2      HO    EAP1\n",
       "6   YKL204W       0.001967  rep1    EAP1    EAP1\n",
       "7   YKL204W       0.001484  rep2    EAP1    EAP1\n",
       "8   YOL139C       0.270439  rep1      HO   eIF4E\n",
       "9   YOL139C       0.287945  rep2      HO   eIF4E\n",
       "10  YOL139C       0.033997  rep1   eIF4E   eIF4E\n",
       "11  YOL139C       0.033419  rep2   eIF4E   eIF4E\n",
       "12  YGR162W       0.089606  rep1      HO  eIF4G1\n",
       "13  YGR162W       0.089966  rep2      HO  eIF4G1\n",
       "14  YGR162W       0.001264  rep1  eIF4G1  eIF4G1\n",
       "15  YGR162W       0.001328  rep2  eIF4G1  eIF4G1\n",
       "16  YER165W       0.858013  rep1      HO    PAB1\n",
       "17  YER165W       0.700748  rep2      HO    PAB1\n",
       "18  YER165W       0.131107  rep1    PAB1    PAB1\n",
       "19  YER165W       0.114185  rep2    PAB1    PAB1\n",
       "20  YNL244C       1.282763  rep1      HO    eIF1\n",
       "21  YNL244C       1.303210  rep2      HO    eIF1\n",
       "22  YNL244C       0.194668  rep1    eIF1    eIF1\n",
       "23  YNL244C       0.435358  rep2    eIF1    eIF1\n",
       "24  YMR260C       0.344222  rep1      HO   eIF1A\n",
       "25  YMR260C       0.374655  rep2      HO   eIF1A\n",
       "26  YMR260C       0.034718  rep1   eIF1A   eIF1A\n",
       "27  YMR260C       0.037287  rep2   eIF1A   eIF1A\n",
       "28  YKR059W       0.082713  rep1      HO   eIF4A\n",
       "29  YKR059W       0.090916  rep2      HO   eIF4A\n",
       "30  YKR059W       0.002942  rep1   eIF4A   eIF4A\n",
       "31  YKR059W       0.001064  rep2   eIF4A   eIF4A\n",
       "32  YPR163C       0.288257  rep1      HO   eIF4B\n",
       "33  YPR163C       0.264323  rep2      HO   eIF4B\n",
       "34  YPR163C       0.000943  rep1   eIF4B   eIF4B\n",
       "35  YPR163C       0.000154  rep2   eIF4B   eIF4B"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/RNA_Abundance/'\n",
    "\n",
    "MutantList = ['CAF20','EAP1','eIF4E','eIF4G1','PAB1','eIF1','eIF1A','eIF4A','eIF4B']             \n",
    "\n",
    "for i in MutantList:\n",
    "    locals()[('mRNA_' + i)] = pd.read_table((path + i + '/' + i + '_RNA_abundance_ACT1normalized_subset.txt'))\n",
    "    locals()[('mRNA_' + i)]['group'] = i\n",
    "\n",
    "AllSample = pd.concat([locals()[('mRNA_CAF20')],locals()[('mRNA_EAP1')],locals()[('mRNA_eIF4E')],locals()[('mRNA_eIF4G1')],locals()[('mRNA_PAB1')],\n",
    "                       locals()[('mRNA_eIF1')],locals()[('mRNA_eIF1A')],locals()[('mRNA_eIF4A')],locals()[('mRNA_eIF4B')]]).reset_index().drop(['index','Unnamed: 0'], axis = 1)\n",
    "AllSample.to_csv((path + 'AllSample_mRNA_ACT1normalized_subset.txt'), sep = '\\t')\n",
    "AllSample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>HO</th>\n",
       "      <th>CAF20</th>\n",
       "      <th>EAP1</th>\n",
       "      <th>eIF4E</th>\n",
       "      <th>eIF4G1</th>\n",
       "      <th>PAB1</th>\n",
       "      <th>eIF1</th>\n",
       "      <th>eIF1A</th>\n",
       "      <th>eIF4A</th>\n",
       "      <th>eIF4B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YGR192C</td>\n",
       "      <td>1.044853</td>\n",
       "      <td>0.830954</td>\n",
       "      <td>0.913147</td>\n",
       "      <td>0.771685</td>\n",
       "      <td>0.541474</td>\n",
       "      <td>0.897885</td>\n",
       "      <td>0.905258</td>\n",
       "      <td>0.793557</td>\n",
       "      <td>0.985544</td>\n",
       "      <td>0.880244</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YLR044C</td>\n",
       "      <td>1.165907</td>\n",
       "      <td>1.565413</td>\n",
       "      <td>1.930788</td>\n",
       "      <td>1.132771</td>\n",
       "      <td>1.025065</td>\n",
       "      <td>1.524017</td>\n",
       "      <td>1.376604</td>\n",
       "      <td>1.113527</td>\n",
       "      <td>1.498104</td>\n",
       "      <td>1.262894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YHR174W</td>\n",
       "      <td>1.380856</td>\n",
       "      <td>1.630936</td>\n",
       "      <td>1.799037</td>\n",
       "      <td>1.606966</td>\n",
       "      <td>0.929578</td>\n",
       "      <td>1.319555</td>\n",
       "      <td>2.195932</td>\n",
       "      <td>1.807041</td>\n",
       "      <td>1.609496</td>\n",
       "      <td>1.859192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YKL060C</td>\n",
       "      <td>1.216527</td>\n",
       "      <td>1.812802</td>\n",
       "      <td>1.668074</td>\n",
       "      <td>2.086814</td>\n",
       "      <td>1.656685</td>\n",
       "      <td>1.267185</td>\n",
       "      <td>1.416241</td>\n",
       "      <td>1.675920</td>\n",
       "      <td>1.468450</td>\n",
       "      <td>2.038270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YOL086C</td>\n",
       "      <td>1.406400</td>\n",
       "      <td>1.581750</td>\n",
       "      <td>1.848760</td>\n",
       "      <td>1.702193</td>\n",
       "      <td>1.071226</td>\n",
       "      <td>1.294545</td>\n",
       "      <td>1.574576</td>\n",
       "      <td>1.387739</td>\n",
       "      <td>1.361455</td>\n",
       "      <td>1.236187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6863</th>\n",
       "      <td>tD(GUC)G2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>9.363076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6864</th>\n",
       "      <td>tG(GCC)C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>196.624592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6865</th>\n",
       "      <td>Q0032</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.044614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6866</th>\n",
       "      <td>tA(AGC)K1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>56.178455</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6867</th>\n",
       "      <td>tR(UCU)D</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28.089227</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6868 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene        HO     CAF20      EAP1     eIF4E    eIF4G1      PAB1  \\\n",
       "0       YGR192C  1.044853  0.830954  0.913147  0.771685  0.541474  0.897885   \n",
       "1       YLR044C  1.165907  1.565413  1.930788  1.132771  1.025065  1.524017   \n",
       "2       YHR174W  1.380856  1.630936  1.799037  1.606966  0.929578  1.319555   \n",
       "3       YKL060C  1.216527  1.812802  1.668074  2.086814  1.656685  1.267185   \n",
       "4       YOL086C  1.406400  1.581750  1.848760  1.702193  1.071226  1.294545   \n",
       "...         ...       ...       ...       ...       ...       ...       ...   \n",
       "6863  tD(GUC)G2       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "6864   tG(GCC)C       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "6865      Q0032       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "6866  tA(AGC)K1       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "6867   tR(UCU)D       NaN       NaN       NaN       NaN       NaN       NaN   \n",
       "\n",
       "          eIF1     eIF1A     eIF4A       eIF4B  \n",
       "0     0.905258  0.793557  0.985544    0.880244  \n",
       "1     1.376604  1.113527  1.498104    1.262894  \n",
       "2     2.195932  1.807041  1.609496    1.859192  \n",
       "3     1.416241  1.675920  1.468450    2.038270  \n",
       "4     1.574576  1.387739  1.361455    1.236187  \n",
       "...        ...       ...       ...         ...  \n",
       "6863       NaN       NaN       NaN    9.363076  \n",
       "6864       NaN       NaN       NaN  196.624592  \n",
       "6865       NaN       NaN       NaN   14.044614  \n",
       "6866       NaN       NaN       NaN   56.178455  \n",
       "6867       NaN       NaN       NaN   28.089227  \n",
       "\n",
       "[6868 rows x 11 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/'\n",
    "\n",
    "#MutantList = ['HO-1','HO-2','CAF20-1','CAF20-2','EAP1-1','EAP1-2','eIF4E-1','eIF4E-2',\\\n",
    "#    'eIF4G1-1','eIF4G1-2','PAB1-1','PAB1-2','eIF1-1','eIF1-2','eIF1A-1','eIF1A-2','eIF4A-1','eIF4A-2','eIF4B-1','eIF4B-2']             \n",
    "\n",
    "MutantList = ['HO','CAF20','EAP1','eIF4E','eIF4G1','PAB1','eIF1','eIF1A','eIF4A','eIF4B']   \n",
    "\n",
    "for i in MutantList:\n",
    "    #print(i)\n",
    "    #locals()[('mut_f_' + i)] = pd.read_table((path + i[:-2] + '/' + i + '-Ribo-Abundance.txt'), usecols = ['gene','RNA_Abundance'])\n",
    "    locals()[('mut_f_' + i)] = pd.read_table((path + 'Ribosome_Density/' + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "    locals()[('mut_f_' + i)].columns = ['gene', i]\n",
    "\n",
    "MutantList.remove('HO')\n",
    "MutantList.remove('CAF20')\n",
    "\n",
    "AllSample = pd.merge(locals()['mut_f_HO'],locals()['mut_f_CAF20'],on = 'gene', how = 'outer')\n",
    "\n",
    "for i in MutantList:\n",
    "    AllSample = pd.merge(AllSample,locals()[('mut_f_' + i)],on = 'gene', how = 'outer')\n",
    "AllSample.to_csv((path + 'TotalMutant-RibosomeDensity.txt'), sep = '\\t', na_rep = 'NA')\n",
    "AllSample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_25597/2148960494.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  locals()[i[:-2]+'_'+i[-1]+'_corr']['sample'] = i\n",
      "/tmp/ipykernel_25597/2148960494.py:13: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  locals()[i[:-2]+'_'+i[-1]+'_corr']['group'] = ''\n",
      "/tmp/ipykernel_25597/2148960494.py:17: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  locals()[i[:-2]+'_'+i[-1]+'_corr']['group'][j] = 'Rep'\n",
      "/tmp/ipykernel_25597/2148960494.py:19: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  locals()[i[:-2]+'_'+i[-1]+'_corr']['group'][j] = 'Strain'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>rho</th>\n",
       "      <th>sample</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HO.2</td>\n",
       "      <td>0.989021</td>\n",
       "      <td>HO.1</td>\n",
       "      <td>Rep</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAF20.1</td>\n",
       "      <td>0.966925</td>\n",
       "      <td>HO.1</td>\n",
       "      <td>Strain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CAF20.2</td>\n",
       "      <td>0.961333</td>\n",
       "      <td>HO.1</td>\n",
       "      <td>Strain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>EAP1.1</td>\n",
       "      <td>0.951856</td>\n",
       "      <td>HO.1</td>\n",
       "      <td>Strain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EAP1.2</td>\n",
       "      <td>0.924846</td>\n",
       "      <td>HO.1</td>\n",
       "      <td>Strain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>375</th>\n",
       "      <td>eIF1A.1</td>\n",
       "      <td>0.962687</td>\n",
       "      <td>eIF4B.2</td>\n",
       "      <td>Strain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>376</th>\n",
       "      <td>eIF1A.2</td>\n",
       "      <td>0.970510</td>\n",
       "      <td>eIF4B.2</td>\n",
       "      <td>Strain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377</th>\n",
       "      <td>eIF4A.1</td>\n",
       "      <td>0.967948</td>\n",
       "      <td>eIF4B.2</td>\n",
       "      <td>Strain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>378</th>\n",
       "      <td>eIF4A.2</td>\n",
       "      <td>0.968872</td>\n",
       "      <td>eIF4B.2</td>\n",
       "      <td>Strain</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>379</th>\n",
       "      <td>eIF4B.1</td>\n",
       "      <td>0.991387</td>\n",
       "      <td>eIF4B.2</td>\n",
       "      <td>Rep</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>380 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       index       rho   sample   group\n",
       "0       HO.2  0.989021     HO.1     Rep\n",
       "1    CAF20.1  0.966925     HO.1  Strain\n",
       "2    CAF20.2  0.961333     HO.1  Strain\n",
       "3     EAP1.1  0.951856     HO.1  Strain\n",
       "4     EAP1.2  0.924846     HO.1  Strain\n",
       "..       ...       ...      ...     ...\n",
       "375  eIF1A.1  0.962687  eIF4B.2  Strain\n",
       "376  eIF1A.2  0.970510  eIF4B.2  Strain\n",
       "377  eIF4A.1  0.967948  eIF4B.2  Strain\n",
       "378  eIF4A.2  0.968872  eIF4B.2  Strain\n",
       "379  eIF4B.1  0.991387  eIF4B.2     Rep\n",
       "\n",
       "[380 rows x 4 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#### spearman correlation between replications and strains\n",
    "\n",
    "RepAndStrain_corr = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/RNA_Abundance/TotalMutant-RNA-Density-correlation.txt')\n",
    "RepAndStrain_corr = RepAndStrain_corr.reset_index()\n",
    "SampleList = ['HO.1','HO.2','CAF20.1','CAF20.2','EAP1.1','EAP1.2','eIF4E.1','eIF4E.2',\\\n",
    "              'eIF4G1.1','eIF4G1.2','PAB1.1','PAB1.2','eIF1.1','eIF1.2','eIF1A.1','eIF1A.2',\\\n",
    "                'eIF4A.1','eIF4A.2','eIF4B.1','eIF4B.2']\n",
    "\n",
    "for i in SampleList:\n",
    "    locals()[i[:-2]+'_'+i[-1]+'_corr'] = RepAndStrain_corr[['index',i]]\n",
    "    locals()[i[:-2]+'_'+i[-1]+'_corr'].columns = ['index','rho']\n",
    "    locals()[i[:-2]+'_'+i[-1]+'_corr']['sample'] = i\n",
    "    locals()[i[:-2]+'_'+i[-1]+'_corr']['group'] = ''\n",
    "\n",
    "    for j in range(len(locals()[i[:-2]+'_'+i[-1]+'_corr']['index'])):\n",
    "        if locals()[i[:-2]+'_'+i[-1]+'_corr']['index'][j][:-2] == i[:-2]:\n",
    "            locals()[i[:-2]+'_'+i[-1]+'_corr']['group'][j] = 'Rep'\n",
    "        else:\n",
    "            locals()[i[:-2]+'_'+i[-1]+'_corr']['group'][j] = 'Strain'\n",
    "\n",
    "RepAndStrain_corr = pd.concat([HO_1_corr,HO_2_corr,CAF20_1_corr,CAF20_2_corr,EAP1_1_corr,EAP1_2_corr,eIF4E_1_corr,eIF4E_2_corr,\\\n",
    "                               eIF4G1_1_corr,eIF4G1_2_corr,PAB1_1_corr,PAB1_2_corr,eIF1_1_corr,eIF1_2_corr,eIF1A_1_corr,eIF1A_2_corr,\\\n",
    "                                eIF4A_1_corr,eIF4A_2_corr,eIF4B_1_corr,eIF4B_2_corr])\n",
    "\n",
    "RepAndStrain_corr = RepAndStrain_corr[RepAndStrain_corr['rho'] != 1].reset_index(drop = True)\n",
    "RepAndStrain_corr.to_csv('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/RNA_Abundance/TotalMutant-RNA-Density-corr.txt', sep = '\\t')\n",
    "RepAndStrain_corr"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### b. Ribosome density and Ribosome density foldchange"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 311,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Calculate Ribosome density\n",
    "\n",
    "sample = 'CAF20'\n",
    "#rep = '1'\n",
    "name = sample# + '-' + rep\n",
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/' ## The file path where -RNA-Abundance.txt and -RPF-Abundance.txt are located\n",
    "\n",
    "\n",
    "mRNA_abundance = pd.read_table((path + 'RNA_Abundance/' + sample + '/' + name + '-RNA-Abundance.txt'), usecols = ['gene','RNA_abundance'])\n",
    "Ribo_abundance = pd.read_table((path + 'Ribo_Abundance/' + sample + '/' + name + '-Ribo-Abundance.txt'), usecols = ['gene','Ribo_abundance'])\n",
    "\n",
    "RibosomeDensity = pd.merge(mRNA_abundance,Ribo_abundance,on = 'gene')\n",
    "RibosomeDensity['Ribosome_Density'] = RibosomeDensity['Ribo_abundance'] / RibosomeDensity['RNA_abundance']\n",
    "RibosomeDensity = RibosomeDensity[['gene','RNA_abundance','Ribo_abundance','Ribosome_Density']]\n",
    "RibosomeDensity.columns = ['gene','RNA_abundance','RPF_abundance','Ribosome_Density']\n",
    "RibosomeDensity = RibosomeDensity[~RibosomeDensity.isin([np.nan, np.inf, -np.inf]).any(1)].reset_index().drop(['index'], axis = 1)\n",
    "RibosomeDensity['sample'] = name\n",
    "RibosomeDensity.to_csv((path + 'Ribosome_Density/' + sample + '/' + name + '-RibosomeDensity.txt'), sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 321,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Calculate Ribosome density foldchange\n",
    "\n",
    "sample = 'PAB1'\n",
    "#rep = '1'\n",
    "name = sample# + '-' + rep\n",
    "path = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/' + sample + '/') ## The file path where -RibosomeDensity.txt are located\n",
    "\n",
    "WT_RibosomeDensity = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/HO/HO-RibosomeDensity.txt',\\\n",
    "                                    usecols = ['gene','RNA_abundance','RPF_abundance','Ribosome_Density']) ## Pathways for wild-type Ribosome Density\n",
    "WT_RibosomeDensity.columns = ['gene','WT_RNA_abundance','WT_RPF_abundance','WT_Ribosome_Density']\n",
    "RibosomeDensity = pd.read_table((path + name + '-RibosomeDensity.txt'), usecols = ['gene','RNA_abundance','RPF_abundance','Ribosome_Density'])\n",
    "\n",
    "RibosomeDensity_FoldChange = pd.merge(RibosomeDensity,WT_RibosomeDensity,on = 'gene')\n",
    "RibosomeDensity_FoldChange['RNA_foldchange'] = RibosomeDensity_FoldChange['RNA_abundance'] / RibosomeDensity_FoldChange['WT_RNA_abundance']\n",
    "RibosomeDensity_FoldChange['RPF_foldchange'] = RibosomeDensity_FoldChange['RPF_abundance'] / RibosomeDensity_FoldChange['WT_RPF_abundance']\n",
    "RibosomeDensity_FoldChange['RibosomeDensity_foldchange'] = RibosomeDensity_FoldChange['Ribosome_Density'] / RibosomeDensity_FoldChange['WT_Ribosome_Density']\n",
    "RibosomeDensity_FoldChange = RibosomeDensity_FoldChange[~RibosomeDensity_FoldChange.isin([np.nan, np.inf, -np.inf]).any(1)].reset_index().drop(['index'], axis = 1)\n",
    "RibosomeDensity_FoldChange.to_csv((path + name + '-RNA_RPF_RibosomeDensity-FoldChange.txt'), sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>RNA_foldchange</th>\n",
       "      <th>RPF_foldchange</th>\n",
       "      <th>RibosomeDensity_foldchange</th>\n",
       "      <th>sample</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tK(CUU)M</td>\n",
       "      <td>0.001283</td>\n",
       "      <td>5.692158</td>\n",
       "      <td>4435.138390</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>top</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tN(GUU)K</td>\n",
       "      <td>0.014053</td>\n",
       "      <td>1.956033</td>\n",
       "      <td>139.184916</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>top</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tS(AGA)D1</td>\n",
       "      <td>0.009734</td>\n",
       "      <td>1.282268</td>\n",
       "      <td>131.730625</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>top</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tG(GCC)D2</td>\n",
       "      <td>0.079949</td>\n",
       "      <td>1.946053</td>\n",
       "      <td>24.341271</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>top</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tR(CCU)J</td>\n",
       "      <td>0.185595</td>\n",
       "      <td>2.816628</td>\n",
       "      <td>15.176189</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>top</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53948</th>\n",
       "      <td>YLR154W-A</td>\n",
       "      <td>27.528981</td>\n",
       "      <td>0.295254</td>\n",
       "      <td>0.010725</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>bottom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53949</th>\n",
       "      <td>YHL037C</td>\n",
       "      <td>1.872720</td>\n",
       "      <td>0.018453</td>\n",
       "      <td>0.009854</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>bottom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53950</th>\n",
       "      <td>tL(CAA)A</td>\n",
       "      <td>16.355086</td>\n",
       "      <td>0.127677</td>\n",
       "      <td>0.007807</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>bottom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53951</th>\n",
       "      <td>snR190</td>\n",
       "      <td>0.915552</td>\n",
       "      <td>0.005464</td>\n",
       "      <td>0.005968</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>bottom</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53952</th>\n",
       "      <td>YJL114W</td>\n",
       "      <td>191.953779</td>\n",
       "      <td>0.787344</td>\n",
       "      <td>0.004102</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>bottom</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>53953 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene  RNA_foldchange  RPF_foldchange  RibosomeDensity_foldchange  \\\n",
       "0       tK(CUU)M        0.001283        5.692158                 4435.138390   \n",
       "1       tN(GUU)K        0.014053        1.956033                  139.184916   \n",
       "2      tS(AGA)D1        0.009734        1.282268                  131.730625   \n",
       "3      tG(GCC)D2        0.079949        1.946053                   24.341271   \n",
       "4       tR(CCU)J        0.185595        2.816628                   15.176189   \n",
       "...          ...             ...             ...                         ...   \n",
       "53948  YLR154W-A       27.528981        0.295254                    0.010725   \n",
       "53949    YHL037C        1.872720        0.018453                    0.009854   \n",
       "53950   tL(CAA)A       16.355086        0.127677                    0.007807   \n",
       "53951     snR190        0.915552        0.005464                    0.005968   \n",
       "53952    YJL114W      191.953779        0.787344                    0.004102   \n",
       "\n",
       "      sample   group  \n",
       "0      CAF20     top  \n",
       "1      CAF20     top  \n",
       "2      CAF20     top  \n",
       "3      CAF20     top  \n",
       "4      CAF20     top  \n",
       "...      ...     ...  \n",
       "53948  eIF4B  bottom  \n",
       "53949  eIF4B  bottom  \n",
       "53950  eIF4B  bottom  \n",
       "53951  eIF4B  bottom  \n",
       "53952  eIF4B  bottom  \n",
       "\n",
       "[53953 rows x 6 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Ribosome Density foldchange bin\n",
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/RNA_Abundance/'\n",
    "\n",
    "MutantList = ['CAF20','EAP1','eIF4E','eIF4G1','PAB1','eIF1','eIF1A','eIF4A','eIF4B']             \n",
    "\n",
    "for i in MutantList:\n",
    "    locals()[('RDfold_' + i)] = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/' + i + '/' + i + '-RNA_RPF_RibosomeDensity-FoldChange.txt', usecols = ['gene','RNA_foldchange','RPF_foldchange','RibosomeDensity_foldchange'])\n",
    "    locals()[('RDfold_' + i)]['sample'] = i\n",
    "    locals()[('RDfold_' + i)] = locals()[('RDfold_' + i)].sort_values('RibosomeDensity_foldchange', ascending=False).reset_index(drop = True)\n",
    "    total_rows = len(locals()[('RDfold_' + i)])\n",
    "    top_30_percent_rows = int(total_rows * 0.3)\n",
    "    bottom_30_percent_rows = int(total_rows * 0.3)\n",
    "    locals()[('RDfold_' + i)]['group'] = ''\n",
    "    locals()[('RDfold_' + i)].loc[:top_30_percent_rows, 'group'] = 'top'\n",
    "    locals()[('RDfold_' + i)].loc[top_30_percent_rows+1:total_rows-bottom_30_percent_rows-2, 'group'] = 'median'\n",
    "    locals()[('RDfold_' + i)].loc[total_rows-bottom_30_percent_rows-1:, 'group'] = 'bottom'\n",
    "    locals()[('RDfold_' + i)].to_csv(('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/' + i + '/' + i + '-RNA_RPF_RibosomeDensity-FoldChange-bin.txt'),sep = '\\t')\n",
    "\n",
    "\n",
    "AllSample = pd.concat([locals()[('RDfold_CAF20')],locals()[('RDfold_EAP1')],locals()[('RDfold_eIF4E')],locals()[('RDfold_eIF4G1')],locals()[('RDfold_PAB1')],\n",
    "                       locals()[('RDfold_eIF1')],locals()[('RDfold_eIF1A')],locals()[('RDfold_eIF4A')],locals()[('RDfold_eIF4B')]]).reset_index().drop(['index'], axis = 1)\n",
    "AllSample.to_csv('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/TotalMutant-RNAandRPF-FoldChange-rbind-RDfold-bin.txt', sep = '\\t')\n",
    "AllSample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_172253/2560827747.py:11: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Foldchange_File['uORF'][i] = 'No'\n",
      "/tmp/ipykernel_172253/2560827747.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Foldchange_File['uORF'][i] = 'Yes'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>sample</th>\n",
       "      <th>RNA_foldchange</th>\n",
       "      <th>RPF_foldchange</th>\n",
       "      <th>uORF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YGR192C</td>\n",
       "      <td>HO</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YLR044C</td>\n",
       "      <td>HO</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YHR174W</td>\n",
       "      <td>HO</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YKL060C</td>\n",
       "      <td>HO</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YOL086C</td>\n",
       "      <td>HO</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60052</th>\n",
       "      <td>snR60</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>0.124848</td>\n",
       "      <td>0.590508</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60053</th>\n",
       "      <td>YCL018W</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>0.374544</td>\n",
       "      <td>1.181016</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60054</th>\n",
       "      <td>YMR323W</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>0.374544</td>\n",
       "      <td>0.393672</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60055</th>\n",
       "      <td>tC(GCA)B</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>0.041616</td>\n",
       "      <td>3.149375</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60056</th>\n",
       "      <td>tD(GUC)I2</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>0.374544</td>\n",
       "      <td>10.629142</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>60057 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene sample  RNA_foldchange  RPF_foldchange uORF\n",
       "0        YGR192C     HO        1.000000        1.000000   No\n",
       "1        YLR044C     HO        1.000000        1.000000   No\n",
       "2        YHR174W     HO        1.000000        1.000000   No\n",
       "3        YKL060C     HO        1.000000        1.000000   No\n",
       "4        YOL086C     HO        1.000000        1.000000   No\n",
       "...          ...    ...             ...             ...  ...\n",
       "60052      snR60  eIF4B        0.124848        0.590508   No\n",
       "60053    YCL018W  eIF4B        0.374544        1.181016   No\n",
       "60054    YMR323W  eIF4B        0.374544        0.393672   No\n",
       "60055   tC(GCA)B  eIF4B        0.041616        3.149375   No\n",
       "60056  tD(GUC)I2  eIF4B        0.374544       10.629142   No\n",
       "\n",
       "[60057 rows x 5 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### uORF\n",
    "Foldchange_File = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/TotalMutant-foldchange-rbind.txt',usecols = ['gene','RNA_foldchange','RPF_foldchange','sample'])\n",
    "Ingolia_uORF_Gene = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uORF_Gene_FristPos_2009-ingolia.txt')\n",
    "Ingolia_uORF_GeneList = Ingolia_uORF_Gene['Gene'].unique()\n",
    "Foldchange_File['uORF'] = ''\n",
    "\n",
    "for i in range(len(Foldchange_File['uORF'])):\n",
    "    if Foldchange_File['gene'][i] in Ingolia_uORF_GeneList:\n",
    "        Foldchange_File['uORF'][i] = 'Yes'\n",
    "    else:\n",
    "        Foldchange_File['uORF'][i] = 'No'\n",
    "Foldchange_File.to_csv('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/TotalMutant-RNAandRPF-FoldChange-rbind-uORF.txt', sep = '\\t')\n",
    "Foldchange_File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>gene</th>\n",
       "      <th>sample</th>\n",
       "      <th>RNA_foldchange</th>\n",
       "      <th>RPF_foldchange</th>\n",
       "      <th>uORF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>30153</th>\n",
       "      <td>30153</td>\n",
       "      <td>YGR192C</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>1.084789</td>\n",
       "      <td>0.932203</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30154</th>\n",
       "      <td>30154</td>\n",
       "      <td>YKL060C</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>1.089743</td>\n",
       "      <td>1.135122</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30155</th>\n",
       "      <td>30155</td>\n",
       "      <td>YLR044C</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>0.906476</td>\n",
       "      <td>1.184901</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30156</th>\n",
       "      <td>30156</td>\n",
       "      <td>YHR174W</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>0.764134</td>\n",
       "      <td>0.730212</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30157</th>\n",
       "      <td>30157</td>\n",
       "      <td>YOL086C</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>0.953922</td>\n",
       "      <td>0.878054</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36181</th>\n",
       "      <td>36181</td>\n",
       "      <td>tD(GUC)N</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>1.192252</td>\n",
       "      <td>0.488894</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36182</th>\n",
       "      <td>36182</td>\n",
       "      <td>YAR010C</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>0.397417</td>\n",
       "      <td>0.570376</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36183</th>\n",
       "      <td>36183</td>\n",
       "      <td>YBR182C-A</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>0.397417</td>\n",
       "      <td>0.570376</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36184</th>\n",
       "      <td>36184</td>\n",
       "      <td>YJR005C-A</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>0.119225</td>\n",
       "      <td>0.570376</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36185</th>\n",
       "      <td>36185</td>\n",
       "      <td>tF(GAA)P2</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>0.596126</td>\n",
       "      <td>0.570376</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5458 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0       gene sample  RNA_foldchange  RPF_foldchange uORF\n",
       "30153       30153    YGR192C   PAB1        1.084789        0.932203   No\n",
       "30154       30154    YKL060C   PAB1        1.089743        1.135122   No\n",
       "30155       30155    YLR044C   PAB1        0.906476        1.184901   No\n",
       "30156       30156    YHR174W   PAB1        0.764134        0.730212   No\n",
       "30157       30157    YOL086C   PAB1        0.953922        0.878054   No\n",
       "...           ...        ...    ...             ...             ...  ...\n",
       "36181       36181   tD(GUC)N   PAB1        1.192252        0.488894   No\n",
       "36182       36182    YAR010C   PAB1        0.397417        0.570376   No\n",
       "36183       36183  YBR182C-A   PAB1        0.397417        0.570376   No\n",
       "36184       36184  YJR005C-A   PAB1        0.119225        0.570376   No\n",
       "36185       36185  tF(GAA)P2   PAB1        0.596126        0.570376   No\n",
       "\n",
       "[5458 rows x 6 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Foldchange_File = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/TotalMutant-RNAandRPF-FoldChange-rbind-uORF.txt')\n",
    "Foldchange_File[(Foldchange_File['sample'] == 'PAB1')&(Foldchange_File['uORF'] == 'No')]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. UMAP and KEGG pathway"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " ### a. Differentialy Expressed Gene and KEGG pathway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>HO_1</th>\n",
       "      <th>HO_2</th>\n",
       "      <th>PAB1_1</th>\n",
       "      <th>PAB1_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YLR044C</td>\n",
       "      <td>55981</td>\n",
       "      <td>55265</td>\n",
       "      <td>36298</td>\n",
       "      <td>48283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YGR192C</td>\n",
       "      <td>53449</td>\n",
       "      <td>62778</td>\n",
       "      <td>47944</td>\n",
       "      <td>57807</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YHR174W</td>\n",
       "      <td>51074</td>\n",
       "      <td>55180</td>\n",
       "      <td>29990</td>\n",
       "      <td>38110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YKL060C</td>\n",
       "      <td>49285</td>\n",
       "      <td>55322</td>\n",
       "      <td>40700</td>\n",
       "      <td>54913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YOL086C</td>\n",
       "      <td>35230</td>\n",
       "      <td>37782</td>\n",
       "      <td>26127</td>\n",
       "      <td>32290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6356</th>\n",
       "      <td>YJL089W</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6357</th>\n",
       "      <td>YKL102C</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6358</th>\n",
       "      <td>YLR154W-F</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6359</th>\n",
       "      <td>YLR255C</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6360</th>\n",
       "      <td>YLR307W</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>12</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6361 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene   HO_1   HO_2  PAB1_1  PAB1_2\n",
       "0       YLR044C  55981  55265   36298   48283\n",
       "1       YGR192C  53449  62778   47944   57807\n",
       "2       YHR174W  51074  55180   29990   38110\n",
       "3       YKL060C  49285  55322   40700   54913\n",
       "4       YOL086C  35230  37782   26127   32290\n",
       "...         ...    ...    ...     ...     ...\n",
       "6356    YJL089W      1      4       6       9\n",
       "6357    YKL102C      1      1       2       3\n",
       "6358  YLR154W-F      1      2       2       3\n",
       "6359    YLR255C      1      6       2       4\n",
       "6360    YLR307W      1      6      12       8\n",
       "\n",
       "[6361 rows x 5 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Merge the RPF Count (All reads length) into a single file (matrix) for differential expression analysis.\n",
    "\n",
    "SeqType = 'RNA'\n",
    "path1 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/' + SeqType + '_Abundance/'\n",
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/4-DiffExpressGene/ColData_and_CountDataframe/' + SeqType + '/'\n",
    "sample = 'PAB1'\n",
    "\n",
    "HO_1_RPF_RNA_Count = pd.read_table((path1 + 'HO' + '/' + 'HO' + '-1_ReadCount.txt'), usecols = ['gene','0'])\n",
    "HO_1_RPF_RNA_Count.columns = ['gene','HO_1']\n",
    "HO_2_RPF_RNA_Count = pd.read_table((path1 + 'HO' + '/' + 'HO' + '-2_ReadCount.txt'), usecols = ['gene','0'])\n",
    "HO_2_RPF_RNA_Count.columns = ['gene','HO_2']\n",
    "\n",
    "Mutant_1_RPF_RNA_Count = pd.read_table((path1 + sample + '/' + sample + '-1_ReadCount.txt'), usecols = ['gene','0'])\n",
    "Mutant_1_RPF_RNA_Count.columns = ['gene',(sample + '_1')]\n",
    "Mutant_2_RPF_RNA_Count = pd.read_table((path1 + sample + '/' + sample + '-2_ReadCount.txt'), usecols = ['gene','0'])\n",
    "Mutant_2_RPF_RNA_Count.columns = ['gene',(sample + '_2')]\n",
    "\n",
    "HO_RPF_RNA_Count = pd.merge(HO_1_RPF_RNA_Count,HO_2_RPF_RNA_Count,on = 'gene')\n",
    "Mutant_RPF_RNA_Count = pd.merge(Mutant_1_RPF_RNA_Count,Mutant_2_RPF_RNA_Count,on = 'gene')\n",
    "\n",
    "RPF_RNA_Count = pd.merge(HO_RPF_RNA_Count,Mutant_RPF_RNA_Count,on = 'gene')\n",
    "RPF_RNA_Count.to_csv((path2 + sample + '_CountData.txt'), sep = '\\t')\n",
    "RPF_RNA_Count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Element 'CAF20' removed from the list.\n",
      "Element 'EAP1' removed from the list.\n",
      "Element 'eIF4E' removed from the list.\n",
      "Element 'eIF4G1' removed from the list.\n",
      "Element 'PAB1' removed from the list.\n",
      "Element 'eIF1' removed from the list.\n",
      "Element 'eIF1A' removed from the list.\n",
      "Element 'eIF4A' removed from the list.\n",
      "Element 'eIF4B' removed from the list.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>row</th>\n",
       "      <th>group</th>\n",
       "      <th>type</th>\n",
       "      <th>sample</th>\n",
       "      <th>Unique</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YLR044C</td>\n",
       "      <td>Up</td>\n",
       "      <td>Significant</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YKL060C</td>\n",
       "      <td>Up</td>\n",
       "      <td>Significant</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RDN37-1</td>\n",
       "      <td>Down</td>\n",
       "      <td>Significant</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YKL056C</td>\n",
       "      <td>Down</td>\n",
       "      <td>Significant</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RDN25-1</td>\n",
       "      <td>Down</td>\n",
       "      <td>Significant</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2408</th>\n",
       "      <td>YOL140W</td>\n",
       "      <td>Up</td>\n",
       "      <td>Significant</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2409</th>\n",
       "      <td>tK(CUU)M</td>\n",
       "      <td>Up</td>\n",
       "      <td>Significant</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2410</th>\n",
       "      <td>YBR013C</td>\n",
       "      <td>Up</td>\n",
       "      <td>Significant</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2411</th>\n",
       "      <td>snR19</td>\n",
       "      <td>Up</td>\n",
       "      <td>Significant</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2412</th>\n",
       "      <td>YDL227C</td>\n",
       "      <td>Up</td>\n",
       "      <td>Significant</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2413 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           row group         type sample Unique\n",
       "0      YLR044C    Up  Significant  CAF20     No\n",
       "1      YKL060C    Up  Significant  CAF20     No\n",
       "2      RDN37-1  Down  Significant  CAF20     No\n",
       "3      YKL056C  Down  Significant  CAF20     No\n",
       "4      RDN25-1  Down  Significant  CAF20     No\n",
       "...        ...   ...          ...    ...    ...\n",
       "2408   YOL140W    Up  Significant  eIF4B    Yes\n",
       "2409  tK(CUU)M    Up  Significant  eIF4B     No\n",
       "2410   YBR013C    Up  Significant  eIF4B     No\n",
       "2411     snR19    Up  Significant  eIF4B     No\n",
       "2412   YDL227C    Up  Significant  eIF4B     No\n",
       "\n",
       "[2413 rows x 5 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/4-DiffExpressGene/'\n",
    "\n",
    "MutantList = ['CAF20','EAP1','eIF4E','eIF4G1','PAB1','eIF1','eIF1A','eIF4A','eIF4B']\n",
    "\n",
    "for i in MutantList:\n",
    "    locals()[(i + '_Diff')] = pd.read_table((path + 'ColData_and_CountDataframe/Ribo/' + i + '_DEseq2_TypeList.txt'))\n",
    "    locals()[(i + '_Diff')] = locals()[(i + '_Diff')][locals()[(i + '_Diff')]['type'] == 'Significant'].reset_index(drop = True)\n",
    "    locals()[(i + '_Diff')]['sample'] = i\n",
    "    \n",
    "    OtherStrain_GeneList = []\n",
    "    MutantList_copy = MutantList.copy()\n",
    "\n",
    "    if i in MutantList_copy:\n",
    "        MutantList_copy.remove(i)\n",
    "        print(f\"Element '{i}' removed from the list.\")\n",
    "    else:\n",
    "        print(f\"Element '{i}' not found in the list.\")\n",
    "    \n",
    "    for j in MutantList_copy:\n",
    "        locals()[(j + '_Subset')] = pd.read_table((path + 'ColData_and_CountDataframe/Ribo/' + j + '_DEseq2_TypeList.txt'))\n",
    "        locals()[(j + '_Subset')] = locals()[(j + '_Subset')][locals()[(j + '_Subset')]['type'] == 'Significant']\n",
    "        OtherStrain_GeneList.extend(list(locals()[(j + '_Subset')]['row'].unique()))\n",
    "\n",
    "    locals()[(i + '_Diff')]['Unique'] = ''\n",
    "    for n in range(len(locals()[(i + '_Diff')]['Unique'])):\n",
    "        gene = locals()[(i + '_Diff')]['row'][n]\n",
    "        if gene in OtherStrain_GeneList:\n",
    "            locals()[(i + '_Diff')]['Unique'][n] = 'No'\n",
    "        elif gene not in OtherStrain_GeneList:\n",
    "            locals()[(i + '_Diff')]['Unique'][n] = 'Yes'\n",
    "    \n",
    "    locals()[(i + '_Diff')].to_csv((path + 'ColData_and_CountDataframe/Ribo/' + i + '_DEseq2_TypeList_unique.txt'), sep = '\\t')\n",
    "\n",
    "DiffExpressGene = pd.concat([CAF20_Diff,EAP1_Diff,eIF4E_Diff,eIF4G1_Diff,\\\n",
    "                             PAB1_Diff,eIF1_Diff,eIF1A_Diff,eIF4A_Diff,eIF4B_Diff]).reset_index(drop = True)\n",
    "DiffExpressGene.to_csv((path + 'ColData_and_CountDataframe/Ribo/' + 'AllSample_DEseq2_TypeList_Unique.txt'), sep = '\\t')\n",
    "DiffExpressGene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_191002/1146808175.py:12: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  locals()[sample + '_KEGG']['Term'][i] = locals()[sample + '_KEGG']['Term'][i].split(' - ')[0]\n",
      "/tmp/ipykernel_191002/1146808175.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  locals()[sample + '_KEGG']['Term'][i] = (locals()[sample + '_KEGG']['Term'][i].split(' - ')[0] + '-' + locals()[sample + '_KEGG']['Term'][i].split(' - ')[1])\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CAF20</th>\n",
       "      <th>Term</th>\n",
       "      <th>EAP1</th>\n",
       "      <th>eIF4E</th>\n",
       "      <th>eIF4G1</th>\n",
       "      <th>PAB1</th>\n",
       "      <th>eIF1</th>\n",
       "      <th>eIF1A</th>\n",
       "      <th>eIF4A</th>\n",
       "      <th>eIF4B</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Nucleotide metabolism</td>\n",
       "      <td>0.001326</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Ribosome</td>\n",
       "      <td>0.001740</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.318967e-06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Biosynthesis of secondary metabolites</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003810</td>\n",
       "      <td>4.741484e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.027062e-06</td>\n",
       "      <td>0.015612</td>\n",
       "      <td>0.000023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Biosynthesis of cofactors</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003810</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Glycolysis / Gluconeogenesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003810</td>\n",
       "      <td>9.299674e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.315315e-07</td>\n",
       "      <td>0.000367</td>\n",
       "      <td>0.000014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Protein processing in endoplasmic reticulum</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.003810</td>\n",
       "      <td>4.289198e-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Various types of N-glycan biosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.007854</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Amino sugar and nucleotide sugar metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010133</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.789296e-03</td>\n",
       "      <td>0.014415</td>\n",
       "      <td>0.003645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Fructose and mannose metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.012910</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.713044e-03</td>\n",
       "      <td>0.029639</td>\n",
       "      <td>0.000468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Biosynthesis of nucleotide sugars</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.012910</td>\n",
       "      <td>3.657908e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.205388e-04</td>\n",
       "      <td>0.004643</td>\n",
       "      <td>0.000468</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Other types of O-glycan biosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.023769</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Carbon metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.026128</td>\n",
       "      <td>9.299674e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.559125e-06</td>\n",
       "      <td>0.038767</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Steroid biosynthesis</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.040672</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Biosynthesis of amino acids</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.041983</td>\n",
       "      <td>9.966787e-04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.611209e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000891</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Glycerolipid metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.027213</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NaN</td>\n",
       "      <td>MAPK signaling pathway-yeast</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.047660</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Galactose metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.789296e-03</td>\n",
       "      <td>0.004643</td>\n",
       "      <td>0.000068</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Glyoxylate and dicarboxylate metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.555173e-03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.042763</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Starch and sucrose metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.356223e-02</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>0.000001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Citrate cycle (TCA cycle)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.112002e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.012055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Longevity regulating pathway-multiple species</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.006069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Cysteine and methionine metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.010701</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Proteasome</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.021123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Glutathione metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.021123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>Sulfur metabolism</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.039073</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    CAF20                                           Term      EAP1     eIF4E  \\\n",
       "0     NaN                          Nucleotide metabolism  0.001326       NaN   \n",
       "1     NaN                                       Ribosome  0.001740       NaN   \n",
       "2     NaN          Biosynthesis of secondary metabolites       NaN  0.003810   \n",
       "3     NaN                      Biosynthesis of cofactors       NaN  0.003810   \n",
       "4     NaN                   Glycolysis / Gluconeogenesis       NaN  0.003810   \n",
       "5     NaN    Protein processing in endoplasmic reticulum       NaN  0.003810   \n",
       "6     NaN         Various types of N-glycan biosynthesis       NaN  0.007854   \n",
       "7     NaN    Amino sugar and nucleotide sugar metabolism       NaN  0.010133   \n",
       "8     NaN                Fructose and mannose metabolism       NaN  0.012910   \n",
       "9     NaN              Biosynthesis of nucleotide sugars       NaN  0.012910   \n",
       "10    NaN           Other types of O-glycan biosynthesis       NaN  0.023769   \n",
       "11    NaN                              Carbon metabolism       NaN  0.026128   \n",
       "12    NaN                           Steroid biosynthesis       NaN  0.040672   \n",
       "13    NaN                    Biosynthesis of amino acids       NaN  0.041983   \n",
       "14    NaN                        Glycerolipid metabolism       NaN       NaN   \n",
       "15    NaN                   MAPK signaling pathway-yeast       NaN       NaN   \n",
       "16    NaN                           Galactose metabolism       NaN       NaN   \n",
       "17    NaN        Glyoxylate and dicarboxylate metabolism       NaN       NaN   \n",
       "18    NaN                  Starch and sucrose metabolism       NaN       NaN   \n",
       "19    NaN                      Citrate cycle (TCA cycle)       NaN       NaN   \n",
       "20    NaN  Longevity regulating pathway-multiple species       NaN       NaN   \n",
       "21    NaN             Cysteine and methionine metabolism       NaN       NaN   \n",
       "22    NaN                                     Proteasome       NaN       NaN   \n",
       "23    NaN                         Glutathione metabolism       NaN       NaN   \n",
       "24    NaN                              Sulfur metabolism       NaN       NaN   \n",
       "\n",
       "          eIF4G1  PAB1      eIF1         eIF1A     eIF4A     eIF4B  \n",
       "0            NaN   NaN       NaN           NaN       NaN       NaN  \n",
       "1   3.318967e-06   NaN       NaN           NaN       NaN       NaN  \n",
       "2   4.741484e-07   NaN       NaN  2.027062e-06  0.015612  0.000023  \n",
       "3            NaN   NaN       NaN           NaN       NaN       NaN  \n",
       "4   9.299674e-04   NaN       NaN  8.315315e-07  0.000367  0.000014  \n",
       "5   4.289198e-03   NaN       NaN           NaN       NaN       NaN  \n",
       "6            NaN   NaN       NaN           NaN       NaN       NaN  \n",
       "7            NaN   NaN       NaN  2.789296e-03  0.014415  0.003645  \n",
       "8            NaN   NaN       NaN  2.713044e-03  0.029639  0.000468  \n",
       "9   3.657908e-02   NaN       NaN  3.205388e-04  0.004643  0.000468  \n",
       "10           NaN   NaN       NaN           NaN       NaN       NaN  \n",
       "11  9.299674e-04   NaN       NaN  4.559125e-06  0.038767  0.000001  \n",
       "12           NaN   NaN       NaN           NaN       NaN       NaN  \n",
       "13  9.966787e-04   NaN       NaN  2.611209e-02       NaN  0.000891  \n",
       "14           NaN   NaN  0.027213           NaN       NaN       NaN  \n",
       "15           NaN   NaN  0.047660           NaN       NaN       NaN  \n",
       "16           NaN   NaN       NaN  2.789296e-03  0.004643  0.000068  \n",
       "17           NaN   NaN       NaN  6.555173e-03       NaN  0.042763  \n",
       "18           NaN   NaN       NaN  3.356223e-02  0.000006  0.000001  \n",
       "19           NaN   NaN       NaN  4.112002e-02       NaN  0.012055  \n",
       "20           NaN   NaN       NaN           NaN       NaN  0.006069  \n",
       "21           NaN   NaN       NaN           NaN       NaN  0.010701  \n",
       "22           NaN   NaN       NaN           NaN       NaN  0.021123  \n",
       "23           NaN   NaN       NaN           NaN       NaN  0.021123  \n",
       "24           NaN   NaN       NaN           NaN       NaN  0.039073  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### KEGG pathway\n",
    "\n",
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/4-DiffExpressGene/KEGG_pathway/Ribo/'\n",
    "MutantList = ['CAF20','EAP1','eIF4E','eIF4G1','PAB1','eIF1','eIF1A','eIF4A','eIF4B']\n",
    "\n",
    "for sample in MutantList:\n",
    "    locals()[sample + '_KEGG'] = pd.read_table((path + sample + '_KEGG_Down.txt'), usecols = ['Description','qvalue'])\n",
    "    locals()[sample + '_KEGG'].columns = ['Term',sample]\n",
    "    locals()[sample + '_KEGG'] = locals()[sample + '_KEGG'][locals()[sample + '_KEGG'][sample] <= 0.05]\n",
    "    for i in range(len(locals()[sample + '_KEGG']['Term'])):\n",
    "        if len(locals()[sample + '_KEGG']['Term'][i].split(' - ')) == 2:\n",
    "            locals()[sample + '_KEGG']['Term'][i] = locals()[sample + '_KEGG']['Term'][i].split(' - ')[0]\n",
    "        if len(locals()[sample + '_KEGG']['Term'][i].split(' - ')) > 2:\n",
    "            locals()[sample + '_KEGG']['Term'][i] = (locals()[sample + '_KEGG']['Term'][i].split(' - ')[0] + '-' + locals()[sample + '_KEGG']['Term'][i].split(' - ')[1])\n",
    "\n",
    "KEGG_merge = pd.merge(locals()['CAF20_KEGG'],locals()['EAP1_KEGG'],on = ['Term'], how = 'outer')\n",
    "\n",
    "MutantList.remove('CAF20')\n",
    "MutantList.remove('EAP1')\n",
    "\n",
    "for sample in MutantList:\n",
    "    KEGG_merge = pd.merge(KEGG_merge,locals()[sample + '_KEGG'],on = ['Term'], how = 'outer')\n",
    "\n",
    "#KEGG_merge['eIF1'] = 1\n",
    "KEGG_merge.set_index('Term').to_csv((path + 'AllSample_KEGG_FDR_Down_0.05.txt'), sep = '\\t', na_rep = 1)\n",
    "KEGG_merge"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 4. Features in different regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. Reads fraction in each region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_191002/3260054085.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uORF_Yes['fraction'] = uORF_Yes['ReadsStartPos'] / uORF_Yes['ReadsStartPos'].sum()\n",
      "/tmp/ipykernel_191002/3260054085.py:38: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uORF_Yes['Group'] = 'Yes'\n",
      "/tmp/ipykernel_191002/3260054085.py:40: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uORF_No['fraction'] = uORF_No['ReadsStartPos'] / uORF_No['ReadsStartPos'].sum()\n",
      "/tmp/ipykernel_191002/3260054085.py:41: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  uORF_No['Group'] = 'No'\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region</th>\n",
       "      <th>uORF</th>\n",
       "      <th>ReadsStartPos</th>\n",
       "      <th>sample</th>\n",
       "      <th>fraction</th>\n",
       "      <th>Group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>Yes</td>\n",
       "      <td>278</td>\n",
       "      <td>HO</td>\n",
       "      <td>0.018957</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>Yes</td>\n",
       "      <td>157</td>\n",
       "      <td>HO</td>\n",
       "      <td>0.010706</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDS</td>\n",
       "      <td>Yes</td>\n",
       "      <td>14230</td>\n",
       "      <td>HO</td>\n",
       "      <td>0.970338</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>9283</td>\n",
       "      <td>HO</td>\n",
       "      <td>0.022675</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>498</td>\n",
       "      <td>HO</td>\n",
       "      <td>0.001216</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>399615</td>\n",
       "      <td>HO</td>\n",
       "      <td>0.976109</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  region uORF  ReadsStartPos sample  fraction Group\n",
       "0   3UTR  Yes            278     HO  0.018957   Yes\n",
       "1   5UTR  Yes            157     HO  0.010706   Yes\n",
       "2    CDS  Yes          14230     HO  0.970338   Yes\n",
       "3   3UTR   No           9283     HO  0.022675    No\n",
       "4   5UTR   No            498     HO  0.001216    No\n",
       "5    CDS   No         399615     HO  0.976109    No"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Distinguishes whether a gene has a uORF\n",
    "\n",
    "### 28nt RPF used in Ribo-seq\n",
    "### all reads used in RNA-seq\n",
    "\n",
    "sample = 'HO'\n",
    "#rep = '1'\n",
    "name = sample# + '-' + rep\n",
    "SeqType = 'Ribo'\n",
    "path1 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity-Region/' + SeqType + '/' + sample + '/') ## The pathway of _GeneRegion_AllReadsPosition.txt\n",
    "path2 = ('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/5-ReadsFraction-Region/' + SeqType + '/' + sample + '/') ## The pathway to output results\n",
    "\n",
    "ReadsPosition = pd.read_table((path1 + name + '_GeneRegion_AllReadsPosition.txt'), usecols = ['ReadsStartPos','ReadsEndPos','gene','region'])\n",
    "ReadsPosition['length'] = ReadsPosition['ReadsEndPos'] - ReadsPosition['ReadsStartPos'] + 1\n",
    "ReadsPosition = ReadsPosition[ReadsPosition['length'] == 28].reset_index(drop = True) \n",
    "\n",
    "Ingolia_uORF_Gene = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uORF_Gene_FristPos_2009-ingolia.txt')\n",
    "Ingolia_uORF_GeneList = Ingolia_uORF_Gene['Gene'].unique()\n",
    "len(Ingolia_uORF_GeneList)\n",
    "\n",
    "ReadsCount_Region_Gene_Length = ReadsPosition.groupby(['region','gene','length'])['ReadsStartPos'].count().reset_index()\n",
    "\n",
    "def assign_value(ReadsCount_Region_Gene_Length):\n",
    "    if ReadsCount_Region_Gene_Length['gene'] in Ingolia_uORF_GeneList:\n",
    "        return 'Yes'\n",
    "    else:\n",
    "        return 'No'\n",
    "\n",
    "ReadsCount_Region_Gene_Length['uORF'] = ReadsCount_Region_Gene_Length.apply(assign_value, axis=1) \n",
    "\n",
    "ReadsCount_Region_Gene_Length_uORF = ReadsCount_Region_Gene_Length.groupby(['region','uORF'])['ReadsStartPos'].sum().reset_index()\n",
    "ReadsCount_Region_Gene_Length_uORF['sample'] = sample\n",
    "#ReadsCount_Region_Gene_Length_uORF['rep'] = ('rep' + rep)\n",
    "ReadsCount_Region_Gene_Length_uORF.to_csv((path2 + name + '_RegionReadsCount.txt'), sep = '\\t')\n",
    "\n",
    "uORF_Yes = ReadsCount_Region_Gene_Length_uORF[(ReadsCount_Region_Gene_Length_uORF['uORF'] == 'Yes')]\n",
    "uORF_Yes['fraction'] = uORF_Yes['ReadsStartPos'] / uORF_Yes['ReadsStartPos'].sum()\n",
    "uORF_Yes['Group'] = 'Yes'\n",
    "uORF_No = ReadsCount_Region_Gene_Length_uORF[(ReadsCount_Region_Gene_Length_uORF['uORF'] == 'No')]\n",
    "uORF_No['fraction'] = uORF_No['ReadsStartPos'] / uORF_No['ReadsStartPos'].sum()\n",
    "uORF_No['Group'] = 'No'\n",
    "\n",
    "uORF_Length = pd.concat([uORF_Yes,uORF_No]).reset_index(drop = True)\n",
    "uORF_Length.to_csv((path2 + name + '_RegionReadsCount_LengthuORF_Fraction.txt'), sep = '\\t')\n",
    "uORF_Length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region</th>\n",
       "      <th>uORF</th>\n",
       "      <th>sample</th>\n",
       "      <th>rep</th>\n",
       "      <th>fraction</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>HO</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.023232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>HO</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.001145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>HO</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.975623</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>HO</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.022498</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>HO</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.001239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>HO</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.976263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.024574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.001612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.973813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.021184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.001420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.977396</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>EAP1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.022624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>EAP1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.002160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>EAP1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.975216</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>EAP1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.019908</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>EAP1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.001155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>EAP1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.978937</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4E</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.020803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4E</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.001695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4E</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.977502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4E</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.018443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4E</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.001283</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4E</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.980273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.033678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.002264</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.964058</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.026184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.001039</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.972777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.022561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.001206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.976232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.025403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.002345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.972252</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.021359</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.002545</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.976096</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.022412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.002121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.975466</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1A</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.021905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1A</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.005915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1A</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.972180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1A</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.018932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1A</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.003919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF1A</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.977149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4A</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.023076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4A</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.003268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4A</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.973656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4A</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.021739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4A</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.001638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4A</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.976624</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.020357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.001367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>rep1</td>\n",
       "      <td>0.978276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>3UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.020383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>5UTR</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.003228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>CDS</td>\n",
       "      <td>No</td>\n",
       "      <td>eIF4B</td>\n",
       "      <td>rep2</td>\n",
       "      <td>0.976388</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   region uORF  sample   rep  fraction\n",
       "0    3UTR   No      HO  rep1  0.023232\n",
       "1    5UTR   No      HO  rep1  0.001145\n",
       "2     CDS   No      HO  rep1  0.975623\n",
       "3    3UTR   No      HO  rep2  0.022498\n",
       "4    5UTR   No      HO  rep2  0.001239\n",
       "5     CDS   No      HO  rep2  0.976263\n",
       "6    3UTR   No   CAF20  rep1  0.024574\n",
       "7    5UTR   No   CAF20  rep1  0.001612\n",
       "8     CDS   No   CAF20  rep1  0.973813\n",
       "9    3UTR   No   CAF20  rep2  0.021184\n",
       "10   5UTR   No   CAF20  rep2  0.001420\n",
       "11    CDS   No   CAF20  rep2  0.977396\n",
       "12   3UTR   No    EAP1  rep1  0.022624\n",
       "13   5UTR   No    EAP1  rep1  0.002160\n",
       "14    CDS   No    EAP1  rep1  0.975216\n",
       "15   3UTR   No    EAP1  rep2  0.019908\n",
       "16   5UTR   No    EAP1  rep2  0.001155\n",
       "17    CDS   No    EAP1  rep2  0.978937\n",
       "18   3UTR   No   eIF4E  rep1  0.020803\n",
       "19   5UTR   No   eIF4E  rep1  0.001695\n",
       "20    CDS   No   eIF4E  rep1  0.977502\n",
       "21   3UTR   No   eIF4E  rep2  0.018443\n",
       "22   5UTR   No   eIF4E  rep2  0.001283\n",
       "23    CDS   No   eIF4E  rep2  0.980273\n",
       "24   3UTR   No  eIF4G1  rep1  0.033678\n",
       "25   5UTR   No  eIF4G1  rep1  0.002264\n",
       "26    CDS   No  eIF4G1  rep1  0.964058\n",
       "27   3UTR   No  eIF4G1  rep2  0.026184\n",
       "28   5UTR   No  eIF4G1  rep2  0.001039\n",
       "29    CDS   No  eIF4G1  rep2  0.972777\n",
       "30   3UTR   No    PAB1  rep1  0.022561\n",
       "31   5UTR   No    PAB1  rep1  0.001206\n",
       "32    CDS   No    PAB1  rep1  0.976232\n",
       "33   3UTR   No    PAB1  rep2  0.025403\n",
       "34   5UTR   No    PAB1  rep2  0.002345\n",
       "35    CDS   No    PAB1  rep2  0.972252\n",
       "36   3UTR   No    eIF1  rep1  0.021359\n",
       "37   5UTR   No    eIF1  rep1  0.002545\n",
       "38    CDS   No    eIF1  rep1  0.976096\n",
       "39   3UTR   No    eIF1  rep2  0.022412\n",
       "40   5UTR   No    eIF1  rep2  0.002121\n",
       "41    CDS   No    eIF1  rep2  0.975466\n",
       "42   3UTR   No   eIF1A  rep1  0.021905\n",
       "43   5UTR   No   eIF1A  rep1  0.005915\n",
       "44    CDS   No   eIF1A  rep1  0.972180\n",
       "45   3UTR   No   eIF1A  rep2  0.018932\n",
       "46   5UTR   No   eIF1A  rep2  0.003919\n",
       "47    CDS   No   eIF1A  rep2  0.977149\n",
       "48   3UTR   No   eIF4A  rep1  0.023076\n",
       "49   5UTR   No   eIF4A  rep1  0.003268\n",
       "50    CDS   No   eIF4A  rep1  0.973656\n",
       "51   3UTR   No   eIF4A  rep2  0.021739\n",
       "52   5UTR   No   eIF4A  rep2  0.001638\n",
       "53    CDS   No   eIF4A  rep2  0.976624\n",
       "54   3UTR   No   eIF4B  rep1  0.020357\n",
       "55   5UTR   No   eIF4B  rep1  0.001367\n",
       "56    CDS   No   eIF4B  rep1  0.978276\n",
       "57   3UTR   No   eIF4B  rep2  0.020383\n",
       "58   5UTR   No   eIF4B  rep2  0.003228\n",
       "59    CDS   No   eIF4B  rep2  0.976388"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/5-ReadsFraction-Region/Ribo/'\n",
    "\n",
    "MutantList = ['HO','CAF20','EAP1','eIF4E','eIF4G1','PAB1','eIF1','eIF1A','eIF4A','eIF4B']\n",
    "\n",
    "for i in MutantList:\n",
    "    locals()[('Reads_Fraction_' + i + '_1')] = pd.read_table((path + i + '/' + i + '-1_RegionReadsCount_LengthuORF_Fraction.txt'), usecols = ['region','uORF','sample','rep','fraction'])\n",
    "    locals()[('Reads_Fraction_' + i + '_1')] = locals()[('Reads_Fraction_' + i + '_1')][locals()[('Reads_Fraction_' + i + '_1')]['uORF'] == 'No']\n",
    "    locals()[('Reads_Fraction_' + i + '_2')] = pd.read_table((path + i + '/' + i + '-2_RegionReadsCount_LengthuORF_Fraction.txt'), usecols = ['region','uORF','sample','rep','fraction'])\n",
    "    locals()[('Reads_Fraction_' + i + '_2')] = locals()[('Reads_Fraction_' + i + '_2')][locals()[('Reads_Fraction_' + i + '_2')]['uORF'] == 'No']\n",
    "\n",
    "\n",
    "Reads_Fraction = pd.concat([Reads_Fraction_HO_1,Reads_Fraction_HO_2,Reads_Fraction_CAF20_1,Reads_Fraction_CAF20_2,Reads_Fraction_EAP1_1,Reads_Fraction_EAP1_2,\\\n",
    "                   Reads_Fraction_eIF4E_1,Reads_Fraction_eIF4E_2,Reads_Fraction_eIF4G1_1,Reads_Fraction_eIF4G1_2,Reads_Fraction_PAB1_1,Reads_Fraction_PAB1_2,\\\n",
    "                    Reads_Fraction_eIF1_1,Reads_Fraction_eIF1_2,Reads_Fraction_eIF1A_1,Reads_Fraction_eIF1A_2,Reads_Fraction_eIF4A_1,Reads_Fraction_eIF4A_2,\\\n",
    "                        Reads_Fraction_eIF4B_1,Reads_Fraction_eIF4B_2]).reset_index(drop = True)\n",
    "Reads_Fraction.to_csv((path + 'AllSample_Reads_Fraction_28nt_NOuORF.txt'),sep = '\\t')\n",
    "Reads_Fraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 5. Identification of genes co-regulated by cooperative factors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. eIF4E/eIF4G1/PAB1/EAP1/CAF20 vs. eIF1/eIF1A/eIF4A/eIF4B"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cooperative Gene identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Dell/Dell15/zhanggy/bin/anaconda3/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3757: RuntimeWarning: Degrees of freedom <= 0 for slice\n",
      "  return _methods._var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/Dell/Dell15/zhanggy/bin/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:256: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ret = ret.dtype.type(ret / rcount)\n"
     ]
    }
   ],
   "source": [
    "path1 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/' ## The pathway of -RibosomeDensity.txt\n",
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/8-CooperativeGene/Five_vs_Four/' ## The pathway to output results\n",
    "\n",
    "### Label 'eIF4E','eIF4G1','PAB1','EAP1','CAF20' as group1\n",
    "group1_List = ['eIF4E','eIF4G1','PAB1','EAP1','CAF20']\n",
    "\n",
    "for i in group1_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "group1 = pd.concat([locals()['eIF4E'],locals()['eIF4G1'],locals()['PAB1'],locals()['EAP1'],locals()['CAF20']])\n",
    "group1['type'] = 'group1'\n",
    "\n",
    "### Label 'eIF1','eIF1A','eIF4A','eIF4B' as group2\n",
    "group2_List = ['eIF1','eIF1A','eIF4A','eIF4B']\n",
    "\n",
    "for i in group2_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "group2 = pd.concat([locals()['eIF1'],locals()['eIF1A'],locals()['eIF4A'],locals()['eIF4B']])\n",
    "group2['type'] = 'group2'\n",
    "\n",
    "total_cluster = pd.concat([group1,group2])\n",
    "gene1 = total_cluster['gene'].unique()\n",
    "\n",
    "### Only consider genes with 5'UTR and 3'UTR annotations and 5'UTR length > 1\n",
    "UTR_gene = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/saccharomyces_cerevisiae_information/Saccharomyces_cerevisiae_UTRInformation.txt')\n",
    "UTR3_gene = UTR_gene[UTR_gene['type'] == 'three_prime_utr']['gene_id'].unique()\n",
    "UTR_5 = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/APOBEC1_data/preview/1-GeneIdentification-Overview/5UTR_seq.txt')\n",
    "UTR_5['length'] = UTR_5['stop'] - UTR_5['start'] + 1\n",
    "UTR_5 = UTR_5[UTR_5['length'] != 1]\n",
    "UTR5_gene = UTR_5['gene_id'].unique()\n",
    "gene2 = list(set(UTR5_gene).intersection(set(UTR3_gene)))\n",
    "gene = list(set(gene1).intersection(set(gene2)))\n",
    "\n",
    "### Genes with significantly different Ribosome Density between group1 and group2 were calculated using t.test\n",
    "\n",
    "gene_pval = open((path2 + 'Ttest_Cluster.txt'),'w')\n",
    "gene_pval.write('gene\\ttstat\\tpval\\n')\n",
    "\n",
    "for i in gene:\n",
    "    gene_subset = total_cluster[total_cluster['gene'] == i]\n",
    "    gene_subset_group1 = gene_subset[gene_subset['type'] == 'group1']['Ribosome_Density'].tolist()\n",
    "    gene_subset_group2 = gene_subset[gene_subset['type'] == 'group2']['Ribosome_Density'].tolist()\n",
    "    tstat, pval = stats.ttest_ind(a = gene_subset_group1, b = gene_subset_group2, alternative = \"two-sided\")\n",
    "    gene_pval.write(i + '\\t' + str(tstat) + '\\t' + str(pval) + '\\n')\n",
    "gene_pval.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_69781/3318458651.py:41: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ttest_Cluster['group'][i] = 'Not_significant'\n",
      "/tmp/ipykernel_69781/3318458651.py:46: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ttest_Cluster['group'][i] = 'Cluster2'\n",
      "/tmp/ipykernel_69781/3318458651.py:44: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ttest_Cluster['group'][i] = 'Cluster1'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster1:  364\n",
      "Cluster2:  205\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>group1_RibosomeDensity_mean</th>\n",
       "      <th>group2_RibosomeDensity_mean</th>\n",
       "      <th>Ribosome_Density_foldchange</th>\n",
       "      <th>pval</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YLR044C</td>\n",
       "      <td>1.435611</td>\n",
       "      <td>1.312782</td>\n",
       "      <td>1.093564</td>\n",
       "      <td>0.554962</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YGR192C</td>\n",
       "      <td>0.791029</td>\n",
       "      <td>0.891151</td>\n",
       "      <td>0.887649</td>\n",
       "      <td>0.271266</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YKL060C</td>\n",
       "      <td>1.698312</td>\n",
       "      <td>1.649720</td>\n",
       "      <td>1.029455</td>\n",
       "      <td>0.810382</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YHR174W</td>\n",
       "      <td>1.457214</td>\n",
       "      <td>1.867915</td>\n",
       "      <td>0.780129</td>\n",
       "      <td>0.083500</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YLR075W</td>\n",
       "      <td>0.859427</td>\n",
       "      <td>0.908874</td>\n",
       "      <td>0.945594</td>\n",
       "      <td>0.740490</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4261</th>\n",
       "      <td>YER011W</td>\n",
       "      <td>1.551796</td>\n",
       "      <td>1.800107</td>\n",
       "      <td>0.862058</td>\n",
       "      <td>0.476815</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4262</th>\n",
       "      <td>YPR145C-A</td>\n",
       "      <td>2.088344</td>\n",
       "      <td>1.059272</td>\n",
       "      <td>1.971489</td>\n",
       "      <td>0.477519</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4263</th>\n",
       "      <td>YJL037W</td>\n",
       "      <td>1.400486</td>\n",
       "      <td>0.434431</td>\n",
       "      <td>3.223723</td>\n",
       "      <td>0.052017</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4264</th>\n",
       "      <td>YML058W-A</td>\n",
       "      <td>1.646590</td>\n",
       "      <td>1.439087</td>\n",
       "      <td>1.144191</td>\n",
       "      <td>0.497375</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4265</th>\n",
       "      <td>YNL093W</td>\n",
       "      <td>1.641978</td>\n",
       "      <td>0.769624</td>\n",
       "      <td>2.133482</td>\n",
       "      <td>0.208727</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4266 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene  group1_RibosomeDensity_mean  group2_RibosomeDensity_mean  \\\n",
       "0       YLR044C                     1.435611                     1.312782   \n",
       "1       YGR192C                     0.791029                     0.891151   \n",
       "2       YKL060C                     1.698312                     1.649720   \n",
       "3       YHR174W                     1.457214                     1.867915   \n",
       "4       YLR075W                     0.859427                     0.908874   \n",
       "...         ...                          ...                          ...   \n",
       "4261    YER011W                     1.551796                     1.800107   \n",
       "4262  YPR145C-A                     2.088344                     1.059272   \n",
       "4263    YJL037W                     1.400486                     0.434431   \n",
       "4264  YML058W-A                     1.646590                     1.439087   \n",
       "4265    YNL093W                     1.641978                     0.769624   \n",
       "\n",
       "      Ribosome_Density_foldchange      pval            group  \n",
       "0                        1.093564  0.554962  Not_significant  \n",
       "1                        0.887649  0.271266  Not_significant  \n",
       "2                        1.029455  0.810382  Not_significant  \n",
       "3                        0.780129  0.083500  Not_significant  \n",
       "4                        0.945594  0.740490  Not_significant  \n",
       "...                           ...       ...              ...  \n",
       "4261                     0.862058  0.476815  Not_significant  \n",
       "4262                     1.971489  0.477519  Not_significant  \n",
       "4263                     3.223723  0.052017  Not_significant  \n",
       "4264                     1.144191  0.497375  Not_significant  \n",
       "4265                     2.133482  0.208727  Not_significant  \n",
       "\n",
       "[4266 rows x 6 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Calculate the average Ribosome Density of genes in group1\n",
    "group1_List = ['eIF4E','eIF4G1','PAB1','EAP1','CAF20']\n",
    "\n",
    "for i in group1_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "    locals()[i].columns = ['gene',i]\n",
    "group1_List.remove('eIF4E')\n",
    "group1_List.remove('eIF4G1')\n",
    "group1 = pd.merge(locals()['eIF4E'],locals()['eIF4G1'],on = 'gene')\n",
    "for i in group1_List:\n",
    "    group1 = pd.merge(group1,locals()[i],on = 'gene')\n",
    "group1['group1_RibosomeDensity_mean'] = (group1['eIF4E'] + group1['eIF4G1'] + group1['PAB1'] + group1['EAP1'] + group1['CAF20']) / 5\n",
    "\n",
    "### Calculate the average Ribosome Density of genes in group2\n",
    "group2_List = ['eIF1','eIF1A','eIF4A','eIF4B']\n",
    "\n",
    "for i in group2_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "    locals()[i].columns = ['gene',i]\n",
    "group2_List.remove('eIF1')\n",
    "group2_List.remove('eIF1A')\n",
    "group2 = pd.merge(locals()['eIF1'],locals()['eIF1A'],on = 'gene')\n",
    "for i in group2_List:\n",
    "    group2 = pd.merge(group2,locals()[i],on = 'gene')\n",
    "group2['group2_RibosomeDensity_mean'] = (group2['eIF1'] + group2['eIF1A'] + group2['eIF4A'] + group2['eIF4B']) / 4\n",
    "\n",
    "### Measures the change in Ribosome Density of genes in group1 relative to group2\n",
    "Groups = pd.merge(group1,group2,on = 'gene')\n",
    "Groups['Ribosome_Density_foldchange'] = Groups['group1_RibosomeDensity_mean'] / Groups['group2_RibosomeDensity_mean']\n",
    "Groups.to_csv((path2 + 'Cluster_Ribosome_Density_foldchange.txt'),sep = '\\t')\n",
    "\n",
    "Ttest_Cluster = pd.read_table((path2 + 'Ttest_Cluster.txt'),sep = '\\t')\n",
    "Ttest_Cluster = pd.merge(Groups, Ttest_Cluster, on = 'gene')\n",
    "Ttest_Cluster = Ttest_Cluster[['gene','group1_RibosomeDensity_mean','group2_RibosomeDensity_mean','Ribosome_Density_foldchange','pval']]\n",
    "Ttest_Cluster = Ttest_Cluster[~Ttest_Cluster.isin([np.nan, np.inf, -np.inf]).any(1)].reset_index(drop = True)\n",
    "\n",
    "Ttest_Cluster['group'] = ''\n",
    "\n",
    "for i in range(len(Ttest_Cluster['gene'])):\n",
    "    if Ttest_Cluster['pval'][i] > 0.05:\n",
    "        Ttest_Cluster['group'][i] = 'Not_significant'\n",
    "    elif Ttest_Cluster['pval'][i] <= 0.05:\n",
    "        if Ttest_Cluster['Ribosome_Density_foldchange'][i] > 1:\n",
    "            Ttest_Cluster['group'][i] = 'Cluster1'\n",
    "        elif Ttest_Cluster['Ribosome_Density_foldchange'][i] < 1:\n",
    "            Ttest_Cluster['group'][i] = 'Cluster2'\n",
    "\n",
    "Ttest_Cluster_selected = Ttest_Cluster[Ttest_Cluster['group'] != 'Not significant'].reset_index(drop = True)\n",
    "print(\"Cluster1: \", str(len(Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster1']['gene'].unique())))\n",
    "print(\"Cluster2: \", str(len(Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster2']['gene'].unique())))\n",
    "\n",
    "Ttest_Cluster.to_csv((path2 + 'RibosomeDensity_foldchange_Pval_Cluster.txt'), sep = '\\t')\n",
    "#Ttest_Cluster\n",
    "Ttest_Cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/8-CooperativeGene/Five_vs_Four/' ## The pathway to output results\n",
    "Ttest_Cluster = pd.read_table((path2 + 'RibosomeDensity_foldchange_Pval_Cluster.txt'), usecols = ['gene','group'])\n",
    "Ttest_Cluster[Ttest_Cluster['group'] == 'Cluster1']['gene'].to_csv((path2 + 'Cluster1_GeneList.txt'), sep = '\\t')\n",
    "Ttest_Cluster[Ttest_Cluster['group'] == 'Cluster2']['gene'].to_csv((path2 + 'Cluster2_GeneList.txt'), sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Features of genes in different cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/8-CooperativeGene/Five_vs_Four/'\n",
    "Ttest_Cluster = pd.read_table((path2 + 'RibosomeDensity_foldchange_Pval_Cluster.txt'), usecols = ['gene','Ribosome_Density_foldchange','pval','group'])\n",
    "\n",
    "### Ribosome Density: Mutant and HO\n",
    "Mutant_RibosomeDensity = pd.read_table((path2 + 'Cluster_Ribosome_Density_foldchange.txt'), usecols = ['gene','group1_RibosomeDensity_mean','group2_RibosomeDensity_mean'])\n",
    "Mutant_Cluster1_RibosomeDensity = Mutant_RibosomeDensity[['gene','group2_RibosomeDensity_mean']]\n",
    "Mutant_Cluster1_RibosomeDensity['sample'] = 'Cluster1_Factor_mean'\n",
    "Mutant_Cluster1_RibosomeDensity.columns = ['gene','RibosomeDensity','sample']\n",
    "\n",
    "Mutant_Cluster2_RibosomeDensity = Mutant_RibosomeDensity[['gene','group1_RibosomeDensity_mean']]\n",
    "Mutant_Cluster2_RibosomeDensity['sample'] = 'Cluster2_Factor_mean'\n",
    "Mutant_Cluster2_RibosomeDensity.columns = ['gene','RibosomeDensity','sample']\n",
    "\n",
    "HO_RibosomeDensity = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/HO/HO-RibosomeDensity.txt',\\\n",
    "                                   usecols = ['gene','Ribosome_Density','sample'])\n",
    "HO_RibosomeDensity.columns = ['gene','RibosomeDensity','sample']\n",
    "\n",
    "RibosomeDensity = pd.concat([HO_RibosomeDensity,Mutant_Cluster1_RibosomeDensity,Mutant_Cluster2_RibosomeDensity]).reset_index(drop = True)\n",
    "RibosomeDensity = pd.merge(RibosomeDensity,Ttest_Cluster[['gene','group']], on = 'gene')\n",
    "RibosomeDensity = RibosomeDensity[RibosomeDensity['group'] != 'Not_significant'].reset_index(drop = True)\n",
    "RibosomeDensity.to_csv((path2 + 'Cluster_HOandMutant_RibosomeDensity.txt'), sep = '\\t')\n",
    "\n",
    "### Length and GC%\n",
    "\n",
    "UTR_5_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/5UTR_seq.txt', usecols = ['gene_id','seq'])\n",
    "UTR_5_annotation['Region'] = \"5'UTR\"\n",
    "\n",
    "CDS_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/CDS_seq.txt', usecols = ['gene_id','seq'])\n",
    "CDS_annotation['Region'] = \"CDS\"\n",
    "\n",
    "UTR_3_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/3UTR_seq.txt', usecols = ['gene_id','seq'])\n",
    "UTR_3_annotation['Region'] = \"3'UTR\"\n",
    "\n",
    "AllRegion_annotation = pd.concat([UTR_3_annotation,UTR_5_annotation,CDS_annotation]).reset_index(drop = True)\n",
    "\n",
    "AllRegion_annotation['length'] = ''\n",
    "AllRegion_annotation['GC'] = ''\n",
    "for i in range(len(AllRegion_annotation['gene_id'])):\n",
    "    AllRegion_annotation['length'][i] = len(AllRegion_annotation['seq'][i])\n",
    "    AllRegion_annotation['GC'][i] = (AllRegion_annotation['seq'][i].count('G') + AllRegion_annotation['seq'][i].count('C')) / len(AllRegion_annotation['seq'][i])\n",
    "AllRegion_annotation['GC'] = AllRegion_annotation['GC']*100\n",
    "AllRegion_annotation = pd.merge(AllRegion_annotation,Ttest_Cluster_selected,left_on = 'gene_id',right_on = 'gene').drop(['gene_id'],axis = 1)\n",
    "\n",
    "AllRegion_annotation[['gene','Region','length','GC','group']].to_csv((path2 + 'Cluster_Length_and_GC.txt'), sep = '\\t')\n",
    "AllRegion_annotation[['gene','Region','length','GC','group']]\n",
    "\n",
    "### Kozak Score\n",
    "\n",
    "KozakScore = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/APOBEC1_data/preview/3-NormalizeFeature/Kozak/initiation_score_6+3.txt',usecols = ['gene','score_xnt'])\n",
    "KozakScore.columns = ['gene','KozakScore']\n",
    "KozakScore = pd.merge(Ttest_Cluster,KozakScore,on = 'gene')\n",
    "KozakScore.to_csv((path2 + 'Cluster_KozakScore_6+3.txt'), sep = '\\t')\n",
    "\n",
    "### PolyA Length and nonA\n",
    "\n",
    "PolyA_length_nonA = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H_polyLength_nonAcontent.txt')\n",
    "PolyA_length_nonA = pd.merge(Ttest_Cluster,PolyA_length_nonA,left_on = 'gene',right_on = 'gene_id').drop(['gene_id','Unnamed: 0'], axis = 1)\n",
    "PolyA_length_nonA.to_csv((path2 + 'Cluster_PolyA.txt'), sep = '\\t')\n",
    "\n",
    "### transcript length\n",
    "\n",
    "transcript_length = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/AllRegionSequence.txt', usecols = ['gene_id','seq'])\n",
    "transcript_length['transcript_length'] = transcript_length['seq'].str.len()\n",
    "transcript_length = pd.merge(Ttest_Cluster,transcript_length[['gene_id','transcript_length']],left_on = 'gene_id',right_on = 'gene').drop(['gene_id'],axis = 1)\n",
    "transcript_length.to_csv((path2 + 'Cluster_transcript_length.txt'), sep = '\\t')\n",
    "\n",
    "### uORF enrichment\n",
    "\n",
    "Ttest_Cluster_selected = Ttest_Cluster[Ttest_Cluster['group'] != 'Not_significant'].reset_index()\n",
    "Ttest_Cluster_selected = Ttest_Cluster_selected[['gene','group']]\n",
    "\n",
    "Ingolia_uORF_Gene = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uORF_Gene_FristPos_2009-ingolia.txt')\n",
    "Ingolia_uORF_GeneList = Ingolia_uORF_Gene['Gene'].unique()\n",
    "Cluster1_GeneList = Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster1']['gene'].unique()\n",
    "Cluster2_GeneList = Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster2']['gene'].unique()\n",
    "print(\"Ingolia uORF: n = \", str(len(Ingolia_uORF_GeneList)))\n",
    "print(\"Cluster1: n = \", str(len(Cluster1_GeneList)))\n",
    "print(\"Cluster2: n = \", str(len(Cluster2_GeneList)))\n",
    "Cluster1_uORF_GeneList = list(set(Ingolia_uORF_GeneList).intersection(set(Cluster1_GeneList)))\n",
    "Cluster2_uORF_GeneList = list(set(Ingolia_uORF_GeneList).intersection(set(Cluster2_GeneList)))\n",
    "print(\"Cluster1 have uORF: n = \", str(len(Cluster1_uORF_GeneList)))\n",
    "print(\"Cluster2 have uORF: n = \", str(len(Cluster2_uORF_GeneList)))\n",
    "\n",
    "uAUG_U3 = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uAUG_Gene-U3.txt')\n",
    "uAUG_U3_AG = uAUG_U3[(uAUG_U3['U3'] == 'A')|(uAUG_U3['U3'] == 'G')]\n",
    "U3_AG_GeneList = uAUG_U3_AG['gene'].unique()\n",
    "Cluster1_uORF_U3_AG_GeneList = list(set(U3_AG_GeneList).intersection(set(Cluster1_GeneList)))\n",
    "Cluster2_uORF_U3_AG_GeneList = list(set(U3_AG_GeneList).intersection(set(Cluster2_GeneList)))\n",
    "print(\"Cluster1 have uORF U3=A/G: n = \", str(len(Cluster1_uORF_U3_AG_GeneList)))\n",
    "print(\"Cluster2 have uORF U3=A/G: n = \", str(len(Cluster2_uORF_U3_AG_GeneList)))\n",
    "\n",
    "uAUG_U3_CT = uAUG_U3[(uAUG_U3['U3'] == 'C')|(uAUG_U3['U3'] == 'T')]\n",
    "U3_CT_GeneList = uAUG_U3_CT['gene'].unique()\n",
    "Cluster1_uORF_U3_CT_GeneList = list(set(U3_CT_GeneList).intersection(set(Cluster1_GeneList)))\n",
    "Cluster2_uORF_U3_CT_GeneList = list(set(U3_CT_GeneList).intersection(set(Cluster2_GeneList)))\n",
    "print(\"Cluster1 have uORF U3=C/T: n = \", str(len(Cluster1_uORF_U3_CT_GeneList)))\n",
    "print(\"Cluster2 have uORF U3=C/T: n = \", str(len(Cluster2_uORF_U3_CT_GeneList)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_323398/1813364056.py:53: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  kmers_count_GCTGAA['Kmer'] = 'GCTGAA'\n",
      "/tmp/ipykernel_323398/1813364056.py:57: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  kmers_count_GAAGAA['Kmer'] = 'GAAGAA'\n",
      "/tmp/ipykernel_323398/1813364056.py:67: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  locals()['kmers_count_' + i]['Kmer'] = i\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>group</th>\n",
       "      <th>value</th>\n",
       "      <th>Kmer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YNL178W</td>\n",
       "      <td>Cluster2</td>\n",
       "      <td>0.007472</td>\n",
       "      <td>GCTGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YHL015W</td>\n",
       "      <td>Cluster2</td>\n",
       "      <td>0.001468</td>\n",
       "      <td>GCTGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YDR064W</td>\n",
       "      <td>Cluster2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCTGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YOR369C</td>\n",
       "      <td>Cluster2</td>\n",
       "      <td>0.001721</td>\n",
       "      <td>GCTGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YCR013C</td>\n",
       "      <td>Cluster2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCTGAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2330619</th>\n",
       "      <td>YEL039C</td>\n",
       "      <td>Cluster1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCGTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2330620</th>\n",
       "      <td>YPR001W</td>\n",
       "      <td>Cluster1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCGTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2330621</th>\n",
       "      <td>YGR016W</td>\n",
       "      <td>Cluster2</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCGTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2330622</th>\n",
       "      <td>YDR070C</td>\n",
       "      <td>Cluster1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCGTAG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2330623</th>\n",
       "      <td>YBR148W</td>\n",
       "      <td>Cluster1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>GCGTAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2330624 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            gene     group     value    Kmer\n",
       "0        YNL178W  Cluster2  0.007472  GCTGAA\n",
       "1        YHL015W  Cluster2  0.001468  GCTGAA\n",
       "2        YDR064W  Cluster2  0.000000  GCTGAA\n",
       "3        YOR369C  Cluster2  0.001721  GCTGAA\n",
       "4        YCR013C  Cluster2  0.000000  GCTGAA\n",
       "...          ...       ...       ...     ...\n",
       "2330619  YEL039C  Cluster1  0.000000  GCGTAG\n",
       "2330620  YPR001W  Cluster1  0.000000  GCGTAG\n",
       "2330621  YGR016W  Cluster2  0.000000  GCGTAG\n",
       "2330622  YDR070C  Cluster1  0.000000  GCGTAG\n",
       "2330623  YBR148W  Cluster1  0.000000  GCGTAG\n",
       "\n",
       "[2330624 rows x 4 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/8-CooperativeGene/Five_vs_Four/'\n",
    "Ttest_Cluster = pd.read_table((path2 + 'RibosomeDensity_foldchange_Pval_Cluster.txt'), usecols = ['gene','pval','group'])\n",
    "Ttest_Cluster = Ttest_Cluster[Ttest_Cluster['group'] != 'Not_significant']\n",
    "\n",
    "\n",
    "### 3-mers\n",
    "\n",
    "transcript_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/AllRegionSequence.txt', usecols = ['gene_id','seq'])\n",
    "transcript_annotation = transcript_annotation[transcript_annotation['gene_id'].isin(Ttest_Cluster['gene'].unique())].reset_index(drop = True)\n",
    "transcript_annotation['length'] = transcript_annotation['seq'].str.len()\n",
    "\n",
    "kmers_dict = {}\n",
    "\n",
    "for i in range(len(transcript_annotation['gene_id'])):\n",
    "    gene = transcript_annotation['gene_id'][i]\n",
    "    seq = transcript_annotation['seq'][i]\n",
    "    length = transcript_annotation['length'][i]\n",
    "    kmers_dict[gene] = []\n",
    "    for j in range(length-2):\n",
    "        kmers_dict[gene].append(seq[j:j+3])\n",
    "\n",
    "kmers_df = pd.DataFrame.from_dict(kmers_dict, orient='index').T\n",
    "\n",
    "kmers_count = pd.DataFrame()\n",
    "for i in Ttest_Cluster['gene'].unique():\n",
    "    kmers_count = kmers_count.append(kmers_df[i].value_counts())\n",
    "kmers_count = kmers_count[kmers_count.columns.drop(list(kmers_count.filter(regex='N')))]\n",
    "kmers_count = kmers_count.fillna(0).reset_index()\n",
    "\n",
    "kmers_count = pd.merge(kmers_count,Ttest_Cluster[['gene','group']],left_on = 'index',right_on = 'gene').drop('index', axis = 1)\n",
    "kmers_count = pd.merge(kmers_count,transcript_annotation[['gene_id','length']],left_on = 'gene',right_on = 'gene_id').drop('gene_id', axis = 1)\n",
    "\n",
    "kmers_list = list(kmers_count.T.reset_index()['index'].unique()[:-3])\n",
    "for i in kmers_list:\n",
    "    kmers_count[i] = kmers_count[i] / (kmers_count['length'] - 2)\n",
    "  \n",
    "kmers_count_AGA = kmers_count[['gene','group','AGA']]\n",
    "kmers_count_AGA.columns = ['gene','group','value']\n",
    "kmers_count_AGA['Kmer'] = 'AGA'\n",
    "\n",
    "kmers_count_AAG = kmers_count[['gene','group','AAG']]\n",
    "kmers_count_AAG.columns = ['gene','group','value']\n",
    "kmers_count_AAG['Kmer'] = 'AAG'\n",
    "\n",
    "kmers_count_rbind = pd.concat([kmers_count_AGA,kmers_count_AAG])\n",
    "\n",
    "kmers_list.remove('AGA')\n",
    "kmers_list.remove('AAG')\n",
    "\n",
    "for i in kmers_list:\n",
    "    locals()['kmers_count_' + i] = kmers_count[['gene','group',i]]\n",
    "    locals()['kmers_count_' + i].columns = ['gene','group','value']\n",
    "    locals()['kmers_count_' + i]['Kmer'] = i\n",
    "    kmers_count_rbind = pd.concat([kmers_count_rbind,locals()['kmers_count_' + i]]).reset_index(drop = True)\n",
    "kmers_count_rbind.to_csv((path2 + 'Cluster_3mer.txt'), sep = '\\t')\n",
    "kmers_count_rbind"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### p value and z score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_323398/3133680708.py:9: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Length_Feature['Feature'][i] = Length_Feature['Feature'][i] + '_Length'\n",
      "/tmp/ipykernel_323398/3133680708.py:14: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  GC_Feature['Feature'][i] = GC_Feature['Feature'][i] + '_GC'\n",
      "/tmp/ipykernel_323398/3133680708.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  PolyA_length_Feature['Feature'] = 'PolyA_length'\n",
      "/tmp/ipykernel_323398/3133680708.py:31: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  nonA_Feature['Feature'] = 'PolyA_nonA'\n"
     ]
    }
   ],
   "source": [
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/8-CooperativeGene/Five_vs_Four/' ## The pathway to output results\n",
    "\n",
    "Length_and_GC = pd.read_table(path2 + 'Cluster_Length_and_GC.txt')\n",
    "Length_and_GC = Length_and_GC[Length_and_GC['group'] != 'Not_significant'].reset_index(drop = True)\n",
    "\n",
    "Length_Feature = Length_and_GC[['gene','group','length','Region']]\n",
    "Length_Feature.columns = ['gene','group','value','Feature']\n",
    "for i in range(len(Length_Feature['gene'])):\n",
    "    Length_Feature['Feature'][i] = Length_Feature['Feature'][i] + '_Length'\n",
    "\n",
    "GC_Feature = Length_and_GC[['gene','group','GC','Region']]\n",
    "GC_Feature.columns = ['gene','group','value','Feature']\n",
    "for i in range(len(GC_Feature['gene'])):\n",
    "    GC_Feature['Feature'][i] = GC_Feature['Feature'][i] + '_GC'\n",
    "\n",
    "KozakScore = pd.read_table(path2 + 'Cluster_KozakScore_6+3.txt')\n",
    "KozakScore = KozakScore[KozakScore['group'] != 'Not_significant'].reset_index(drop = True)\n",
    "KozakScore = KozakScore[['gene','group','KozakScore']]\n",
    "KozakScore.columns = ['gene','group','value']\n",
    "KozakScore['Feature'] = 'KozakScore'\n",
    "\n",
    "PolyA_length_nonA = pd.read_table(path2 + 'Cluster_PolyA.txt')\n",
    "PolyA_length_nonA = PolyA_length_nonA[PolyA_length_nonA['group'] != 'Not_significant'].reset_index(drop = True)\n",
    "\n",
    "PolyA_length_Feature = PolyA_length_nonA[['gene','group','poly_length_average']]\n",
    "PolyA_length_Feature.columns = ['gene','group','value']\n",
    "PolyA_length_Feature['Feature'] = 'PolyA_length'\n",
    "\n",
    "nonA_Feature = PolyA_length_nonA[['gene','group','nonA_freq_average']]\n",
    "nonA_Feature.columns = ['gene','group','value']\n",
    "nonA_Feature['Feature'] = 'PolyA_nonA'\n",
    "\n",
    "transcript_length = pd.read_table((path2 + 'Cluster_transcript_length.txt'))\n",
    "transcript_length = transcript_length[['gene','group','transcript_length']]\n",
    "transcript_length.columns = ['gene','group','value']\n",
    "transcript_length['Feature'] = 'transcript_length'\n",
    "transcript_length = transcript_length[transcript_length['group'] != 'Not_significant'].reset_index(drop = True)\n",
    "\n",
    "kmers_count_rbind = pd.read_table(path2 + 'Cluster_3mer.txt', usecols = ['gene','group','value','Kmer'])\n",
    "kmers_count_rbind.columns = ['gene','group','value','Feature']\n",
    "\n",
    "\n",
    "Feature_rbind = pd.concat([Length_Feature,GC_Feature,transcript_length,KozakScore,PolyA_length_Feature,nonA_Feature,kmers_count_rbind]).reset_index(drop = True)\n",
    "Feature_rbind.to_csv((path2 + 'Feature_rbind_3kmer.txt'),sep = '\\t')\n",
    "\n",
    "### perform wilcox.test in R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. eIF1/eIF1A vs. eIF4A/eIF4B"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cooperative Gene identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Dell/Dell15/zhanggy/bin/anaconda3/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3757: RuntimeWarning: Degrees of freedom <= 0 for slice\n",
      "  return _methods._var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/Dell/Dell15/zhanggy/bin/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:256: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ret = ret.dtype.type(ret / rcount)\n"
     ]
    }
   ],
   "source": [
    "path1 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/' ## The pathway of -RibosomeDensity.txt\n",
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/8-CooperativeGene/Two_vs_Two/' ## The pathway to output results\n",
    "\n",
    "group1_List = ['eIF1','eIF1A']\n",
    "\n",
    "for i in group1_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "group1 = pd.concat([locals()['eIF1'],locals()['eIF1A']])\n",
    "group1['type'] = 'group1'\n",
    "\n",
    "group2_List = ['eIF4A','eIF4B']\n",
    "\n",
    "for i in group2_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "group2 = pd.concat([locals()['eIF4A'],locals()['eIF4B']])\n",
    "group2['type'] = 'group2'\n",
    "\n",
    "total_cluster = pd.concat([group1,group2])\n",
    "gene1 = total_cluster['gene'].unique()\n",
    "\n",
    "### Only consider genes with 5'UTR and 3'UTR annotations and 5'UTR length > 1\n",
    "UTR_gene = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/saccharomyces_cerevisiae_information/Saccharomyces_cerevisiae_UTRInformation.txt')\n",
    "UTR3_gene = UTR_gene[UTR_gene['type'] == 'three_prime_utr']['gene_id'].unique()\n",
    "UTR_5 = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/APOBEC1_data/preview/1-GeneIdentification-Overview/5UTR_seq.txt')\n",
    "UTR_5['length'] = UTR_5['stop'] - UTR_5['start'] + 1\n",
    "UTR_5 = UTR_5[UTR_5['length'] != 1]\n",
    "UTR5_gene = UTR_5['gene_id'].unique()\n",
    "gene2 = list(set(UTR5_gene).intersection(set(UTR3_gene)))\n",
    "gene = list(set(gene1).intersection(set(gene2)))\n",
    "\n",
    "### Genes with significantly different Ribosome Density between group1 and group2 were calculated using t.test\n",
    "\n",
    "gene_pval = open((path2 + 'Ttest_Cluster.txt'),'w')\n",
    "gene_pval.write('gene\\ttstat\\tpval\\n')\n",
    "\n",
    "for i in gene:\n",
    "    gene_subset = total_cluster[total_cluster['gene'] == i]\n",
    "    gene_subset_group1 = gene_subset[gene_subset['type'] == 'group1']['Ribosome_Density'].tolist()\n",
    "    gene_subset_group2 = gene_subset[gene_subset['type'] == 'group2']['Ribosome_Density'].tolist()\n",
    "    tstat, pval = stats.ttest_ind(a = gene_subset_group1, b = gene_subset_group2, alternative = \"two-sided\")\n",
    "    gene_pval.write(i + '\\t' + str(tstat) + '\\t' + str(pval) + '\\n')\n",
    "gene_pval.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_69781/2598026937.py:34: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ttest_Cluster['group'][i] = 'Not_significant'\n",
      "/tmp/ipykernel_69781/2598026937.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ttest_Cluster['group'][i] = 'Cluster2'\n",
      "/tmp/ipykernel_69781/2598026937.py:37: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ttest_Cluster['group'][i] = 'Cluster1'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster1:  178\n",
      "Cluster2:  50\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>group1_RibosomeDensity_mean</th>\n",
       "      <th>group2_RibosomeDensity_mean</th>\n",
       "      <th>Ribosome_Density_foldchange</th>\n",
       "      <th>pval</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YGR192C</td>\n",
       "      <td>0.849407</td>\n",
       "      <td>0.932894</td>\n",
       "      <td>0.910508</td>\n",
       "      <td>0.390342</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YLR044C</td>\n",
       "      <td>1.245066</td>\n",
       "      <td>1.380499</td>\n",
       "      <td>0.901896</td>\n",
       "      <td>0.522983</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YKL060C</td>\n",
       "      <td>1.546081</td>\n",
       "      <td>1.753360</td>\n",
       "      <td>0.881782</td>\n",
       "      <td>0.576034</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YLR075W</td>\n",
       "      <td>0.944501</td>\n",
       "      <td>0.873248</td>\n",
       "      <td>1.081595</td>\n",
       "      <td>0.818001</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YHR174W</td>\n",
       "      <td>2.001487</td>\n",
       "      <td>1.734344</td>\n",
       "      <td>1.154031</td>\n",
       "      <td>0.367091</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4273</th>\n",
       "      <td>YPR145C-A</td>\n",
       "      <td>0.847491</td>\n",
       "      <td>1.271053</td>\n",
       "      <td>0.666763</td>\n",
       "      <td>0.225026</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4274</th>\n",
       "      <td>Q0275</td>\n",
       "      <td>13.277022</td>\n",
       "      <td>15.911548</td>\n",
       "      <td>0.834427</td>\n",
       "      <td>0.879634</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4275</th>\n",
       "      <td>YDR034W-B</td>\n",
       "      <td>4.774208</td>\n",
       "      <td>3.202471</td>\n",
       "      <td>1.490789</td>\n",
       "      <td>0.692750</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4276</th>\n",
       "      <td>Q0085</td>\n",
       "      <td>4.198422</td>\n",
       "      <td>6.230817</td>\n",
       "      <td>0.673816</td>\n",
       "      <td>0.749564</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4277</th>\n",
       "      <td>YGL088W</td>\n",
       "      <td>3.941973</td>\n",
       "      <td>10.411717</td>\n",
       "      <td>0.378609</td>\n",
       "      <td>0.019875</td>\n",
       "      <td>Cluster2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4278 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene  group1_RibosomeDensity_mean  group2_RibosomeDensity_mean  \\\n",
       "0       YGR192C                     0.849407                     0.932894   \n",
       "1       YLR044C                     1.245066                     1.380499   \n",
       "2       YKL060C                     1.546081                     1.753360   \n",
       "3       YLR075W                     0.944501                     0.873248   \n",
       "4       YHR174W                     2.001487                     1.734344   \n",
       "...         ...                          ...                          ...   \n",
       "4273  YPR145C-A                     0.847491                     1.271053   \n",
       "4274      Q0275                    13.277022                    15.911548   \n",
       "4275  YDR034W-B                     4.774208                     3.202471   \n",
       "4276      Q0085                     4.198422                     6.230817   \n",
       "4277    YGL088W                     3.941973                    10.411717   \n",
       "\n",
       "      Ribosome_Density_foldchange      pval            group  \n",
       "0                        0.910508  0.390342  Not_significant  \n",
       "1                        0.901896  0.522983  Not_significant  \n",
       "2                        0.881782  0.576034  Not_significant  \n",
       "3                        1.081595  0.818001  Not_significant  \n",
       "4                        1.154031  0.367091  Not_significant  \n",
       "...                           ...       ...              ...  \n",
       "4273                     0.666763  0.225026  Not_significant  \n",
       "4274                     0.834427  0.879634  Not_significant  \n",
       "4275                     1.490789  0.692750  Not_significant  \n",
       "4276                     0.673816  0.749564  Not_significant  \n",
       "4277                     0.378609  0.019875         Cluster2  \n",
       "\n",
       "[4278 rows x 6 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Calculate the average Ribosome Density of genes in group1\n",
    "group1_List = ['eIF1','eIF1A']\n",
    "\n",
    "for i in group1_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "    locals()[i].columns = ['gene',i]\n",
    "\n",
    "group1 = pd.merge(locals()['eIF1'],locals()['eIF1A'],on = 'gene')\n",
    "group1['group1_RibosomeDensity_mean'] = (group1['eIF1'] + group1['eIF1A']) / 2\n",
    "\n",
    "### Calculate the average Ribosome Density of genes in group2\n",
    "group2_List = ['eIF4A','eIF4B']\n",
    "\n",
    "for i in group2_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "    locals()[i].columns = ['gene',i]\n",
    "group2 = pd.merge(locals()['eIF4A'],locals()['eIF4B'],on = 'gene')\n",
    "group2['group2_RibosomeDensity_mean'] = (group2['eIF4A'] + group2['eIF4B']) / 2\n",
    "\n",
    "### Measures the change in Ribosome Density of genes in group1 relative to group2\n",
    "Groups = pd.merge(group1,group2,on = 'gene')\n",
    "Groups['Ribosome_Density_foldchange'] = Groups['group1_RibosomeDensity_mean'] / Groups['group2_RibosomeDensity_mean']\n",
    "Groups.to_csv((path2 + 'Cluster_Ribosome_Density_foldchange.txt'),sep = '\\t')\n",
    "\n",
    "Ttest_Cluster = pd.read_table((path2 + 'Ttest_Cluster.txt'),sep = '\\t')\n",
    "Ttest_Cluster = pd.merge(Groups, Ttest_Cluster, on = 'gene')\n",
    "Ttest_Cluster = Ttest_Cluster[['gene','group1_RibosomeDensity_mean','group2_RibosomeDensity_mean','Ribosome_Density_foldchange','pval']]\n",
    "Ttest_Cluster = Ttest_Cluster[~Ttest_Cluster.isin([np.nan, np.inf, -np.inf]).any(1)].reset_index(drop = True)\n",
    "\n",
    "Ttest_Cluster['group'] = ''\n",
    "\n",
    "for i in range(len(Ttest_Cluster['gene'])):\n",
    "    if Ttest_Cluster['pval'][i] > 0.05:\n",
    "        Ttest_Cluster['group'][i] = 'Not_significant'\n",
    "    elif Ttest_Cluster['pval'][i] <= 0.05:\n",
    "        if Ttest_Cluster['Ribosome_Density_foldchange'][i] > 1:\n",
    "            Ttest_Cluster['group'][i] = 'Cluster1'\n",
    "        elif Ttest_Cluster['Ribosome_Density_foldchange'][i] < 1:\n",
    "            Ttest_Cluster['group'][i] = 'Cluster2'\n",
    "\n",
    "Ttest_Cluster_selected = Ttest_Cluster[Ttest_Cluster['group'] != 'Not_significant'].reset_index(drop = True)\n",
    "print(\"Cluster1: \", str(len(Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster1']['gene'].unique())))\n",
    "print(\"Cluster2: \", str(len(Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster2']['gene'].unique())))\n",
    "\n",
    "Ttest_Cluster.to_csv((path2 + 'RibosomeDensity_foldchange_Pval_Cluster.txt'), sep = '\\t')\n",
    "Ttest_Cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Features of genes in different cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ingolia uORF: n =  597\n",
      "Cluster1: n =  178\n",
      "Cluster2: n =  50\n",
      "Cluster1 have uORF: n =  16\n",
      "Cluster2 have uORF: n =  9\n",
      "Cluster1 have uORF U3=A/G: n =  7\n",
      "Cluster2 have uORF U3=A/G: n =  1\n",
      "Cluster1 have uORF U3=C/T: n =  9\n",
      "Cluster2 have uORF U3=C/T: n =  8\n"
     ]
    }
   ],
   "source": [
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/8-CooperativeGene/Two_vs_Two/' ## The pathway to output results\n",
    "Ttest_Cluster = pd.read_table((path2 + 'RibosomeDensity_foldchange_Pval_Cluster.txt'), usecols = ['gene','Ribosome_Density_foldchange','pval','group'])\n",
    "\n",
    "### Ribosome Density: Mutant and HO\n",
    "Mutant_RibosomeDensity = pd.read_table((path2 + 'Cluster_Ribosome_Density_foldchange.txt'), usecols = ['gene','group1_RibosomeDensity_mean','group2_RibosomeDensity_mean'])\n",
    "Mutant_Cluster1_RibosomeDensity = Mutant_RibosomeDensity[['gene','group2_RibosomeDensity_mean']]\n",
    "Mutant_Cluster1_RibosomeDensity['sample'] = 'Cluster1_Factor_mean'\n",
    "Mutant_Cluster1_RibosomeDensity.columns = ['gene','RibosomeDensity','sample']\n",
    "\n",
    "Mutant_Cluster2_RibosomeDensity = Mutant_RibosomeDensity[['gene','group1_RibosomeDensity_mean']]\n",
    "Mutant_Cluster2_RibosomeDensity['sample'] = 'Cluster2_Factor_mean'\n",
    "Mutant_Cluster2_RibosomeDensity.columns = ['gene','RibosomeDensity','sample']\n",
    "\n",
    "HO_RibosomeDensity = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/HO/HO-RibosomeDensity.txt',\\\n",
    "                                   usecols = ['gene','Ribosome_Density','sample'])\n",
    "HO_RibosomeDensity.columns = ['gene','RibosomeDensity','sample']\n",
    "\n",
    "RibosomeDensity = pd.concat([HO_RibosomeDensity,Mutant_Cluster1_RibosomeDensity,Mutant_Cluster2_RibosomeDensity]).reset_index(drop = True)\n",
    "RibosomeDensity = pd.merge(RibosomeDensity,Ttest_Cluster[['gene','group']], on = 'gene')\n",
    "RibosomeDensity = RibosomeDensity[RibosomeDensity['group'] != 'Not_significant'].reset_index(drop = True)\n",
    "RibosomeDensity.to_csv((path2 + 'Cluster_HOandMutant_RibosomeDensity.txt'), sep = '\\t')\n",
    "\n",
    "### Length and GC%\n",
    "\n",
    "UTR_5_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/5UTR_seq.txt', usecols = ['gene_id','seq'])\n",
    "UTR_5_annotation['Region'] = \"5'UTR\"\n",
    "\n",
    "CDS_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/CDS_seq.txt', usecols = ['gene_id','seq'])\n",
    "CDS_annotation['Region'] = \"CDS\"\n",
    "\n",
    "UTR_3_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/3UTR_seq.txt', usecols = ['gene_id','seq'])\n",
    "UTR_3_annotation['Region'] = \"3'UTR\"\n",
    "\n",
    "AllRegion_annotation = pd.concat([UTR_3_annotation,UTR_5_annotation,CDS_annotation]).reset_index(drop = True)\n",
    "\n",
    "AllRegion_annotation['length'] = ''\n",
    "AllRegion_annotation['GC'] = ''\n",
    "for i in range(len(AllRegion_annotation['gene_id'])):\n",
    "    AllRegion_annotation['length'][i] = len(AllRegion_annotation['seq'][i])\n",
    "    AllRegion_annotation['GC'][i] = (AllRegion_annotation['seq'][i].count('G') + AllRegion_annotation['seq'][i].count('C')) / len(AllRegion_annotation['seq'][i])\n",
    "AllRegion_annotation['GC'] = AllRegion_annotation['GC']*100\n",
    "AllRegion_annotation = pd.merge(AllRegion_annotation,Ttest_Cluster,left_on = 'gene_id',right_on = 'gene').drop(['gene_id'],axis = 1)\n",
    "\n",
    "AllRegion_annotation[['gene','Region','length','GC','group']].to_csv((path2 + 'Cluster_Length_and_GC.txt'), sep = '\\t')\n",
    "AllRegion_annotation[['gene','Region','length','GC','group']]\n",
    "\n",
    "### Kozak Score\n",
    "\n",
    "KozakScore = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/APOBEC1_data/preview/3-NormalizeFeature/Kozak/initiation_score_6+3.txt',usecols = ['gene','score_xnt'])\n",
    "KozakScore.columns = ['gene','KozakScore']\n",
    "KozakScore = pd.merge(Ttest_Cluster,KozakScore,on = 'gene')\n",
    "KozakScore.to_csv((path2 + 'Cluster_KozakScore_6+3.txt'), sep = '\\t')\n",
    "\n",
    "### PolyA Length and nonA\n",
    "\n",
    "PolyA_length_nonA = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H_polyLength_nonAcontent.txt')\n",
    "PolyA_length_nonA = pd.merge(Ttest_Cluster,PolyA_length_nonA,left_on = 'gene',right_on = 'gene_id').drop(['gene_id','Unnamed: 0'], axis = 1)\n",
    "PolyA_length_nonA.to_csv((path2 + 'Cluster_PolyA.txt'), sep = '\\t')\n",
    "\n",
    "### transcript length\n",
    "\n",
    "transcript_length = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/AllRegionSequence.txt', usecols = ['gene_id','seq'])\n",
    "transcript_length['transcript_length'] = transcript_length['seq'].str.len()\n",
    "transcript_length = pd.merge(Ttest_Cluster,transcript_length[['gene_id','transcript_length']],left_on = 'gene',right_on = 'gene_id').drop(['gene_id'],axis = 1)\n",
    "transcript_length.to_csv((path2 + 'Cluster_transcript_length.txt'), sep = '\\t')\n",
    "\n",
    "### uORF enrichment\n",
    "\n",
    "Ttest_Cluster_selected = Ttest_Cluster[Ttest_Cluster['group'] != 'Not_significant'].reset_index()\n",
    "Ttest_Cluster_selected = Ttest_Cluster_selected[['gene','group']]\n",
    "\n",
    "Ingolia_uORF_Gene = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uORF_Gene_FristPos_2009-ingolia.txt')\n",
    "Ingolia_uORF_GeneList = Ingolia_uORF_Gene['Gene'].unique()\n",
    "Cluster1_GeneList = Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster1']['gene'].unique()\n",
    "Cluster2_GeneList = Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster2']['gene'].unique()\n",
    "print(\"Ingolia uORF: n = \", str(len(Ingolia_uORF_GeneList)))\n",
    "print(\"Cluster1: n = \", str(len(Cluster1_GeneList)))\n",
    "print(\"Cluster2: n = \", str(len(Cluster2_GeneList)))\n",
    "Cluster1_uORF_GeneList = list(set(Ingolia_uORF_GeneList).intersection(set(Cluster1_GeneList)))\n",
    "Cluster2_uORF_GeneList = list(set(Ingolia_uORF_GeneList).intersection(set(Cluster2_GeneList)))\n",
    "print(\"Cluster1 have uORF: n = \", str(len(Cluster1_uORF_GeneList)))\n",
    "print(\"Cluster2 have uORF: n = \", str(len(Cluster2_uORF_GeneList)))\n",
    "\n",
    "uAUG_U3 = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uAUG_Gene-U3.txt')\n",
    "uAUG_U3_AG = uAUG_U3[(uAUG_U3['U3'] == 'A')|(uAUG_U3['U3'] == 'G')]\n",
    "U3_AG_GeneList = uAUG_U3_AG['gene'].unique()\n",
    "Cluster1_uORF_U3_AG_GeneList = list(set(U3_AG_GeneList).intersection(set(Cluster1_GeneList)))\n",
    "Cluster2_uORF_U3_AG_GeneList = list(set(U3_AG_GeneList).intersection(set(Cluster2_GeneList)))\n",
    "print(\"Cluster1 have uORF U3=A/G: n = \", str(len(Cluster1_uORF_U3_AG_GeneList)))\n",
    "print(\"Cluster2 have uORF U3=A/G: n = \", str(len(Cluster2_uORF_U3_AG_GeneList)))\n",
    "\n",
    "uAUG_U3_CT = uAUG_U3[(uAUG_U3['U3'] == 'C')|(uAUG_U3['U3'] == 'T')]\n",
    "U3_CT_GeneList = uAUG_U3_CT['gene'].unique()\n",
    "Cluster1_uORF_U3_CT_GeneList = list(set(U3_CT_GeneList).intersection(set(Cluster1_GeneList)))\n",
    "Cluster2_uORF_U3_CT_GeneList = list(set(U3_CT_GeneList).intersection(set(Cluster2_GeneList)))\n",
    "print(\"Cluster1 have uORF U3=C/T: n = \", str(len(Cluster1_uORF_U3_CT_GeneList)))\n",
    "print(\"Cluster2 have uORF U3=C/T: n = \", str(len(Cluster2_uORF_U3_CT_GeneList)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c. eIF4G1/PAB1 vs. eIF4E/EAP1/CAF20"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Cooperative Gene identification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Dell/Dell15/zhanggy/bin/anaconda3/lib/python3.9/site-packages/numpy/core/fromnumeric.py:3757: RuntimeWarning: Degrees of freedom <= 0 for slice\n",
      "  return _methods._var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,\n",
      "/Dell/Dell15/zhanggy/bin/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:256: RuntimeWarning: invalid value encountered in double_scalars\n",
      "  ret = ret.dtype.type(ret / rcount)\n"
     ]
    }
   ],
   "source": [
    "path1 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/' ## The pathway of -RibosomeDensity.txt\n",
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/8-CooperativeGene/Two_vs_Three/' ## The pathway to output results\n",
    "\n",
    "group1_List = ['eIF4G1','PAB1']\n",
    "\n",
    "for i in group1_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "group1 = pd.concat([locals()['eIF4G1'],locals()['PAB1']])\n",
    "group1['type'] = 'group1'\n",
    "\n",
    "group2_List = ['eIF4E','EAP1','CAF20']\n",
    "\n",
    "for i in group2_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "group2 = pd.concat([locals()['eIF4E'],locals()['EAP1'],locals()['CAF20']])\n",
    "group2['type'] = 'group2'\n",
    "\n",
    "total_cluster = pd.concat([group1,group2])\n",
    "gene1 = total_cluster['gene'].unique()\n",
    "\n",
    "### Only consider genes with 5'UTR and 3'UTR annotations and 5'UTR length > 1\n",
    "UTR_gene = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/saccharomyces_cerevisiae_information/Saccharomyces_cerevisiae_UTRInformation.txt')\n",
    "UTR3_gene = UTR_gene[UTR_gene['type'] == 'three_prime_utr']['gene_id'].unique()\n",
    "UTR_5 = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/APOBEC1_data/preview/1-GeneIdentification-Overview/5UTR_seq.txt')\n",
    "UTR_5['length'] = UTR_5['stop'] - UTR_5['start'] + 1\n",
    "UTR_5 = UTR_5[UTR_5['length'] != 1]\n",
    "UTR5_gene = UTR_5['gene_id'].unique()\n",
    "gene2 = list(set(UTR5_gene).intersection(set(UTR3_gene)))\n",
    "gene = list(set(gene1).intersection(set(gene2)))\n",
    "\n",
    "### Genes with significantly different Ribosome Density between group1 and group2 were calculated using t.test\n",
    "\n",
    "gene_pval = open((path2 + 'Ttest_Cluster.txt'),'w')\n",
    "gene_pval.write('gene\\ttstat\\tpval\\n')\n",
    "\n",
    "for i in gene:\n",
    "    gene_subset = total_cluster[total_cluster['gene'] == i]\n",
    "    gene_subset_group1 = gene_subset[gene_subset['type'] == 'group1']['Ribosome_Density'].tolist()\n",
    "    gene_subset_group2 = gene_subset[gene_subset['type'] == 'group2']['Ribosome_Density'].tolist()\n",
    "    tstat, pval = stats.ttest_ind(a = gene_subset_group1, b = gene_subset_group2, alternative = \"two-sided\")\n",
    "    gene_pval.write(i + '\\t' + str(tstat) + '\\t' + str(pval) + '\\n')\n",
    "gene_pval.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_69781/552633228.py:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ttest_Cluster['group'][i] = 'Not_significant'\n",
      "/tmp/ipykernel_69781/552633228.py:41: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ttest_Cluster['group'][i] = 'Cluster2'\n",
      "/tmp/ipykernel_69781/552633228.py:39: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  Ttest_Cluster['group'][i] = 'Cluster1'\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cluster1:  272\n",
      "Cluster2:  97\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>group1_RibosomeDensity_mean</th>\n",
       "      <th>group2_RibosomeDensity_mean</th>\n",
       "      <th>Ribosome_Density_foldchange</th>\n",
       "      <th>pval</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YGR192C</td>\n",
       "      <td>0.719680</td>\n",
       "      <td>0.838595</td>\n",
       "      <td>0.858197</td>\n",
       "      <td>0.466627</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YHR174W</td>\n",
       "      <td>1.124567</td>\n",
       "      <td>1.678980</td>\n",
       "      <td>0.669792</td>\n",
       "      <td>0.043697</td>\n",
       "      <td>Cluster2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YKL060C</td>\n",
       "      <td>1.461935</td>\n",
       "      <td>1.855897</td>\n",
       "      <td>0.787725</td>\n",
       "      <td>0.164192</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YLR044C</td>\n",
       "      <td>1.274541</td>\n",
       "      <td>1.542991</td>\n",
       "      <td>0.826020</td>\n",
       "      <td>0.500095</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YLR075W</td>\n",
       "      <td>0.717548</td>\n",
       "      <td>0.954012</td>\n",
       "      <td>0.752137</td>\n",
       "      <td>0.250734</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4309</th>\n",
       "      <td>YGR249W</td>\n",
       "      <td>2.189377</td>\n",
       "      <td>2.133946</td>\n",
       "      <td>1.025976</td>\n",
       "      <td>0.965942</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4310</th>\n",
       "      <td>YPR145C-A</td>\n",
       "      <td>3.786854</td>\n",
       "      <td>0.956003</td>\n",
       "      <td>3.961132</td>\n",
       "      <td>0.309235</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4311</th>\n",
       "      <td>YGL032C</td>\n",
       "      <td>4.773787</td>\n",
       "      <td>1.756722</td>\n",
       "      <td>2.717440</td>\n",
       "      <td>0.113953</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4312</th>\n",
       "      <td>Q0085</td>\n",
       "      <td>6.240546</td>\n",
       "      <td>5.654282</td>\n",
       "      <td>1.103685</td>\n",
       "      <td>0.828873</td>\n",
       "      <td>Not_significant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4313</th>\n",
       "      <td>YDR034W-B</td>\n",
       "      <td>13.255942</td>\n",
       "      <td>1.625854</td>\n",
       "      <td>8.153218</td>\n",
       "      <td>0.000240</td>\n",
       "      <td>Cluster1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4314 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene  group1_RibosomeDensity_mean  group2_RibosomeDensity_mean  \\\n",
       "0       YGR192C                     0.719680                     0.838595   \n",
       "1       YHR174W                     1.124567                     1.678980   \n",
       "2       YKL060C                     1.461935                     1.855897   \n",
       "3       YLR044C                     1.274541                     1.542991   \n",
       "4       YLR075W                     0.717548                     0.954012   \n",
       "...         ...                          ...                          ...   \n",
       "4309    YGR249W                     2.189377                     2.133946   \n",
       "4310  YPR145C-A                     3.786854                     0.956003   \n",
       "4311    YGL032C                     4.773787                     1.756722   \n",
       "4312      Q0085                     6.240546                     5.654282   \n",
       "4313  YDR034W-B                    13.255942                     1.625854   \n",
       "\n",
       "      Ribosome_Density_foldchange      pval            group  \n",
       "0                        0.858197  0.466627  Not_significant  \n",
       "1                        0.669792  0.043697         Cluster2  \n",
       "2                        0.787725  0.164192  Not_significant  \n",
       "3                        0.826020  0.500095  Not_significant  \n",
       "4                        0.752137  0.250734  Not_significant  \n",
       "...                           ...       ...              ...  \n",
       "4309                     1.025976  0.965942  Not_significant  \n",
       "4310                     3.961132  0.309235  Not_significant  \n",
       "4311                     2.717440  0.113953  Not_significant  \n",
       "4312                     1.103685  0.828873  Not_significant  \n",
       "4313                     8.153218  0.000240         Cluster1  \n",
       "\n",
       "[4314 rows x 6 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Calculate the average Ribosome Density of genes in group1\n",
    "group1_List = ['eIF4G1','PAB1']\n",
    "\n",
    "for i in group1_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "    locals()[i].columns = ['gene',i]\n",
    "\n",
    "group1 = pd.merge(locals()['eIF4G1'],locals()['PAB1'],on = 'gene')\n",
    "group1['group1_RibosomeDensity_mean'] = (group1['eIF4G1'] + group1['PAB1']) / 2\n",
    "\n",
    "### Calculate the average Ribosome Density of genes in group2\n",
    "group2_List = ['eIF4E','EAP1','CAF20']\n",
    "\n",
    "for i in group2_List:\n",
    "    locals()[i] = pd.read_table((path1 + i + '/' + i + '-RibosomeDensity.txt'), usecols = ['gene','Ribosome_Density'])\n",
    "    locals()[i].columns = ['gene',i]\n",
    "group2 = pd.merge(locals()['eIF4E'],locals()['EAP1'],on = 'gene')\n",
    "group2 = pd.merge(group2,locals()['CAF20'],on = 'gene')\n",
    "\n",
    "group2['group2_RibosomeDensity_mean'] = (group2['eIF4E'] + group2['EAP1'] + group2['CAF20']) / 3\n",
    "\n",
    "### Measures the change in Ribosome Density of genes in group1 relative to group2\n",
    "Groups = pd.merge(group1,group2,on = 'gene')\n",
    "Groups['Ribosome_Density_foldchange'] = Groups['group1_RibosomeDensity_mean'] / Groups['group2_RibosomeDensity_mean']\n",
    "Groups.to_csv((path2 + 'Cluster_Ribosome_Density_foldchange.txt'),sep = '\\t')\n",
    "\n",
    "Ttest_Cluster = pd.read_table((path2 + 'Ttest_Cluster.txt'),sep = '\\t')\n",
    "Ttest_Cluster = pd.merge(Groups, Ttest_Cluster, on = 'gene')\n",
    "Ttest_Cluster = Ttest_Cluster[['gene','group1_RibosomeDensity_mean','group2_RibosomeDensity_mean','Ribosome_Density_foldchange','pval']]\n",
    "Ttest_Cluster = Ttest_Cluster[~Ttest_Cluster.isin([np.nan, np.inf, -np.inf]).any(1)].reset_index(drop = True)\n",
    "\n",
    "Ttest_Cluster['group'] = ''\n",
    "\n",
    "for i in range(len(Ttest_Cluster['gene'])):\n",
    "    if Ttest_Cluster['pval'][i] > 0.05:\n",
    "        Ttest_Cluster['group'][i] = 'Not_significant'\n",
    "    elif Ttest_Cluster['pval'][i] <= 0.05:\n",
    "        if Ttest_Cluster['Ribosome_Density_foldchange'][i] > 1:\n",
    "            Ttest_Cluster['group'][i] = 'Cluster1'\n",
    "        elif Ttest_Cluster['Ribosome_Density_foldchange'][i] < 1:\n",
    "            Ttest_Cluster['group'][i] = 'Cluster2'\n",
    "\n",
    "Ttest_Cluster_selected = Ttest_Cluster[Ttest_Cluster['group'] != 'Not_significant'].reset_index(drop = True)\n",
    "print(\"Cluster1: \", str(len(Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster1']['gene'].unique())))\n",
    "print(\"Cluster2: \", str(len(Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster2']['gene'].unique())))\n",
    "\n",
    "Ttest_Cluster.to_csv((path2 + 'RibosomeDensity_foldchange_Pval_Cluster.txt'), sep = '\\t')\n",
    "Ttest_Cluster"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Features of genes in different cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Ingolia uORF: n =  597\n",
      "Cluster1: n =  272\n",
      "Cluster2: n =  97\n",
      "Cluster1 have uORF: n =  35\n",
      "Cluster2 have uORF: n =  8\n",
      "Cluster1 have uORF U3=A/G: n =  9\n",
      "Cluster2 have uORF U3=A/G: n =  3\n",
      "Cluster1 have uORF U3=C/T: n =  26\n",
      "Cluster2 have uORF U3=C/T: n =  5\n"
     ]
    }
   ],
   "source": [
    "Ttest_Cluster = pd.read_table((path2 + 'RibosomeDensity_foldchange_Pval_Cluster.txt'), usecols = ['gene','Ribosome_Density_foldchange','pval','group'])\n",
    "\n",
    "### Ribosome Density: Mutant and HO\n",
    "Mutant_RibosomeDensity = pd.read_table((path2 + 'Cluster_Ribosome_Density_foldchange.txt'), usecols = ['gene','group1_RibosomeDensity_mean','group2_RibosomeDensity_mean'])\n",
    "Mutant_Cluster1_RibosomeDensity = Mutant_RibosomeDensity[['gene','group2_RibosomeDensity_mean']]\n",
    "Mutant_Cluster1_RibosomeDensity['sample'] = 'Cluster1_Factor_mean'\n",
    "Mutant_Cluster1_RibosomeDensity.columns = ['gene','RibosomeDensity','sample']\n",
    "\n",
    "Mutant_Cluster2_RibosomeDensity = Mutant_RibosomeDensity[['gene','group1_RibosomeDensity_mean']]\n",
    "Mutant_Cluster2_RibosomeDensity['sample'] = 'Cluster2_Factor_mean'\n",
    "Mutant_Cluster2_RibosomeDensity.columns = ['gene','RibosomeDensity','sample']\n",
    "\n",
    "HO_RibosomeDensity = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/HO/HO-RibosomeDensity.txt',\\\n",
    "                                   usecols = ['gene','Ribosome_Density','sample'])\n",
    "HO_RibosomeDensity.columns = ['gene','RibosomeDensity','sample']\n",
    "\n",
    "RibosomeDensity = pd.concat([HO_RibosomeDensity,Mutant_Cluster1_RibosomeDensity,Mutant_Cluster2_RibosomeDensity]).reset_index(drop = True)\n",
    "RibosomeDensity = pd.merge(RibosomeDensity,Ttest_Cluster[['gene','group']], on = 'gene')\n",
    "RibosomeDensity = RibosomeDensity[RibosomeDensity['group'] != 'Not_significant'].reset_index(drop = True)\n",
    "RibosomeDensity.to_csv((path2 + 'Cluster_HOandMutant_RibosomeDensity.txt'), sep = '\\t')\n",
    "\n",
    "### Length and GC%\n",
    "\n",
    "UTR_5_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/5UTR_seq.txt', usecols = ['gene_id','seq'])\n",
    "UTR_5_annotation['Region'] = \"5'UTR\"\n",
    "\n",
    "CDS_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/CDS_seq.txt', usecols = ['gene_id','seq'])\n",
    "CDS_annotation['Region'] = \"CDS\"\n",
    "\n",
    "UTR_3_annotation = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/3UTR_seq.txt', usecols = ['gene_id','seq'])\n",
    "UTR_3_annotation['Region'] = \"3'UTR\"\n",
    "\n",
    "AllRegion_annotation = pd.concat([UTR_3_annotation,UTR_5_annotation,CDS_annotation]).reset_index(drop = True)\n",
    "\n",
    "AllRegion_annotation['length'] = ''\n",
    "AllRegion_annotation['GC'] = ''\n",
    "for i in range(len(AllRegion_annotation['gene_id'])):\n",
    "    AllRegion_annotation['length'][i] = len(AllRegion_annotation['seq'][i])\n",
    "    AllRegion_annotation['GC'][i] = (AllRegion_annotation['seq'][i].count('G') + AllRegion_annotation['seq'][i].count('C')) / len(AllRegion_annotation['seq'][i])\n",
    "AllRegion_annotation['GC'] = AllRegion_annotation['GC']*100\n",
    "AllRegion_annotation = pd.merge(AllRegion_annotation,Ttest_Cluster_selected,left_on = 'gene_id',right_on = 'gene').drop(['gene_id'],axis = 1)\n",
    "\n",
    "AllRegion_annotation[['gene','Region','length','GC','group']].to_csv((path2 + 'Cluster_Length_and_GC.txt'), sep = '\\t')\n",
    "AllRegion_annotation[['gene','Region','length','GC','group']]\n",
    "\n",
    "### Kozak Score\n",
    "\n",
    "KozakScore = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/APOBEC1_data/preview/3-NormalizeFeature/Kozak/initiation_score_6+3.txt',usecols = ['gene','score_xnt'])\n",
    "KozakScore.columns = ['gene','KozakScore']\n",
    "KozakScore = pd.merge(Ttest_Cluster,KozakScore,on = 'gene')\n",
    "KozakScore.to_csv((path2 + 'Cluster_KozakScore_6+3.txt'), sep = '\\t')\n",
    "\n",
    "### PolyA Length and nonA\n",
    "\n",
    "PolyA_length_nonA = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H_polyLength_nonAcontent.txt')\n",
    "PolyA_length_nonA = pd.merge(Ttest_Cluster,PolyA_length_nonA,left_on = 'gene',right_on = 'gene_id').drop(['gene_id','Unnamed: 0'], axis = 1)\n",
    "PolyA_length_nonA.to_csv((path2 + 'Cluster_PolyA.txt'), sep = '\\t')\n",
    "\n",
    "### uORF enrichment\n",
    "\n",
    "Ttest_Cluster_selected = Ttest_Cluster[Ttest_Cluster['group'] != 'Not_significant'].reset_index()\n",
    "Ttest_Cluster_selected = Ttest_Cluster_selected[['gene','group']]\n",
    "\n",
    "Ingolia_uORF_Gene = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uORF_Gene_FristPos_2009-ingolia.txt')\n",
    "Ingolia_uORF_GeneList = Ingolia_uORF_Gene['Gene'].unique()\n",
    "Cluster1_GeneList = Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster1']['gene'].unique()\n",
    "Cluster2_GeneList = Ttest_Cluster_selected[Ttest_Cluster_selected['group'] == 'Cluster2']['gene'].unique()\n",
    "print(\"Ingolia uORF: n = \", str(len(Ingolia_uORF_GeneList)))\n",
    "print(\"Cluster1: n = \", str(len(Cluster1_GeneList)))\n",
    "print(\"Cluster2: n = \", str(len(Cluster2_GeneList)))\n",
    "Cluster1_uORF_GeneList = list(set(Ingolia_uORF_GeneList).intersection(set(Cluster1_GeneList)))\n",
    "Cluster2_uORF_GeneList = list(set(Ingolia_uORF_GeneList).intersection(set(Cluster2_GeneList)))\n",
    "print(\"Cluster1 have uORF: n = \", str(len(Cluster1_uORF_GeneList)))\n",
    "print(\"Cluster2 have uORF: n = \", str(len(Cluster2_uORF_GeneList)))\n",
    "\n",
    "uAUG_U3 = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/5-InitiationFactor/1-TripletPeriodicity/uAUG_Gene-U3.txt')\n",
    "uAUG_U3_AG = uAUG_U3[(uAUG_U3['U3'] == 'A')|(uAUG_U3['U3'] == 'G')]\n",
    "U3_AG_GeneList = uAUG_U3_AG['gene'].unique()\n",
    "Cluster1_uORF_U3_AG_GeneList = list(set(U3_AG_GeneList).intersection(set(Cluster1_GeneList)))\n",
    "Cluster2_uORF_U3_AG_GeneList = list(set(U3_AG_GeneList).intersection(set(Cluster2_GeneList)))\n",
    "print(\"Cluster1 have uORF U3=A/G: n = \", str(len(Cluster1_uORF_U3_AG_GeneList)))\n",
    "print(\"Cluster2 have uORF U3=A/G: n = \", str(len(Cluster2_uORF_U3_AG_GeneList)))\n",
    "\n",
    "uAUG_U3_CT = uAUG_U3[(uAUG_U3['U3'] == 'C')|(uAUG_U3['U3'] == 'T')]\n",
    "U3_CT_GeneList = uAUG_U3_CT['gene'].unique()\n",
    "Cluster1_uORF_U3_CT_GeneList = list(set(U3_CT_GeneList).intersection(set(Cluster1_GeneList)))\n",
    "Cluster2_uORF_U3_CT_GeneList = list(set(U3_CT_GeneList).intersection(set(Cluster2_GeneList)))\n",
    "print(\"Cluster1 have uORF U3=C/T: n = \", str(len(Cluster1_uORF_U3_CT_GeneList)))\n",
    "print(\"Cluster2 have uORF U3=C/T: n = \", str(len(Cluster2_uORF_U3_CT_GeneList)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 6. Binding Affinity"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. Ribosome Density foldchange & Binding Affinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data/2-Binding_affinity/'\n",
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/'\n",
    "path3 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/6-BindingAffinity/'\n",
    "\n",
    "SampleList = ['CAF20','EAP1','eIF4E','eIF4G1','PAB1']\n",
    "\n",
    "for i in SampleList:\n",
    "    locals()['BindAffi_' + i] = pd.read_table((path + 'binding_affinity_' + i + '.txt'), usecols = ['genes','logFC','FDR'])\n",
    "    locals()['BindAffi_' + i].columns = ['gene','logFC','FDR']\n",
    "    locals()['BindAffi_' + i]['sample'] = i\n",
    "    locals()['BindAffi_' + i]['group'] = ''\n",
    "\n",
    "BindAffi = pd.concat([locals()['BindAffi_CAF20'],locals()['BindAffi_EAP1'],locals()['BindAffi_eIF4E'],\\\n",
    "    locals()['BindAffi_eIF4G1'],locals()['BindAffi_PAB1']]).reset_index(drop =  True)\n",
    "BindAffi.to_csv(path3 + 'binding_affinity.txt', sep = '\\t')\n",
    "\n",
    "AllSample_RibosomeDensity_fold = pd.read_table(path2 + 'TotalMutant-foldchange-rbind.txt', usecols = ['gene','sample','RNA_foldchange','RPF_foldchange','RibosomeDensity_foldchange'])\n",
    "AllSample_RibosomeDensity_fold = AllSample_RibosomeDensity_fold[AllSample_RibosomeDensity_fold['sample'].isin(SampleList)].reset_index(drop =  True)\n",
    "AllSample_RibosomeDensity_fold = pd.merge(BindAffi,AllSample_RibosomeDensity_fold,on = ['gene','sample'])\n",
    "AllSample_RibosomeDensity_fold.to_csv(path3 + 'TotalMutant-RNA_RPF_TE-Density-Fold-BindAffi.txt', sep = '\\t')\n",
    "AllSample_RibosomeDensity_fold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. PolyA length & PAB1 Binding Affinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>logFC</th>\n",
       "      <th>sample</th>\n",
       "      <th>group</th>\n",
       "      <th>poly_length_average</th>\n",
       "      <th>nonA_freq_average</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>YNR025C</td>\n",
       "      <td>2.379484</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>51.035985</td>\n",
       "      <td>0.001187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>YNR025C</td>\n",
       "      <td>2.093916</td>\n",
       "      <td>EAP1</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>51.035985</td>\n",
       "      <td>0.001187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>YNR025C</td>\n",
       "      <td>0.413868</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>51.035985</td>\n",
       "      <td>0.001187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>YOR183W</td>\n",
       "      <td>1.799892</td>\n",
       "      <td>CAF20</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>41.251667</td>\n",
       "      <td>0.001611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>YOR183W</td>\n",
       "      <td>0.809261</td>\n",
       "      <td>eIF4E</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>41.251667</td>\n",
       "      <td>0.001611</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5829</th>\n",
       "      <td>YCL041C</td>\n",
       "      <td>1.028989</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>36.613462</td>\n",
       "      <td>0.001857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5830</th>\n",
       "      <td>YCL041C</td>\n",
       "      <td>0.504321</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>36.613462</td>\n",
       "      <td>0.001857</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5831</th>\n",
       "      <td>YIL047C-A</td>\n",
       "      <td>0.850737</td>\n",
       "      <td>eIF4G1</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>47.437500</td>\n",
       "      <td>0.001423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5832</th>\n",
       "      <td>YIL047C-A</td>\n",
       "      <td>2.301891</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>47.437500</td>\n",
       "      <td>0.001423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5833</th>\n",
       "      <td>YJL135W</td>\n",
       "      <td>0.334676</td>\n",
       "      <td>PAB1</td>\n",
       "      <td>Not_Significant</td>\n",
       "      <td>42.574585</td>\n",
       "      <td>0.002107</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5834 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene     logFC  sample            group  poly_length_average  \\\n",
       "0       YNR025C  2.379484   CAF20  Not_Significant            51.035985   \n",
       "1       YNR025C  2.093916    EAP1  Not_Significant            51.035985   \n",
       "2       YNR025C  0.413868    PAB1  Not_Significant            51.035985   \n",
       "3       YOR183W  1.799892   CAF20  Not_Significant            41.251667   \n",
       "4       YOR183W  0.809261   eIF4E  Not_Significant            41.251667   \n",
       "...         ...       ...     ...              ...                  ...   \n",
       "5829    YCL041C  1.028989  eIF4G1  Not_Significant            36.613462   \n",
       "5830    YCL041C  0.504321    PAB1  Not_Significant            36.613462   \n",
       "5831  YIL047C-A  0.850737  eIF4G1  Not_Significant            47.437500   \n",
       "5832  YIL047C-A  2.301891    PAB1  Not_Significant            47.437500   \n",
       "5833    YJL135W  0.334676    PAB1  Not_Significant            42.574585   \n",
       "\n",
       "      nonA_freq_average  \n",
       "0              0.001187  \n",
       "1              0.001187  \n",
       "2              0.001187  \n",
       "3              0.001611  \n",
       "4              0.001611  \n",
       "...                 ...  \n",
       "5829           0.001857  \n",
       "5830           0.001857  \n",
       "5831           0.001423  \n",
       "5832           0.001423  \n",
       "5833           0.002107  \n",
       "\n",
       "[5834 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "BindAffi = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/6-BindingAffinity/binding_affinity.txt', usecols = ['gene','logFC','sample','group'])\n",
    "PolyA_Length = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H_polyLength_nonAcontent.txt')\n",
    "PolyA_Length = pd.merge(BindAffi,PolyA_Length,left_on = 'gene',right_on = 'gene_id').drop('gene_id', axis = 1)\n",
    "PolyA_Length.drop('Unnamed: 0', axis = 1).to_csv('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/6-BindingAffinity/PolyA_length_nonA_BindAffi.txt', sep = '\\t')\n",
    "PolyA_Length.drop('Unnamed: 0', axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 7. PolyA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. calculate polyA Length and non-A frequency"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/'\n",
    "name = 'H-1'\n",
    "\n",
    "PacbioGeneSeq = pd.read_table((path + name + '_mapping.txt'),usecols = ['strand','gene_id','cigar','seq'])\n",
    "PacbioGeneSeq_1 = PacbioGeneSeq[PacbioGeneSeq['strand'] == '+'].reset_index().drop(['index'], axis = 1)\n",
    "PacbioGeneSeq_1['poly'] = ''\n",
    "\n",
    "for i in range(len(PacbioGeneSeq_1['gene_id'])):\n",
    "    cigar = PacbioGeneSeq_1['cigar'][i]\n",
    "    soft_clip = cigar.split('=')[-1]\n",
    "    if \"S\" in soft_clip:\n",
    "        soft_clip_len = - int(soft_clip.strip('S'))\n",
    "        PacbioGeneSeq_1['poly'][i] = PacbioGeneSeq_1['seq'][i][soft_clip_len:]\n",
    "    else:\n",
    "        PacbioGeneSeq_1['poly'][i] = 'none'\n",
    "\n",
    "\n",
    "PacbioGeneSeq_0 = PacbioGeneSeq[PacbioGeneSeq['strand'] == '-'].reset_index().drop(['index'], axis = 1)\n",
    "PacbioGeneSeq_0['poly'] = ''\n",
    "\n",
    "for i in range(len(PacbioGeneSeq_0['gene_id'])):\n",
    "    cigar = PacbioGeneSeq_0['cigar'][i]\n",
    "    soft_clip = cigar.split('=')[0]\n",
    "    if \"S\" in soft_clip:\n",
    "        soft_clip = soft_clip.split('S')[0]\n",
    "        soft_clip_len = int(soft_clip)\n",
    "        poly_seq = PacbioGeneSeq_0['seq'][i][:soft_clip_len]\n",
    "        poly_seq = poly_seq[::-1].replace('A','t').replace('C','g').replace('G','c').replace('T','a').upper()\n",
    "        PacbioGeneSeq_0['poly'][i] = poly_seq\n",
    "    else:\n",
    "        PacbioGeneSeq_0['poly'][i] = 'none'\n",
    "\n",
    "PacbioGeneSeq = pd.concat([PacbioGeneSeq_1,PacbioGeneSeq_0]).reset_index().drop(['index'], axis = 1)\n",
    "PacbioGeneSeq['poly_length'] = PacbioGeneSeq['poly'].apply(len)\n",
    "PacbioGeneSeq.to_csv((path + name + '_polySeq.txt'), sep = '\\t')\n",
    "\n",
    "PacbioGeneSeq_step2 = PacbioGeneSeq[['gene_id','poly','poly_length']]\n",
    "\n",
    "PacbioGeneSeq_step2['polyA_A_freq'] = ''\n",
    "PacbioGeneSeq_step2['polyA_nonA_count'] = ''\n",
    "\n",
    "for i in range(len(PacbioGeneSeq_step2['poly'])):\n",
    "    PolyA_seq = PacbioGeneSeq_step2['poly'][i]\n",
    "    PolyA_len = PacbioGeneSeq_step2['poly_length'][i]\n",
    "    PacbioGeneSeq_step2['polyA_A_freq'][i] = PolyA_seq.count('A') / PolyA_len\n",
    "    PacbioGeneSeq_step2['polyA_nonA_count'][i] = (PolyA_seq.count('G') + PolyA_seq.count('C') + PolyA_seq.count('T'))\n",
    "\n",
    "PacbioGeneSeq_cutoff = PacbioGeneSeq_step2[(PacbioGeneSeq_step2['poly_length'] >= 15) & (PacbioGeneSeq['poly_length'] <= 500) & \\\n",
    "    (PacbioGeneSeq_step2['polyA_A_freq'] >= 0.5) & (PacbioGeneSeq_step2['polyA_nonA_count'] <= 20)].reset_index().drop(['index'], axis = 1)\n",
    "\n",
    "num_df = PacbioGeneSeq_cutoff['gene_id'].value_counts()\n",
    "num_df = num_df.reset_index()\n",
    "num_df.columns = ['gene','num']\n",
    "num_df = num_df[num_df['num'] >= 10].reset_index().drop(['index'], axis = 1)\n",
    "gene = num_df['gene'].unique()\n",
    "\n",
    "PacbioGeneSeq_cutoff = PacbioGeneSeq_cutoff[PacbioGeneSeq_cutoff['gene_id'].isin(gene)].reset_index().drop(['index'], axis = 1)\n",
    "PacbioGeneSeq_cutoff['nonA_freq'] = 1 - PacbioGeneSeq_cutoff['polyA_A_freq']\n",
    "PacbioGeneSeq_cutoff.to_csv((path + name + '_polySeq_before_average.txt'), sep = '\\t')\n",
    "\n",
    "length_df = PacbioGeneSeq_cutoff[['gene_id','poly_length']].groupby(['gene_id']).mean()\n",
    "nonA_df = PacbioGeneSeq_cutoff[['gene_id','nonA_freq']].groupby(['gene_id']).mean()\n",
    "\n",
    "df_length_nonA = pd.merge(length_df,nonA_df,on = 'gene_id')\n",
    "df_length_nonA.to_csv((path + name + '_polyLength_nonAcontent.txt'), sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample1 = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H-1_polyLength_nonAcontent.txt')\n",
    "sample2 = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H-2_polyLength_nonAcontent.txt')\n",
    "sample3 = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H-3_polyLength_nonAcontent.txt')\n",
    "\n",
    "total_1 = pd.merge(sample1,sample2,on = 'gene_id')\n",
    "total = pd.merge(total_1,sample3,on = 'gene_id')\n",
    "\n",
    "total['poly_length_average'] = (total['poly_length_x'] + total['poly_length_y'] + total['poly_length']) / 3\n",
    "total['nonA_freq_average'] = (total['nonA_freq_x'] + total['nonA_freq_y'] + total['nonA_freq']) / 3\n",
    "total = total[['gene_id','poly_length_average','nonA_freq_average']]\n",
    "total.to_csv('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H_polyLength_nonAcontent.txt', sep = '\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### b. polyA Length and Ribosome Density(fold change)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/'\n",
    "path3 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/7-PolyA/'\n",
    "\n",
    "polyA_length = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H_polyLength_nonAcontent.txt', usecols = ['gene_id','poly_length_average'])\n",
    "#polyA_length = pd.read_csv('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/7-PolyA/paper_data/NormalizedCountsMeanMedianTable1.csv', usecols = ['Systematic.Name','WT.BY.mean'])\n",
    "polyA_length.columns = ['gene_id','poly_length_average']\n",
    "\n",
    "polyA_length_sorted = polyA_length.sort_values('poly_length_average').reset_index(drop = True)\n",
    "num_bins = 8\n",
    "bin_size = len(polyA_length_sorted) // num_bins\n",
    "\n",
    "bins = [i * bin_size for i in range(num_bins + 1)]\n",
    "bin_labels = [f'bin{i + 1}' for i in range(num_bins)]\n",
    "polyA_length_sorted['bin_label'] = pd.cut(polyA_length_sorted.index, bins=bins, labels=bin_labels, right=False)\n",
    "polyA_length_sorted = polyA_length_sorted.fillna(bin_labels[-1])\n",
    "\n",
    "# Ribosome Density\n",
    "Allsample_RibosomeDensity = pd.read_table((path2 + 'TotalMutant-foldchange-rbind.txt'), usecols = ['gene','Ribosome_Density','sample'])\n",
    "\n",
    "polyA_length_sorted_1 = pd.merge(polyA_length_sorted,Allsample_RibosomeDensity,left_on = 'gene_id',right_on = 'gene').drop('gene_id', axis = 1)\n",
    "polyA_length_sorted_1.to_csv((path3 + 'PolyALength_and_RibosomeDensity.txt'), sep = '\\t')\n",
    "\n",
    "# Ribosome Density foldchange\n",
    "Allsample_RibosomeDensity_fold = pd.read_table((path2 + 'TotalMutant-foldchange-rbind.txt'), usecols = ['gene','RibosomeDensity_foldchange','sample'])\n",
    "Allsample_RibosomeDensity_fold = Allsample_RibosomeDensity_fold[~Allsample_RibosomeDensity_fold.isin([np.nan, np.inf, -np.inf]).any(1)].reset_index().drop(['index'], axis = 1)\n",
    "polyA_length_sorted_2 = pd.merge(polyA_length_sorted,Allsample_RibosomeDensity_fold,left_on = 'gene_id',right_on = 'gene').drop(['gene_id'], axis = 1)\n",
    "polyA_length_sorted_2.to_csv((path3 + 'PolyALength_and_RibosomeDensity-fold.txt'), sep = '\\t')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### c. polyA Length and nonA content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "46.08437037379161"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "polyA_length_nonA = pd.read_table('/Dell/Dell15/zhanggy/project/3-mRNA_looping/tmp/Pacbio_data/PolyA_analyze/H_polyLength_nonAcontent.txt')\n",
    "np.mean(polyA_length_nonA['poly_length_average'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 8. mRNA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### a. CDS Length ~ Ribosome Density(fold change)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "path2 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/3-RibosomeDensity/Ribosome_Density/'\n",
    "path3 = '/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/Ribo-seq_data_final/9-mRNA/'\n",
    "\n",
    "CDS_length = pd.read_table('/Dell/Dell15/zhanggy/project/4-Ribo-seq-data/tmp/information/CDS_seq.txt', usecols = ['gene_id','seq'])\n",
    "CDS_length['length'] = CDS_length['seq'].str.len()\n",
    "CDS_length = CDS_length[['gene_id','length']]\n",
    "CDS_length.columns = ['gene_id','length']\n",
    "\n",
    "CDS_length_sorted = CDS_length.sort_values('length').reset_index(drop = True)\n",
    "num_bins = 8\n",
    "bin_size = len(CDS_length_sorted) // num_bins\n",
    "\n",
    "bins = [i * bin_size for i in range(num_bins + 1)]\n",
    "bin_labels = [f'bin{i + 1}' for i in range(num_bins)]\n",
    "CDS_length_sorted['bin_label'] = pd.cut(CDS_length_sorted.index, bins=bins, labels=bin_labels, right=False)\n",
    "CDS_length_sorted = CDS_length_sorted.fillna(bin_labels[-1])\n",
    "\n",
    "# Ribosome Density\n",
    "Allsample_RibosomeDensity = pd.read_table((path2 + 'TotalMutant-foldchange-rbind.txt'), usecols = ['gene','Ribosome_Density','sample'])\n",
    "\n",
    "CDS_length_sorted_1 = pd.merge(CDS_length_sorted,Allsample_RibosomeDensity,left_on = 'gene_id',right_on = 'gene').drop('gene_id', axis = 1)\n",
    "CDS_length_sorted_1.to_csv((path3 + 'CDSLength_and_RibosomeDensity.txt'), sep = '\\t')\n",
    "\n",
    "# Ribosome Density foldchange\n",
    "Allsample_RibosomeDensity_fold = pd.read_table((path2 + 'TotalMutant-foldchange-rbind.txt'), usecols = ['gene','RibosomeDensity_foldchange','sample'])\n",
    "Allsample_RibosomeDensity_fold = Allsample_RibosomeDensity_fold[~Allsample_RibosomeDensity_fold.isin([np.nan, np.inf, -np.inf]).any(1)].reset_index().drop(['index'], axis = 1)\n",
    "CDS_length_sorted_2 = pd.merge(CDS_length_sorted,Allsample_RibosomeDensity_fold,left_on = 'gene_id',right_on = 'gene').drop(['gene_id'], axis = 1)\n",
    "CDS_length_sorted_2.to_csv((path3 + 'CDSLength_and_RibosomeDensity-fold.txt'), sep = '\\t')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "8382f34c52c9aef1025ee680a4004b675487cb355e47984d0f24a909fc4bbbb3"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
